{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "8f4264d7",
   "metadata": {},
   "source": [
    "## PERMGWAS \n",
    "In the method obtained as a result of the GWAS analysis, it was thought that the false positive SNPs were more than expected. It is thought that the presence of too many regulator genes in the network map to be created and the false positive error being more than expected. For this reason, using permgwas instead of classical bonferroni correction, other filtering steps were repeated according to these new permGWAS results. In this study, the results before repeating the prefiletering, which was done before the GWAS analysis in the previous method, were used. In this way, permGWAS analysis was performed with significant 738 phenotype files. The phenotype and genotype files used are the same, and permGWAS does not need a kinship file. The permGWAS used in this study is an open source tool, downloaded from https://github.com/grimmlab/permGWAS.git and detailed information can be found here. After the necessary tools are downloaded, the phenotype and genotype file locations are specified, the appropriate permutation number (perm has been determined as 100 for this study) is entered and python is run in the batch environment. In this way, PermGWAS analysis is done."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "323006d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "FILES=(AT*.csv)\n",
    "FILE=${FILES[$SLURM_ARRAY_TASK_ID]}\n",
    "singularity exec --bind /storage/full-share/permGWAS permgwas.sing python3 permGWAS/permGWAS.py -x F95.h5 -y ${FILE}  --out_dir outfile_${FILE} --perm 100\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6de5d39",
   "metadata": {},
   "source": [
    "### 1. PermGWAS Threshold \n",
    "In this section, new thresold are found for each of the perm GWAS results seperatly. The reason of the the using different threshold each of the result is reduce false positive ratio, compare to the bonferroni correction.Every min p value file contain only first minimum 100 p values. Min p values are reordered accorging to their p values. After that cut off value are decided to 5 th of the p value (0.05. These p values are decided each of the permGWAS result seperatly. Each of the thresholds are used to filter permGWAS results and saved (filter_permgwas.r)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c85d3b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "##permgwas:find pvalue_threshold each phenotype separetly then apply this threshold value for filtering step\n",
    "library(tidyverse)\n",
    "library(fs)\n",
    "setwd(\"~/lastversion_gwas_analysis/PermGWAS_Scripts/\")\n",
    "\n",
    "list_of_files <- list.files(path = \"~/PermGWAS/\",\n",
    "                            recursive = TRUE,\n",
    "                            pattern = \"*min_\",\n",
    "                            full.names = TRUE)\n",
    "Y <- lapply(list_of_files, function(x) {\n",
    "  file <- read.csv(x)\n",
    "  return(file)\n",
    "})\n",
    "###find perm_threshold for each phenotype\n",
    "new_list<-vector()\n",
    "for (i in 1:length((list_of_files))){\n",
    "  A<-Y[[i]][order(Y[[i]][,2]),]\n",
    "  perm_thres<-A[5,2]\n",
    "  new_list<-c(new_list,perm_thres)\n",
    "  new_list<-as.matrix(new_list, sep=\",\")\n",
    "  new<-as.data.frame(new_list)\n",
    "}\n",
    "####use this threshold and filter perm_gwas result\n",
    "\n",
    "csv_files <- list.files(path = \"~/PermGWAS/\",\n",
    "                            recursive = TRUE,\n",
    "                            pattern = \"*gwasp_\",\n",
    "                            full.names = TRUE)\n",
    "Y_ <- lapply(csv_files, function(x) {\n",
    "  file <- read.csv(x)\n",
    "  return(file)\n",
    "})\n",
    "a<-read.csv(\"~/lastversion_gwas_analysis/names.csv\")\n",
    "for (i in 1:length((csv_files))){\n",
    "  R<-subset(Y_[[i]],Y_[[i]][[\"p_value\"]]<(new[i,]))\n",
    "  write.csv(R, file=paste(a$x[i] , sep=\"\"))\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "681f74b2",
   "metadata": {},
   "source": [
    "### 2. MAF filter\n",
    "In this part, MAF (Minnor Allele Frequency) value is less than 5 is discarded from insdide of the all files and saved. Minnor Allele Frequency is the frequency that second most found allele occured in a population, studies showed that it has a major affect on traits (). The rest of the steps are continue with this new filtered files (mac_maf.r)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "820e9d22",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "#maf filtered >0.05\n",
    "\n",
    "library(tidyverse)\n",
    "library(fs)\n",
    "setwd(\"~/lastversion_gwas_analysis/PermGWAS_Scripts/permgwas_filtered_pvalue/\")\n",
    "\n",
    "Y <- list()\n",
    "csvfiles<-list.files(pattern = \"*.csv\")\n",
    "\n",
    "Y <- lapply(csvfiles, function(x) {\n",
    "  file <- read.csv(x)\n",
    "  return(file)\n",
    "})\n",
    "\n",
    "for (i in 1:length((list_of_files))){\n",
    "  R<-subset(Y[[i]],Y[[i]][[\"maf\"]]>0.05)\n",
    "  write.csv(R, file=paste( \"filtered_\",csvfiles[i],\".csv\", sep=\"\"))\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d9dadbc",
   "metadata": {},
   "source": [
    "### Count hits\n",
    "Stringent bonferroni correction files are used to count cis and trans hits. Each stringent bonferroni corrected files are used. The second gene name is taken from the name of these files. Each SNPs position given in these files is valued according to the Start and Stop positions in the gene location file. If these SNPs positions are between the Start and Stop position, the SNPs at that position are called cis_hits, otherwise they are called trans hits. In this way, the number of cis hits for each file is subtracted from the total number of SNPs (hits) and the number of trans hits is found.\n",
    "### Count cis and trans hits\n",
    "After counting cis hits, trans hits counted to sustraction of total hits to cis hits. New column added to the list as when the file contain cis hits it represented with 1, othervise 0. New column added to the list as when the file contain trans hits it represented with 1, othervise 0. Last column added to the list as when both trans and cis hits founded in the same file, it represented with 1, othervise 0."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "dadaaef4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import sys\n",
    "import glob"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c141a143",
   "metadata": {},
   "source": [
    "#### Number of Cis Hits\n",
    "\n",
    "Stringent bonferroni correction files are used to count cis and trans hits. Each stringent bonferroni corrected files are used. The second gene name is taken from the name of these files. Each SNPs position given in these files is valued according to the Start and Stop positions in the gene location file. If these SNPs positions are between the Start and Stop position, the SNPs at that position are called cis_hits, otherwise they are called trans hits. In this way, the number of cis hits for each file is subtracted from the total number of SNPs (hits) and the number of trans hits is found (cis_trans.r)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9a87255",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "##count cis hits\n",
    "library(tidyverse)\n",
    "library(fs)\n",
    "setwd(\"~/lastversion_gwas_analysis/Over_Occ/stringent_correction_738/\")\n",
    "ara11<-read.csv(\"~/lastversion_gwas_analysis/ara11_50.csv\")\n",
    "\n",
    "small_gene<-read.csv(\"~/lastversion_gwas_analysis/Over_Occ/Small_gene.csv\")\n",
    "\n",
    "Y <- list()\n",
    "csvfiles<-list.files(pattern = \"*.csv\")\n",
    "Y <- lapply(csvfiles, function(x) {\n",
    "  file <- read.csv(x)\n",
    "  return(file)\n",
    "})\n",
    "\n",
    "new_list<-list()\n",
    "for (i in 1:length((csvfiles))){\n",
    "  b<-ara11[which(ara11$Gene==small_gene$second_gene_name[i]),]\n",
    "  c<-sum(b$Start<Y[[i]][[\"Pos\"]] & Y[[i]][[\"Pos\"]]<b$Stop)\n",
    "  new_list<-c(new_list,c)\n",
    "  new_list<-as.matrix(new_list, sep=\",\")\n",
    "  write.csv( new_list, \"small_gene_cis.csv\")\n",
    "}\n",
    "####change the folder name########\n",
    "folder = \"~/lastversion_gwas_analysis/PermGWAS_Scripts/deneme2//\"\n",
    "files <- list.files(folder,pattern = \"trans_.*.csv\",full.names = T) \n",
    "sapply(files,FUN=function(eachPath){ \n",
    "  file.rename(from=eachPath,to= sub(pattern=\"trans_\", paste0(\"\"),eachPath))\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "62165cc4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_name</th>\n",
       "      <th>second_gene_name</th>\n",
       "      <th>total_hit</th>\n",
       "      <th>cis_hits</th>\n",
       "      <th>trans_hits</th>\n",
       "      <th>cis_1</th>\n",
       "      <th>trans_1</th>\n",
       "      <th>cis_trans_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT1G04380_AT2G27340.csv</td>\n",
       "      <td>AT2G27340</td>\n",
       "      <td>782</td>\n",
       "      <td>71</td>\n",
       "      <td>711</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT1G04380_AT4G01925.csv</td>\n",
       "      <td>AT4G01925</td>\n",
       "      <td>234</td>\n",
       "      <td>84</td>\n",
       "      <td>150</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT1G04380_AT4G04220.csv</td>\n",
       "      <td>AT4G04220</td>\n",
       "      <td>806</td>\n",
       "      <td>189</td>\n",
       "      <td>617</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT1G04380_AT5G51795.csv</td>\n",
       "      <td>AT5G51795</td>\n",
       "      <td>2067</td>\n",
       "      <td>70</td>\n",
       "      <td>1997</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT1G11180_AT1G29710.csv</td>\n",
       "      <td>AT1G29710</td>\n",
       "      <td>116</td>\n",
       "      <td>79</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>733</th>\n",
       "      <td>AT5G66630_AT5G20450.csv</td>\n",
       "      <td>AT5G20450</td>\n",
       "      <td>26</td>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>734</th>\n",
       "      <td>AT5G66630_AT5G27110.csv</td>\n",
       "      <td>AT5G27110</td>\n",
       "      <td>387</td>\n",
       "      <td>386</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>735</th>\n",
       "      <td>AT5G66630_AT5G45150.csv</td>\n",
       "      <td>AT5G45150</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>736</th>\n",
       "      <td>AT5G66630_AT5G48350.csv</td>\n",
       "      <td>AT5G48350</td>\n",
       "      <td>26</td>\n",
       "      <td>10</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>737</th>\n",
       "      <td>AT5G66630_AT5G54020.csv</td>\n",
       "      <td>AT5G54020</td>\n",
       "      <td>136</td>\n",
       "      <td>135</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>738 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   file_name second_gene_name  total_hit  cis_hits  \\\n",
       "0    AT1G04380_AT2G27340.csv        AT2G27340        782        71   \n",
       "1    AT1G04380_AT4G01925.csv        AT4G01925        234        84   \n",
       "2    AT1G04380_AT4G04220.csv        AT4G04220        806       189   \n",
       "3    AT1G04380_AT5G51795.csv        AT5G51795       2067        70   \n",
       "4    AT1G11180_AT1G29710.csv        AT1G29710        116        79   \n",
       "..                       ...              ...        ...       ...   \n",
       "733  AT5G66630_AT5G20450.csv        AT5G20450         26        26   \n",
       "734  AT5G66630_AT5G27110.csv        AT5G27110        387       386   \n",
       "735  AT5G66630_AT5G45150.csv        AT5G45150          9         9   \n",
       "736  AT5G66630_AT5G48350.csv        AT5G48350         26        10   \n",
       "737  AT5G66630_AT5G54020.csv        AT5G54020        136       135   \n",
       "\n",
       "     trans_hits  cis_1  trans_1  cis_trans_1  \n",
       "0           711      1        1          1.0  \n",
       "1           150      1        1          1.0  \n",
       "2           617      1        1          1.0  \n",
       "3          1997      1        1          1.0  \n",
       "4            37      1        1          1.0  \n",
       "..          ...    ...      ...          ...  \n",
       "733           0      1        0          0.0  \n",
       "734           1      1        1          1.0  \n",
       "735           0      1        0          0.0  \n",
       "736          16      1        1          1.0  \n",
       "737           1      1        1          1.0  \n",
       "\n",
       "[738 rows x 8 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "##add total SNPs hits, cis hits and trans hits as a new cloumns into result(Small_gene.csv) then sum only trans only cis and non hits\n",
    "os.chdir('~/lastversion_gwas_analysis/PermGWAS_Scripts/maf_filtered/')\n",
    "total_hits = pd.DataFrame(columns=('file_name', 'total_hits'))\n",
    "\n",
    "for index,i in enumerate(os.listdir('.')):\n",
    "    total_hits.loc[index] = [i,len(pd.read_csv(i).index)]\n",
    "\n",
    "total_hits=total_hits.sort_values(\"file_name\")\n",
    "total_hits\n",
    "total_hits.to_csv('~/lastversion_gwas_analysis/PermGWAS_Scripts/total_hit.csv', index=False)\n",
    "result=pd.read_csv(\"~/lastversion_gwas_analysis/Over_Occ/Small_gene.csv\")\n",
    "total_hits=pd.read_csv(\"~/lastversion_gwas_analysis/PermGWAS_Scripts/total_hit.csv\")\n",
    "total_hits=total_hits[\"total_hits\"]\n",
    "result[\"total_hit\"]=total_hits\n",
    "\n",
    "cis_hits=pd.read_csv(\"~/lastversion_gwas_analysis/PermGWAS_Scripts/small_gene_cis.csv\")\n",
    "cis_hits=cis_hits[\"V1\"]\n",
    "cis_hits\n",
    "result[\"cis_hits\"]=cis_hits\n",
    "result\n",
    "result[\"trans_hits\"]=result[\"total_hit\"]-result[\"cis_hits\"]\n",
    "result\n",
    "##add new column if cis_hits>0 and trans_hits>0:\n",
    "result[\"cis_1\"]=np.where(result['cis_hits'] >0, 1, 0)\n",
    "result[\"trans_1\"]=np.where(result['trans_hits'] >0, 1, 0)\n",
    "result.loc[(result['cis_hits'] > 0)  & (result['trans_hits'] >0), 'cis_trans_1'] = 1\n",
    "result.loc[(result['cis_hits'] <= 0) | (result['trans_hits'] <=0), \"cis_trans_1\"] = 0\n",
    "result.to_csv('~/lastversion_gwas_analysis/PermGWAS_Scripts/cis_trans_hit.csv', index=False)\n",
    "result"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4353a367",
   "metadata": {},
   "source": [
    "#### Filter \n",
    "The gene AT2G15930 is not found on ara11 notation file that cause a problem to count allthe SNPs as a trans even some of them cis gits. It is inside tair10 file but all our analysis done for ara11 so file which contain as a second gene \"AT2G15930\" discarded to continue of this study. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3488ddd4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_name</th>\n",
       "      <th>second_gene_name</th>\n",
       "      <th>total_hit</th>\n",
       "      <th>cis_hits</th>\n",
       "      <th>trans_hits</th>\n",
       "      <th>cis_1</th>\n",
       "      <th>trans_1</th>\n",
       "      <th>cis_trans_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT1G04380_AT2G27340.csv</td>\n",
       "      <td>AT2G27340</td>\n",
       "      <td>782</td>\n",
       "      <td>71</td>\n",
       "      <td>711</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT1G04380_AT4G01925.csv</td>\n",
       "      <td>AT4G01925</td>\n",
       "      <td>234</td>\n",
       "      <td>84</td>\n",
       "      <td>150</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT1G04380_AT4G04220.csv</td>\n",
       "      <td>AT4G04220</td>\n",
       "      <td>806</td>\n",
       "      <td>189</td>\n",
       "      <td>617</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT1G04380_AT5G51795.csv</td>\n",
       "      <td>AT5G51795</td>\n",
       "      <td>2067</td>\n",
       "      <td>70</td>\n",
       "      <td>1997</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT1G11180_AT1G29710.csv</td>\n",
       "      <td>AT1G29710</td>\n",
       "      <td>116</td>\n",
       "      <td>79</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>733</th>\n",
       "      <td>AT5G66630_AT5G20450.csv</td>\n",
       "      <td>AT5G20450</td>\n",
       "      <td>26</td>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>734</th>\n",
       "      <td>AT5G66630_AT5G27110.csv</td>\n",
       "      <td>AT5G27110</td>\n",
       "      <td>387</td>\n",
       "      <td>386</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>735</th>\n",
       "      <td>AT5G66630_AT5G45150.csv</td>\n",
       "      <td>AT5G45150</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>736</th>\n",
       "      <td>AT5G66630_AT5G48350.csv</td>\n",
       "      <td>AT5G48350</td>\n",
       "      <td>26</td>\n",
       "      <td>10</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>737</th>\n",
       "      <td>AT5G66630_AT5G54020.csv</td>\n",
       "      <td>AT5G54020</td>\n",
       "      <td>136</td>\n",
       "      <td>135</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>724 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   file_name second_gene_name  total_hit  cis_hits  \\\n",
       "0    AT1G04380_AT2G27340.csv        AT2G27340        782        71   \n",
       "1    AT1G04380_AT4G01925.csv        AT4G01925        234        84   \n",
       "2    AT1G04380_AT4G04220.csv        AT4G04220        806       189   \n",
       "3    AT1G04380_AT5G51795.csv        AT5G51795       2067        70   \n",
       "4    AT1G11180_AT1G29710.csv        AT1G29710        116        79   \n",
       "..                       ...              ...        ...       ...   \n",
       "733  AT5G66630_AT5G20450.csv        AT5G20450         26        26   \n",
       "734  AT5G66630_AT5G27110.csv        AT5G27110        387       386   \n",
       "735  AT5G66630_AT5G45150.csv        AT5G45150          9         9   \n",
       "736  AT5G66630_AT5G48350.csv        AT5G48350         26        10   \n",
       "737  AT5G66630_AT5G54020.csv        AT5G54020        136       135   \n",
       "\n",
       "     trans_hits  cis_1  trans_1  cis_trans_1  \n",
       "0           711      1        1          1.0  \n",
       "1           150      1        1          1.0  \n",
       "2           617      1        1          1.0  \n",
       "3          1997      1        1          1.0  \n",
       "4            37      1        1          1.0  \n",
       "..          ...    ...      ...          ...  \n",
       "733           0      1        0          0.0  \n",
       "734           1      1        1          1.0  \n",
       "735           0      1        0          0.0  \n",
       "736          16      1        1          1.0  \n",
       "737           1      1        1          1.0  \n",
       "\n",
       "[724 rows x 8 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "##files does not contain gene \"AT2G15930\" filtered to continue next steps\n",
    "#filter=\"AT2G15930\"\n",
    "filtered_files=(result.loc[result['second_gene_name'] != \"AT2G15930\"])\n",
    "filtered_files=filtered_files.sort_values(\"file_name\")\n",
    "filtered_files.to_csv('~/lastversion_gwas_analysis/PermGWAS_Scripts/cis_trans_filtered_hit.csv', index=False)\n",
    "filtered_files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a4573b5",
   "metadata": {},
   "source": [
    "#### Count hits \n",
    "In this section only cis, only trans, both trans and cis hits contains file, and non hit contain number of files are counted."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "cb53bfc4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hits           hit_sums\n",
      "-------------  -----------\n",
      "cis_only       62.0\n",
      "trans_only     7.0\n",
      "cis_trans_sum  638.0\n",
      "non_hit        17.0\n",
      "------------   -----------\n",
      "total_hit      724\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from tabulate import tabulate\n",
    "result=pd.read_csv(\"~/lastversion_gwas_analysis/PermGWAS_Scripts/cis_trans_filtered_hit.csv\")\n",
    "\n",
    "hit_sum_result=[]\n",
    "cis_sum=result.iloc[1:724, 5].sum()\n",
    "trans_sum=result.iloc[1:724, 6].sum()\n",
    "cis_trans_sum=result.iloc[1:724, 7].sum()\n",
    "only_cis=cis_sum-cis_trans_sum\n",
    "only_trans=trans_sum-cis_trans_sum\n",
    "non_hit=724-only_cis-only_trans-cis_trans_sum\n",
    "hit_sum_result=[\"cis_only\",only_cis],[\"trans_only\",only_trans],[\"cis_trans_sum\",cis_trans_sum],[\"non_hit\",non_hit],[\"------------\",\"-----------\"],[\"total_hit\",724]\n",
    "hit_sum_result\n",
    "\n",
    "col_names=[\"hits\",\"hit_sums\"]\n",
    "print(tabulate(hit_sum_result, headers=col_names))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "d253639d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_name</th>\n",
       "      <th>second_gene_name</th>\n",
       "      <th>total_hit</th>\n",
       "      <th>cis_hits</th>\n",
       "      <th>trans_hits</th>\n",
       "      <th>cis_1</th>\n",
       "      <th>trans_1</th>\n",
       "      <th>cis_trans_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT1G04380_AT2G27340.csv</td>\n",
       "      <td>AT2G27340</td>\n",
       "      <td>824</td>\n",
       "      <td>71</td>\n",
       "      <td>753</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT1G04380_AT4G01925.csv</td>\n",
       "      <td>AT4G01925</td>\n",
       "      <td>234</td>\n",
       "      <td>84</td>\n",
       "      <td>150</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT1G04380_AT4G04220.csv</td>\n",
       "      <td>AT4G04220</td>\n",
       "      <td>806</td>\n",
       "      <td>189</td>\n",
       "      <td>617</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT1G04380_AT5G51795.csv</td>\n",
       "      <td>AT5G51795</td>\n",
       "      <td>2067</td>\n",
       "      <td>70</td>\n",
       "      <td>1997</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT1G11180_AT1G29710.csv</td>\n",
       "      <td>AT1G29710</td>\n",
       "      <td>116</td>\n",
       "      <td>79</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>719</th>\n",
       "      <td>AT5G66630_AT5G20450.csv</td>\n",
       "      <td>AT5G20450</td>\n",
       "      <td>84</td>\n",
       "      <td>26</td>\n",
       "      <td>58</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>720</th>\n",
       "      <td>AT5G66630_AT5G27110.csv</td>\n",
       "      <td>AT5G27110</td>\n",
       "      <td>387</td>\n",
       "      <td>386</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>721</th>\n",
       "      <td>AT5G66630_AT5G45150.csv</td>\n",
       "      <td>AT5G45150</td>\n",
       "      <td>57</td>\n",
       "      <td>9</td>\n",
       "      <td>48</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>722</th>\n",
       "      <td>AT5G66630_AT5G48350.csv</td>\n",
       "      <td>AT5G48350</td>\n",
       "      <td>50</td>\n",
       "      <td>10</td>\n",
       "      <td>40</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>723</th>\n",
       "      <td>AT5G66630_AT5G54020.csv</td>\n",
       "      <td>AT5G54020</td>\n",
       "      <td>136</td>\n",
       "      <td>135</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>724 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   file_name second_gene_name  total_hit  cis_hits  \\\n",
       "0    AT1G04380_AT2G27340.csv        AT2G27340        824        71   \n",
       "1    AT1G04380_AT4G01925.csv        AT4G01925        234        84   \n",
       "2    AT1G04380_AT4G04220.csv        AT4G04220        806       189   \n",
       "3    AT1G04380_AT5G51795.csv        AT5G51795       2067        70   \n",
       "4    AT1G11180_AT1G29710.csv        AT1G29710        116        79   \n",
       "..                       ...              ...        ...       ...   \n",
       "719  AT5G66630_AT5G20450.csv        AT5G20450         84        26   \n",
       "720  AT5G66630_AT5G27110.csv        AT5G27110        387       386   \n",
       "721  AT5G66630_AT5G45150.csv        AT5G45150         57         9   \n",
       "722  AT5G66630_AT5G48350.csv        AT5G48350         50        10   \n",
       "723  AT5G66630_AT5G54020.csv        AT5G54020        136       135   \n",
       "\n",
       "     trans_hits  cis_1  trans_1  cis_trans_1  \n",
       "0           753      1        1          1.0  \n",
       "1           150      1        1          1.0  \n",
       "2           617      1        1          1.0  \n",
       "3          1997      1        1          1.0  \n",
       "4            37      1        1          1.0  \n",
       "..          ...    ...      ...          ...  \n",
       "719          58      1        1          1.0  \n",
       "720           1      1        1          1.0  \n",
       "721          48      1        1          1.0  \n",
       "722          40      1        1          1.0  \n",
       "723           1      1        1          1.0  \n",
       "\n",
       "[724 rows x 8 columns]"
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df=pd.read_csv(\"~/lastversion_gwas_analysis/PermGWAS_Scripts/cis_trans_filtered_hit.csv\")\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c12ce27c",
   "metadata": {},
   "source": [
    "#### Filter Trans hits\n",
    "In this section only files contain trans hits are filtered and it will continue rest of the anlysis with these files. In GWAS analysis trans hits is important role on trait"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "eb2dc0c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_name</th>\n",
       "      <th>second_gene_name</th>\n",
       "      <th>total_hit</th>\n",
       "      <th>cis_hits</th>\n",
       "      <th>trans_hits</th>\n",
       "      <th>cis_1</th>\n",
       "      <th>trans_1</th>\n",
       "      <th>cis_trans_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT1G04380_AT2G27340.csv</td>\n",
       "      <td>AT2G27340</td>\n",
       "      <td>824</td>\n",
       "      <td>71</td>\n",
       "      <td>753</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT1G04380_AT4G01925.csv</td>\n",
       "      <td>AT4G01925</td>\n",
       "      <td>234</td>\n",
       "      <td>84</td>\n",
       "      <td>150</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT1G04380_AT4G04220.csv</td>\n",
       "      <td>AT4G04220</td>\n",
       "      <td>806</td>\n",
       "      <td>189</td>\n",
       "      <td>617</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT1G04380_AT5G51795.csv</td>\n",
       "      <td>AT5G51795</td>\n",
       "      <td>2067</td>\n",
       "      <td>70</td>\n",
       "      <td>1997</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT1G11180_AT1G29710.csv</td>\n",
       "      <td>AT1G29710</td>\n",
       "      <td>116</td>\n",
       "      <td>79</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>719</th>\n",
       "      <td>AT5G66630_AT5G20450.csv</td>\n",
       "      <td>AT5G20450</td>\n",
       "      <td>84</td>\n",
       "      <td>26</td>\n",
       "      <td>58</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>720</th>\n",
       "      <td>AT5G66630_AT5G27110.csv</td>\n",
       "      <td>AT5G27110</td>\n",
       "      <td>387</td>\n",
       "      <td>386</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>721</th>\n",
       "      <td>AT5G66630_AT5G45150.csv</td>\n",
       "      <td>AT5G45150</td>\n",
       "      <td>57</td>\n",
       "      <td>9</td>\n",
       "      <td>48</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>722</th>\n",
       "      <td>AT5G66630_AT5G48350.csv</td>\n",
       "      <td>AT5G48350</td>\n",
       "      <td>50</td>\n",
       "      <td>10</td>\n",
       "      <td>40</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>723</th>\n",
       "      <td>AT5G66630_AT5G54020.csv</td>\n",
       "      <td>AT5G54020</td>\n",
       "      <td>136</td>\n",
       "      <td>135</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>667 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   file_name second_gene_name  total_hit  cis_hits  \\\n",
       "0    AT1G04380_AT2G27340.csv        AT2G27340        824        71   \n",
       "1    AT1G04380_AT4G01925.csv        AT4G01925        234        84   \n",
       "2    AT1G04380_AT4G04220.csv        AT4G04220        806       189   \n",
       "3    AT1G04380_AT5G51795.csv        AT5G51795       2067        70   \n",
       "4    AT1G11180_AT1G29710.csv        AT1G29710        116        79   \n",
       "..                       ...              ...        ...       ...   \n",
       "719  AT5G66630_AT5G20450.csv        AT5G20450         84        26   \n",
       "720  AT5G66630_AT5G27110.csv        AT5G27110        387       386   \n",
       "721  AT5G66630_AT5G45150.csv        AT5G45150         57         9   \n",
       "722  AT5G66630_AT5G48350.csv        AT5G48350         50        10   \n",
       "723  AT5G66630_AT5G54020.csv        AT5G54020        136       135   \n",
       "\n",
       "     trans_hits  cis_1  trans_1  cis_trans_1  \n",
       "0           753      1        1          1.0  \n",
       "1           150      1        1          1.0  \n",
       "2           617      1        1          1.0  \n",
       "3          1997      1        1          1.0  \n",
       "4            37      1        1          1.0  \n",
       "..          ...    ...      ...          ...  \n",
       "719          58      1        1          1.0  \n",
       "720           1      1        1          1.0  \n",
       "721          48      1        1          1.0  \n",
       "722          40      1        1          1.0  \n",
       "723           1      1        1          1.0  \n",
       "\n",
       "[667 rows x 8 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###filter trans files \n",
    "trans_files=(result.loc[result['trans_1'] == 1])\n",
    "trans_files.to_csv('~/lastversion_gwas_analysis/PermGWAS_Scripts/trans_filtered_hit.csv', index=False)\n",
    "trans_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8ca52ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "###filter non_trans files \n",
    "non_files=(result.loc[result['trans_1'] != 1])\n",
    "non_files.to_csv('~/lastversion_gwas_analysis/PermGWAS_Scripts/non_filtered_hit.csv', index=False)\n",
    "non_files=non_files[\"file_name\"]\n",
    "non_files\n",
    "non_files.to_csv('~/lastversion_gwas_analysis/PermGWAS_Scripts/discard_list.csv', index=False,  header=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3713c60",
   "metadata": {},
   "source": [
    "### Cis hits are removed from files\n",
    "After filtering trans files, still some of the files contain cis hits inside. With this step rest of the cis hits are removed from files (cis_removed.r)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4acd62a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "####filter stringent result files, only contain trans hits\n",
    "####filter stringent result files, only contain trans hits\n",
    "setwd(\"~/lastversion_gwas_analysis/Over_Occ/\")\n",
    "small_gene<-read.csv(\"~/lastversion_gwas_analysis/Over_Occ/trans_filtered_hit.csv\")\n",
    "ara11<-read.csv(\"~/lastversion_gwas_analysis/ara11_50.csv\")\n",
    "\n",
    "Y <- list()\n",
    "csvfiles<-list.files(pattern = \"*.csv\")\n",
    "Y <- lapply(csvfiles, function(x) {\n",
    "  file <- read.csv(x)\n",
    "  return(file)\n",
    "})\n",
    "for (i in 1:length((csvfiles))){\n",
    "  b<-ara11[which(ara11$Gene==S$second_gene_name[i]),]\n",
    "  trials<-subset(Y[[i]],!(b$Start<Y[[i]][[\"Pos\"]] & Y[[i]][[\"Pos\"]]<b$Stop))\n",
    "  write.csv(trials, file=paste( \"filtered_\",csvfiles[i],\".csv\", sep=\"\"))\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ed827e5",
   "metadata": {},
   "source": [
    "### Count number of trans genes, \n",
    "SNPs in files containing trans hits are determined to belong to which genes, and added to the trans_gene list. For this, it is found on which gene the SNPs in the trans files are located in the ara11_10 file, which contains all the SNPs. In this way, if there is a gene corresponding to all SNPs, those genes are added to the list. It is found in the genes in the 10 kb window and added to the list (count_trans_genes.r)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "330d502e",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "##filtering overlapping genes in 10 kb window\n",
    "library(tidyverse)\n",
    "SNPs<-read.csv('~/lastversion_gwas_analysis/SNPs_11_10.csv')\n",
    "ara11<-read.csv(\"~/lastversion_gwas_analysis/ara11_50.csv\")\n",
    "small_gene<-pd.read_csv(\"~/lastversion_gwas_analysis/Over_Occ/Small_gene.csv\")\n",
    "\n",
    "A<-read.csv('~/lastversion_gwas_analysis/PermGWAS_Scripts/trans_filtered_hit.csv')\n",
    "S<-A[,1:5]\n",
    "S$gens_trans<-NA\n",
    "S$gens_trans_10kb<-NA\n",
    "\n",
    "####apply all files see if it is any pattern\n",
    "setwd(\"~/lastversion_gwas_analysis/Over_Occ/trans_hits/\")\n",
    "Trans <- list()\n",
    "csvfiles<-list.files(pattern = \"*filtered_\")\n",
    "Trans <- lapply(csvfiles, function(x) {\n",
    "  file <- read.csv(x)\n",
    "  return(file)\n",
    "})\n",
    "\n",
    "for (i in 1:length((csvfiles))){\n",
    "  Z<-merge(Trans[[i]][,c(4,9)],SNPs[,c(5,17,18)],by='SNP')\n",
    "  S[i,6]<-paste(as.character(na.omit(unique(Z$gene_ara11))),collapse=',')\n",
    "  S[i,7]<-paste(unique(unlist(strsplit(as.character(na.omit(unique(Z$genes_10kb__ara11))),split=','))),collapse=',')\n",
    "  cat(i,'\\n')\n",
    "}\n",
    "a<-nchar(gsub('[^,]','',S$gens_trans))+1\n",
    "df<-as.data.frame(a)\n",
    "S['gene_count']<-NA\n",
    "S['gene_count']<-df\n",
    "write.csv(S, \"~/lastversion_gwas_analysis/overlapping_result/perm_overlapping_genes.csv\")\n",
    "b<-S[duplicated(S$gens_trans),]\n",
    "c<-overlapping_genes[duplicated(overlapping_genes$gens_trans_10kb),]\n",
    "d<-overlapping_genes[duplicated(overlapping_genes$gens_trans) & duplicated(overlapping_genes$gens_trans_10kb),]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d824e7c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, glob\n",
    "\n",
    "###discard files from non_filetered_hits.csv\n",
    "path = ('~/lastversion_gwas_analysis/PermGWAS_Scripts/maf_filtered/')\n",
    "os.chdir(path)\n",
    "with open('non_list.csv', \"r\") as list_file:\n",
    "    _list = list_file.read().splitlines()\n",
    "    [os.remove(os.path.join(path,f)) for f in _list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "ef8b9724",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('~/lastversion_gwas_analysis/PermGWAS_Scripts/MAC_filtered/')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "e4fb15cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_name</th>\n",
       "      <th>second_gene_name</th>\n",
       "      <th>total_hit</th>\n",
       "      <th>cis_hits</th>\n",
       "      <th>trans_hits</th>\n",
       "      <th>cis_1</th>\n",
       "      <th>trans_1</th>\n",
       "      <th>cis_trans_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AT1G11180_AT2G15930.csv</td>\n",
       "      <td>AT2G15930</td>\n",
       "      <td>251.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>251.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>AT1G29710_AT2G15930.csv</td>\n",
       "      <td>AT2G15930</td>\n",
       "      <td>373.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>373.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>AT3G19040_AT2G15930.csv</td>\n",
       "      <td>AT2G15930</td>\n",
       "      <td>5364.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5364.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336</th>\n",
       "      <td>AT3G26240_AT2G15930.csv</td>\n",
       "      <td>AT2G15930</td>\n",
       "      <td>222.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>222.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>420</th>\n",
       "      <td>AT3G58270_AT2G15930.csv</td>\n",
       "      <td>AT2G15930</td>\n",
       "      <td>152.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>152.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>435</th>\n",
       "      <td>AT4G00970_AT2G15930.csv</td>\n",
       "      <td>AT2G15930</td>\n",
       "      <td>329.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>329.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>500</th>\n",
       "      <td>AT4G09012_AT2G15930.csv</td>\n",
       "      <td>AT2G15930</td>\n",
       "      <td>12985.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12985.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>506</th>\n",
       "      <td>AT4G10970_AT2G15930.csv</td>\n",
       "      <td>AT2G15930</td>\n",
       "      <td>187.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>187.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>540</th>\n",
       "      <td>AT4G23180_AT2G15930.csv</td>\n",
       "      <td>AT2G15930</td>\n",
       "      <td>657.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>657.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>545</th>\n",
       "      <td>AT4G34460_AT2G15930.csv</td>\n",
       "      <td>AT2G15930</td>\n",
       "      <td>2996.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2996.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>592</th>\n",
       "      <td>AT5G05750_AT2G15930.csv</td>\n",
       "      <td>AT2G15930</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>627</th>\n",
       "      <td>AT5G20450_AT2G15930.csv</td>\n",
       "      <td>AT2G15930</td>\n",
       "      <td>4576.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4576.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>672</th>\n",
       "      <td>AT5G45150_AT2G15930.csv</td>\n",
       "      <td>AT2G15930</td>\n",
       "      <td>136.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>136.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>721</th>\n",
       "      <td>AT5G63630_AT2G15930.csv</td>\n",
       "      <td>AT2G15930</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   file_name second_gene_name  total_hit  cis_hits  \\\n",
       "6    AT1G11180_AT2G15930.csv        AT2G15930      251.0       0.0   \n",
       "61   AT1G29710_AT2G15930.csv        AT2G15930      373.0       0.0   \n",
       "295  AT3G19040_AT2G15930.csv        AT2G15930     5364.0       0.0   \n",
       "336  AT3G26240_AT2G15930.csv        AT2G15930      222.0       0.0   \n",
       "420  AT3G58270_AT2G15930.csv        AT2G15930      152.0       0.0   \n",
       "435  AT4G00970_AT2G15930.csv        AT2G15930      329.0       0.0   \n",
       "500  AT4G09012_AT2G15930.csv        AT2G15930    12985.0       0.0   \n",
       "506  AT4G10970_AT2G15930.csv        AT2G15930      187.0       0.0   \n",
       "540  AT4G23180_AT2G15930.csv        AT2G15930      657.0       0.0   \n",
       "545  AT4G34460_AT2G15930.csv        AT2G15930     2996.0       0.0   \n",
       "592  AT5G05750_AT2G15930.csv        AT2G15930        8.0       0.0   \n",
       "627  AT5G20450_AT2G15930.csv        AT2G15930     4576.0       0.0   \n",
       "672  AT5G45150_AT2G15930.csv        AT2G15930      136.0       0.0   \n",
       "721  AT5G63630_AT2G15930.csv        AT2G15930        NaN       NaN   \n",
       "\n",
       "     trans_hits  cis_1  trans_1  cis_trans_1  \n",
       "6         251.0      0        1          0.0  \n",
       "61        373.0      0        1          0.0  \n",
       "295      5364.0      0        1          0.0  \n",
       "336       222.0      0        1          0.0  \n",
       "420       152.0      0        1          0.0  \n",
       "435       329.0      0        1          0.0  \n",
       "500     12985.0      0        1          0.0  \n",
       "506       187.0      0        1          0.0  \n",
       "540       657.0      0        1          0.0  \n",
       "545      2996.0      0        1          0.0  \n",
       "592         8.0      0        1          0.0  \n",
       "627      4576.0      0        1          0.0  \n",
       "672       136.0      0        1          0.0  \n",
       "721         NaN      0        0          NaN  "
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "##files does not contain gene \"AT2G15930\" filtered to continue next steps\n",
    "#filter=\"AT2G15930\"\n",
    "filtered_files=(result.loc[result['second_gene_name'] == \"AT2G15930\"])\n",
    "filtered_files=filtered_files.sort_values(\"file_name\")\n",
    "filtered_files.to_csv('~/lastversion_gwas_analysis/PermGWAS_Scripts/filtered_hit.csv', index=False)\n",
    "filtered_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "id": "c14efe13",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6      AT1G11180_AT2G15930.csv\n",
       "61     AT1G29710_AT2G15930.csv\n",
       "295    AT3G19040_AT2G15930.csv\n",
       "336    AT3G26240_AT2G15930.csv\n",
       "420    AT3G58270_AT2G15930.csv\n",
       "435    AT4G00970_AT2G15930.csv\n",
       "500    AT4G09012_AT2G15930.csv\n",
       "506    AT4G10970_AT2G15930.csv\n",
       "540    AT4G23180_AT2G15930.csv\n",
       "545    AT4G34460_AT2G15930.csv\n",
       "592    AT5G05750_AT2G15930.csv\n",
       "627    AT5G20450_AT2G15930.csv\n",
       "672    AT5G45150_AT2G15930.csv\n",
       "721    AT5G63630_AT2G15930.csv\n",
       "Name: file_name, dtype: object"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_files=filtered_files[\"file_name\"]\n",
    "filtered_files\n",
    "filtered_files.to_csv('~/lastversion_gwas_analysis/PermGWAS_Scripts/non_list.csv', index=False,  header=False)\n",
    "filtered_files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "d641f217",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('~/lastversion_gwas_analysis/PermGWAS_Scripts/')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "57ecb04b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "path = ('~/PermGWAS/PermGWAS_OverCo/')\n",
    "files = [file for file in os.listdir(path) if file.endswith(\".csv\")]\n",
    "\n",
    "output = pd.Series(name=\"Rows\", dtype=int)\n",
    "for file in files:\n",
    "    df = pd.read_csv(os.path.join(path, file))\n",
    "    output.at[file.replace(\".csv\", \"\")] = df.shape[0]\n",
    "#output.at[\"Total\"] = output.sum()\n",
    "\n",
    "output.to_csv('~/lastversion_gwas_analysis/PermGWAS_Scripts/Allele_list.csv',  header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "df881183",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Rows</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT2G24600_AT2G16365</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT4G00980_AT5G51795</td>\n",
       "      <td>204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT2G24600_AT5G45500</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT5G10800_AT1G76820</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT4G00970_AT2G41710</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>733</th>\n",
       "      <td>AT3G28610_AT4G18840</td>\n",
       "      <td>265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>734</th>\n",
       "      <td>AT2G42170_AT2G42730</td>\n",
       "      <td>137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>735</th>\n",
       "      <td>AT3G19515_AT2G42640</td>\n",
       "      <td>189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>736</th>\n",
       "      <td>AT2G27340_AT5G45150</td>\n",
       "      <td>126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>737</th>\n",
       "      <td>AT3G32920_AT1G29710</td>\n",
       "      <td>88</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>738 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Unnamed: 0  Rows\n",
       "0    AT2G24600_AT2G16365    76\n",
       "1    AT4G00980_AT5G51795   204\n",
       "2    AT2G24600_AT5G45500    76\n",
       "3    AT5G10800_AT1G76820    80\n",
       "4    AT4G00970_AT2G41710    50\n",
       "..                   ...   ...\n",
       "733  AT3G28610_AT4G18840   265\n",
       "734  AT2G42170_AT2G42730   137\n",
       "735  AT3G19515_AT2G42640   189\n",
       "736  AT2G27340_AT5G45150   126\n",
       "737  AT3G32920_AT1G29710    88\n",
       "\n",
       "[738 rows x 2 columns]"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result=pd.read_csv(\"~/lastversion_gwas_analysis/PermGWAS_Scripts/Allele_list.csv\")\n",
    "result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "27b78f11",
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_name</th>\n",
       "      <th>allele_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>399</th>\n",
       "      <td>AT1G04380_AT2G27340</td>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>523</th>\n",
       "      <td>AT1G04380_AT4G01925</td>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>359</th>\n",
       "      <td>AT1G04380_AT4G04220</td>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77</th>\n",
       "      <td>AT1G04380_AT5G51795</td>\n",
       "      <td>73</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>600</th>\n",
       "      <td>AT1G11180_AT1G29710</td>\n",
       "      <td>107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>AT5G66630_AT5G20450</td>\n",
       "      <td>324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>154</th>\n",
       "      <td>AT5G66630_AT5G27110</td>\n",
       "      <td>324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593</th>\n",
       "      <td>AT5G66630_AT5G45150</td>\n",
       "      <td>324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>389</th>\n",
       "      <td>AT5G66630_AT5G48350</td>\n",
       "      <td>324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>646</th>\n",
       "      <td>AT5G66630_AT5G54020</td>\n",
       "      <td>324</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>738 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               file_name  allele_count\n",
       "399  AT1G04380_AT2G27340            73\n",
       "523  AT1G04380_AT4G01925            73\n",
       "359  AT1G04380_AT4G04220            73\n",
       "77   AT1G04380_AT5G51795            73\n",
       "600  AT1G11180_AT1G29710           107\n",
       "..                   ...           ...\n",
       "181  AT5G66630_AT5G20450           324\n",
       "154  AT5G66630_AT5G27110           324\n",
       "593  AT5G66630_AT5G45150           324\n",
       "389  AT5G66630_AT5G48350           324\n",
       "646  AT5G66630_AT5G54020           324\n",
       "\n",
       "[738 rows x 2 columns]"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result.columns=[\"file_name\", \"allele_count\"]\n",
    "result=result.sort_values(\"file_name\")\n",
    "result.to_csv('~/lastversion_gwas_analysis/PermGWAS_Scripts/Allele_count.csv', index=False ,header=True)\n",
    "result"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcfffb79",
   "metadata": {},
   "source": [
    "## Fake genes\n",
    "Genes are counted most significant and repeated 32 genes are picked from duplicated_overlapping gene file. Calculated its mean and standart deviation. It also need to be statistical significant these genesdid not pick by random chance. For this reason in duplicated_overlapping_gene file added new column is called fake_genes-. Number of trans genes counted for the same number of randomly picked genes from ara11 gene file to put into fake_gene column. Also mean and standart deviation for fake genes are colculated (duplicated_genes.r)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ff08633",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "#####duplicated_genes\n",
    "library(fs)\n",
    "library(tidyverse)\n",
    "library(dplyr)\n",
    "duplicated_genes<-read.csv(\"/storage/full-share/evolgen/leyla/overlapping_result/duplicated_overlapping_genes.csv\")\n",
    "\n",
    "S<-duplicated_genes[which(!duplicated_genes$duplicated==\"\"),]\n",
    "\n",
    "a<-nchar(gsub('[^,]','',S$duplicated))+1\n",
    "df<-as.data.frame(a)\n",
    "S['gene_count']<-NA\n",
    "S['gene_count']<-df\n",
    "S['fake_genes']<-NA\n",
    "w1<-unlist(strsplit(S$duplicated, \",\"))\n",
    "W1<-table(w1)\n",
    "W1<-sort(W1, decreasing = TRUE)\n",
    "W_1<-W1[1:32]\n",
    "setwd(\"~/lastversion_gwas_analysis/PermGWAS_Scripts/fake_genes_duplicated/\")\n",
    "load(\"/storage/full-share/evolgen/results_110522.rda\")\n",
    "a<-1\n",
    "#d:file contain the file names, trans genes, number of trans genes, column 5 has information number of trans genes, column 13\n",
    "#put fake genes according to column 12: number of rel trans genes, \n",
    "for(a in 1:100){\n",
    "  u<-1\n",
    "  for (u in 1:nrow(S)){\n",
    "    i<-sample(1:33056, S[u,5])\n",
    "    S[u,6]<-as.character(paste(ara11[i,5],collapse =\",\"))\n",
    "  }\n",
    "  w<-unlist(strsplit(S$fake_genes, \",\"))\n",
    "  \n",
    "  W<-table(w)\n",
    "  W<-sort(W, decreasing = TRUE)\n",
    "  W[1:32]\n",
    "  W_<-W[1:32]\n",
    "  W_<-W[1:32]\n",
    "  W_<-as.matrix(W_, sep=\",\")\n",
    "  new<-as.data.frame(W_)\n",
    "  write.csv(new, file=paste( \"filtered_\",a,\".csv\", sep=\"\"))\n",
    "}\n",
    "data_all <- list.files(path = \"~/lastversion_gwas_analysis/PermGWAS_Scripts/fake_genes_duplicated/\",  # Identify all CSV files\n",
    "                       pattern = \"*.csv\", full.names = TRUE) %>% \n",
    "  lapply(read_csv) %>%                              # Store all files in list\n",
    "  bind_rows                                         # Combine data sets into one data set \n",
    "data_all \n",
    "\n",
    "S_<-S[,2:6]\n",
    "write.csv(a_, \"~/lastversion_gwas_analysis/overlapping_result/ger_1_2.csv\", row.names = FALSE)\n",
    "\n",
    "mean(data_all$W1_1)\n",
    "#6.1875\n",
    "sd(data_all$W_1)\n",
    "#0.5922892\n",
    "\n",
    "#fake data results\n",
    "mean(data_all$V1)\n",
    "#3.036875\n",
    "sd(data_all$V1)\n",
    "# 0.2473042\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "028580df",
   "metadata": {},
   "source": [
    "## Gene list create for phenotype\n",
    "In this section, the phenotype file causing the 32 most important trans genes and these trans genes are shown."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "1d0766fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "c=pd.read_csv(\"~/lastversion_gwas_analysis/PermGWAS_Scripts/duplicated_file_names.csv\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b16d7c14",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_1</th>\n",
       "      <th>file_2</th>\n",
       "      <th>gens_trans</th>\n",
       "      <th>gene_count</th>\n",
       "      <th>fake_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT1G13430_AT2G16365.csv</td>\n",
       "      <td>AT2G16365_AT1G13430.csv</td>\n",
       "      <td>AT1G08840</td>\n",
       "      <td>922</td>\n",
       "      <td>AT1G21590,AT5G00610,AT4G34360,AT2G15450,AT4G09...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>AT3G47010_AT2G16365.csv</td>\n",
       "      <td>AT2G16365_AT3G47010.csv</td>\n",
       "      <td>AT1G08840</td>\n",
       "      <td>345</td>\n",
       "      <td>AT4G31690,AT5G17410,AT1G76770,AT5G55190,AT5G13...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>AT4G00970_AT3G47010.csv</td>\n",
       "      <td>AT3G47010_AT4G00970.csv</td>\n",
       "      <td>AT1G08840</td>\n",
       "      <td>584</td>\n",
       "      <td>AT5G59820,AT1G32520,AT2G37030,AT1G07950,AT1G09...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AT2G16365_AT1G13430.csv</td>\n",
       "      <td>AT1G13430_AT2G16365.csv</td>\n",
       "      <td>AT1G08840</td>\n",
       "      <td>922</td>\n",
       "      <td>AT4G31615,AT3G51135,AT2G07673,AT5G36140,AT3G03...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>AT3G47010_AT4G00970.csv</td>\n",
       "      <td>AT4G00970_AT3G47010.csv</td>\n",
       "      <td>AT1G08840</td>\n",
       "      <td>584</td>\n",
       "      <td>AT1G66600,AT1G25211,AT1G06087,AT4G00165,AT3G05...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AT1G42680_AT4G00970.csv</td>\n",
       "      <td>AT4G00970_AT1G42680.csv</td>\n",
       "      <td>AT5G41210</td>\n",
       "      <td>405</td>\n",
       "      <td>AT1G32550,AT2G45320,AT4G38740,AT3G04350,AT3G13...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AT2G16365_AT2G24600.csv</td>\n",
       "      <td>AT2G24600_AT2G16365.csv</td>\n",
       "      <td>AT5G41210</td>\n",
       "      <td>100</td>\n",
       "      <td>AT5G05330,AT5G08165,AT2G01630,AT2G35430,AT5G15...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>AT2G16365_AT3G47010.csv</td>\n",
       "      <td>AT3G47010_AT2G16365.csv</td>\n",
       "      <td>AT5G41210</td>\n",
       "      <td>345</td>\n",
       "      <td>AT1G04537,AT1G72080,AT5G40060,AT4G02290,AT5G11...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>AT2G24600_AT2G16365.csv</td>\n",
       "      <td>AT2G16365_AT2G24600.csv</td>\n",
       "      <td>AT5G41210</td>\n",
       "      <td>100</td>\n",
       "      <td>AT5G65330,AT4G09025,AT4G20480,AT3G13226,AT3G08...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>AT4G00970_AT1G42680.csv</td>\n",
       "      <td>AT1G42680_AT4G00970.csv</td>\n",
       "      <td>AT5G41210</td>\n",
       "      <td>405</td>\n",
       "      <td>AT1G53870,AT1G79485,AT2G30730,AT3G22860,AT3G28...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>198 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     file_1                   file_2 gens_trans  gene_count  \\\n",
       "3   AT1G13430_AT2G16365.csv  AT2G16365_AT1G13430.csv  AT1G08840         922   \n",
       "26  AT3G47010_AT2G16365.csv  AT2G16365_AT3G47010.csv  AT1G08840         345   \n",
       "33  AT4G00970_AT3G47010.csv  AT3G47010_AT4G00970.csv  AT1G08840         584   \n",
       "8   AT2G16365_AT1G13430.csv  AT1G13430_AT2G16365.csv  AT1G08840         922   \n",
       "27  AT3G47010_AT4G00970.csv  AT4G00970_AT3G47010.csv  AT1G08840         584   \n",
       "..                      ...                      ...        ...         ...   \n",
       "7   AT1G42680_AT4G00970.csv  AT4G00970_AT1G42680.csv  AT5G41210         405   \n",
       "9   AT2G16365_AT2G24600.csv  AT2G24600_AT2G16365.csv  AT5G41210         100   \n",
       "11  AT2G16365_AT3G47010.csv  AT3G47010_AT2G16365.csv  AT5G41210         345   \n",
       "14  AT2G24600_AT2G16365.csv  AT2G16365_AT2G24600.csv  AT5G41210         100   \n",
       "29  AT4G00970_AT1G42680.csv  AT1G42680_AT4G00970.csv  AT5G41210         405   \n",
       "\n",
       "                                           fake_genes  \n",
       "3   AT1G21590,AT5G00610,AT4G34360,AT2G15450,AT4G09...  \n",
       "26  AT4G31690,AT5G17410,AT1G76770,AT5G55190,AT5G13...  \n",
       "33  AT5G59820,AT1G32520,AT2G37030,AT1G07950,AT1G09...  \n",
       "8   AT4G31615,AT3G51135,AT2G07673,AT5G36140,AT3G03...  \n",
       "27  AT1G66600,AT1G25211,AT1G06087,AT4G00165,AT3G05...  \n",
       "..                                                ...  \n",
       "7   AT1G32550,AT2G45320,AT4G38740,AT3G04350,AT3G13...  \n",
       "9   AT5G05330,AT5G08165,AT2G01630,AT2G35430,AT5G15...  \n",
       "11  AT1G04537,AT1G72080,AT5G40060,AT4G02290,AT5G11...  \n",
       "14  AT5G65330,AT4G09025,AT4G20480,AT3G13226,AT3G08...  \n",
       "29  AT1G53870,AT1G79485,AT2G30730,AT3G22860,AT3G28...  \n",
       "\n",
       "[198 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "c=c.explode('gens_trans')\n",
    "c\n",
    "c_=(c.loc[(c[\"gens_trans\"]== \"AT2G21830\") | (c[\"gens_trans\"]== \"AT2G42860\")| (c[\"gens_trans\"]== \"AT4G01930\")| (c[\"gens_trans\"]== \"AT1G08840\")| (c[\"gens_trans\"]== \"AT1G09880\")| (c[\"gens_trans\"]== \"AT2G14080\")| (c[\"gens_trans\"]== \"AT2G18380\")| (c[\"gens_trans\"]== \"AT2G18735\")| (c[\"gens_trans\"]== \"AT2G19110\")| (c[\"gens_trans\"]== \"AT2G21160\")])\n",
    "#b_=(b.loc[(b[\"gens_trans\"]== \"AT3G44630\") | (b[\"gens_trans\"]== \"AT4G16920\")])\n",
    "#result.loc[(result['cis_hits'] > 0)  | (result['trans_hits'] >0), 'cis_trans_1'] = 1\n",
    "c_=c_.sort_values(\"gens_trans\")\n",
    "c_.to_csv('/home/s417377/lastversion_gwas_analysis/overlapping_result/gene1_2_trans_result_genes.csv', index=False)\n",
    "\n",
    "c_1=(c.loc[(c[\"gens_trans\"]== \"AT2G21170 \") | (c[\"gens_trans\"]== \"AT2G21330\")| (c[\"gens_trans\"]== \"AT2G21340\")| (c[\"gens_trans\"]== \"AT2G21350\")| (c[\"gens_trans\"]== \"AT2G21370\")| (c[\"gens_trans\"]== \"AT2G21470\")| (c[\"gens_trans\"]== \"AT2G21550\")| (c[\"gens_trans\"]== \"AT2G21870\")| (c[\"gens_trans\"]== \"AT2G21920\")| (c[\"gens_trans\"]== \"AT2G29200\")|(c[\"gens_trans\"]==\"AT2G21170\")])\n",
    "#b_=(b.loc[(b[\"gens_trans\"]== \"AT3G44630\") | (b[\"gens_trans\"]== \"AT4G16920\")])\n",
    "\n",
    "#result.loc[(result['cis_hits'] > 0)  | (result['trans_hits'] >0), 'cis_trans_1'] = 1\n",
    "c_1=c_1.sort_values(\"gens_trans\")\n",
    "c_1.to_csv('/home/s417377/lastversion_gwas_analysis/overlapping_result/gene1_2_trans_result_genes2.csv', index=False)\n",
    "\n",
    "c_2=(c.loc[(c[\"gens_trans\"]== \"AT2G43850\") | (c[\"gens_trans\"]== \"AT2G43860\")| (c[\"gens_trans\"]== \"AT3G10440\")| (c[\"gens_trans\"]== \"AT3G52140\")| (c[\"gens_trans\"]== \"AT3G55200\")| (c[\"gens_trans\"]== \"AT4G02190\")| (c[\"gens_trans\"]== \"AT4G04015\")| (c[\"gens_trans\"]== \"AT4G08990\")| (c[\"gens_trans\"]== \"AT5G41180\")| (c[\"gens_trans\"]== \"AT5G41190\")| (c[\"gens_trans\"]== \"AT5G41200\")| (c[\"gens_trans\"]== \"AT5G41210\")])\n",
    "#b_=(b.loc[(b[\"gens_trans\"]== \"AT3G44630\") | (b[\"gens_trans\"]== \"AT4G16920\")])\n",
    "#result.loc[(result['cis_hits'] > 0)  | (result['trans_hits'] >0), 'cis_trans_1'] = 1\n",
    "c_2=c_2.sort_values(\"gens_trans\")\n",
    "c_2.to_csv('/home/s417377/lastversion_gwas_analysis/overlapping_result/gene1_2_trans_result_genes3.csv', index=False)\n",
    "\n",
    "frames = [c_, c_1, c_2]\n",
    "result = pd.concat(frames)\n",
    "display(result)\n",
    "result.to_csv('/home/s417377/lastversion_gwas_analysis/overlapping_result/gene1_2_trans_result_genes.csv', index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c94ad47d",
   "metadata": {},
   "source": [
    "## AC_1 and AC_0 count\n",
    "Genotype files of SNPs that cause trans gene formation are created. From these genotype files, only those corresponding to the accessions in the phenotype files are retrieved. In this way, against the phenotype file, the genotype file is created for the same accesions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ee684b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "\n",
    "library(tidyverse)\n",
    "SNPs<-read.csv('~/lastversion_gwas_analysis/SNPs_11_10.csv')\n",
    "a<-read.csv(\"~/lastversion_gwas_analysis/overlapping_result/ger_1_1_32.csv\")\n",
    "\n",
    "#AT2G21830\n",
    "AT2G21830<- SNPs[which(SNPs$gene == \"AT2G21830\"),]\n",
    "setwd(\"~/trans_genes/AT2G21830\")\n",
    "Y <- list()\n",
    "csvfiles<-list.files(pattern = \"*.csv\")\n",
    "Y <- lapply(csvfiles, function(x) {\n",
    "  file <- read.csv(x)\n",
    "  return(file)\n",
    "})\n",
    "for (i in 1:length((csvfiles))){\n",
    "  Y[[i]]<-mutate(Y[[i]], SNP=paste(CHR, sep=\"- \", POS))\n",
    "}\n",
    "for (i in 1:length(Y)){\n",
    "  Z<-merge(AT2G21830[,c(4,5)],Y[[i]][,c(13,14)],by='SNP')\n",
    "  b[i]<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "  cat(i,'\\n')\n",
    "}\n",
    "load(\"~/lastversion_gwas_analysis/F_95/X_f95_2.rda\")\n",
    "b1<-X[,\"2- 9304447\"]\n",
    "b2<-X[,\"2- 9304693\"]\n",
    "b1<-as.data.frame(b1)\n",
    "b2<-as.data.frame(b2)\n",
    "write.csv( b1,\"~/lastversion_gwas_analysis/PermGWAS_Scripts/AC_1_0_files/filtered_2_9304447.csv\", row.names = TRUE)\n",
    "write.csv( b2,\"~/lastversion_gwas_analysis/PermGWAS_Scripts/AC_1_0_files/filtered_2_9304693.csv\", row.names = TRUE)\n",
    "\n",
    "\n",
    "setwd(\"~/lastversion_gwas_analysis/PermGWAS_Scripts/AC_1_0_files/\")\n",
    "SNP <- list()\n",
    "csvfiles<-list.files(pattern = \"*.csv\")\n",
    "SNP <- lapply(csvfiles, function(x) {\n",
    "  file <- read.csv(x)\n",
    "  return(file)\n",
    "})\n",
    "\n",
    "e<-merge(Y[[1]], SNP[[1]], by.=\"X\",)\n",
    "e<-merge(Y[[1]], SNP[[1]], by=\"X\")\n",
    "e1<-merge(Y[[2]], SNP[[1]], by=\"X\")\n",
    "e2<-merge(Y[[3]], SNP[[1]], by=\"X\")\n",
    "e3<-merge(Y[[4]], SNP[[1]], by=\"X\")\n",
    "e4<-merge(Y[[5]], SNP[[1]], by=\"X\")\n",
    "e5<-merge(Y[[6]], SNP[[2]], by=\"X\")\n",
    "e6<-merge(Y[[7]], SNP[[1]], by=\"X\")\n",
    "e7<-merge(Y[[8]], SNP[[2]], by=\"X\")\n",
    "\n",
    "write.csv( e,\"~/lastversion_gwas_analysis/PermGWAS_Scripts/AC_1_0_files/AT1G42680_AT4G00970_2_9304447.csv\", row.names = TRUE)\n",
    "write.csv( e1,\"~/lastversion_gwas_analysis/PermGWAS_Scripts/AC_1_0_files/AT2G24600_AT2G29710_2_9304447.csv\", row.names = TRUE)\n",
    "write.csv( e2,\"~/lastversion_gwas_analysis/PermGWAS_Scripts/AC_1_0_files/AT2G24600_AT4G00970_2_9304447.csv\", row.names = TRUE)\n",
    "write.csv( e3,\"~/lastversion_gwas_analysis/PermGWAS_Scripts/AC_1_0_files/AT2G29710_AT2G24600_2_9304447.csv\", row.names = TRUE)\n",
    "write.csv( e4,\"~/lastversion_gwas_analysis/PermGWAS_Scripts/AC_1_0_files/AT3G47010_AT4G00970_2_9304447.csv\", row.names = TRUE)\n",
    "write.csv( e5,\"~/lastversion_gwas_analysis/PermGWAS_Scripts/AC_1_0_files/AT4G00970_AT1G42680_2_9304693.csv\", row.names = TRUE)\n",
    "write.csv( e6,\"~/lastversion_gwas_analysis/PermGWAS_Scripts/AC_1_0_files/AT4G00970_AT2G24600_2_9304447.csv\", row.names = TRUE)\n",
    "write.csv( e7,\"~/lastversion_gwas_analysis/PermGWAS_Scripts/AC_1_0_files/AT4G00970_AT3G47010_2_9304693.csv\", row.names = TRUE)\n",
    "#\n",
    "AT2G18735<- SNPs[which(SNPs$gene == \"AT2G18735\"),]\n",
    "setwd(\"/home/s417377/trans_genes/AT2G18735\")\n",
    "Y <- list()\n",
    "csvfiles<-list.files(pattern = \"*.csv\")\n",
    "Y <- lapply(csvfiles, function(x) {\n",
    "  file <- read.csv(x)\n",
    "  return(file)\n",
    "})\n",
    "for (i in 1:length((csvfiles))){\n",
    "  Y[[i]]<-mutate(Y[[i]], SNP=paste(CHR, sep=\"- \", POS))\n",
    "}\n",
    "for (i in 1:length(Y)){\n",
    "  Z<-merge(AT2G18735[,c(4,5)],Y[[i]][,c(13,14)],by='SNP')\n",
    "  b[i]<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "  cat(i,'\\n')\n",
    "}\n",
    "load(\"~/lastversion_gwas_analysis/F_95/X_f95_2.rda\")\n",
    "b1<-X[,\"2- 8122148\"]\n",
    "b2<-X[,\"2- 8122723\"]\n",
    "b1<-as.data.frame(b1)\n",
    "b2<-as.data.frame(b2)\n",
    "write.csv( b1,\"~/lastversion_gwas_analysis/PermGWAS_Scripts/AC_1_0_files/filtered_2_8122148.csv\", row.names = TRUE)\n",
    "write.csv( b2,\"~/lastversion_gwas_analysis/PermGWAS_Scripts/AC_1_0_files/filtered_2_8122723.csv\", row.names = TRUE)\n",
    "write.csv( b,\"~/lastversion_gwas_analysis/PermGWAS_Scripts/AC_1_0_files/SNPsAT2G18735_.csv\", row.names = TRUE)\n",
    "#\n",
    "AT1G08840<- SNPs[which(SNPs$gene == \"AT1G08840\"),]\n",
    "setwd(\"~/trans_genes/AT1G08840\")\n",
    "Y <- list()\n",
    "csvfiles<-list.files(pattern = \"*.csv\")\n",
    "Y <- lapply(csvfiles, function(x) {\n",
    "  file <- read.csv(x)\n",
    "  return(file)\n",
    "})\n",
    "for (i in 1:length((csvfiles))){\n",
    "  Y[[i]]<-mutate(Y[[i]], SNP=paste(CHR, sep=\"- \", POS))\n",
    "}\n",
    "for (i in 1:length(Y)){\n",
    "  Z<-merge(AT1G08840[,c(4,5)],Y[[i]][,c(13,14)],by='SNP')\n",
    "  b[i]<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "  cat(i,'\\n')\n",
    "}\n",
    "load(\"~/lastversion_gwas_analysis/F_95/X_f95_1.rda\")\n",
    "b1<-X[,\"1- 2830188\"]\n",
    "b2<-X[,\"1- 2833798\"]\n",
    "b3<-X[,\"1- 2832154\"]\n",
    "b1<-as.data.frame(b1)\n",
    "b2<-as.data.frame(b2)\n",
    "b3<-as.data.frame(b3)\n",
    "write.csv( b1,\"~/lastversion_gwas_analysis/PermGWAS_Scripts/AC_1_0_files/filtered_1_2830188.csv\", row.names = TRUE)\n",
    "write.csv( b2,\"~/lastversion_gwas_analysis/PermGWAS_Scripts/AC_1_0_files/filtered_1_2833798.csv\", row.names = TRUE)\n",
    "write.csv( b3,\"~/lastversion_gwas_analysis/PermGWAS_Scripts/AC_1_0_files/filtered_1_2832154.csv\", row.names = TRUE)\n",
    "write.csv( b,\"~/lastversion_gwas_analysis/PermGWAS_Scripts/AC_1_0_files/SNPsAT1G08840_.csv\", row.names = TRUE)\n",
    "\n",
    "#AT1G08840\n",
    "#\"AT1G13430_AT2G16365.csv\"\n",
    "setwd(\"~/trans_genes/AT1G08840\")\n",
    "Y <- list()\n",
    "csvfiles<-list.files(pattern = \"*.csv\")\n",
    "Y <- lapply(csvfiles, function(x) {\n",
    "  file <- read.csv(x)\n",
    "  return(file)\n",
    "})\n",
    "for (i in 1:length((csvfiles))){\n",
    "  Y[[i]]<-mutate(Y[[i]], SNP=paste(CHR, sep=\"- \", POS))\n",
    "}\n",
    "d<-a[which(\"AT1G13430_AT2G16365.csv\"==a$file_1),]\n",
    "d<-unique(d$gens_trans)\n",
    "d<-as.data.frame(d)\n",
    "colnames(d)<-\"gene\"\n",
    "mylist<-c()\n",
    "for (i in 1:nrow(d)){\n",
    "  e<-SNPs[which(d$gene[i]==SNPs$gene),]\n",
    "  mylist[[length(mylist)+1]]<-e\n",
    "  cat(i,'\\n')\n",
    "}\n",
    "\n",
    "for (i in 1:length((mylist))){\n",
    "  Z<-merge(mylist[[i]][,c(4,5)],Y[[1]][,c(5,14)],by='SNP')\n",
    "  b[i]<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "  b<-as.data.frame(b)\n",
    "}  \n",
    "  Z<-merge(mylist[[1]][,c(4,5)],Y[[1]][,c(5,14)],by='SNP')\n",
    "  b<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "write.csv( b,\"~/lastversion_gwas_analysis/PermGWAS_Scripts/AC_1_0_files/SNPsAT1G08840_.csv\", row.names = TRUE)\n",
    "  \n",
    "#\"AT2G16365_AT1G13430.csv\" \n",
    "d<-a[which(\"AT2G16365_AT1G13430.csv\"==a$file_1),]\n",
    "d<-unique(d$gens_trans)\n",
    "d<-as.data.frame(d)\n",
    "colnames(d)<-\"gene\"\n",
    "mylist<-c()\n",
    "for (i in 1:nrow(d)){\n",
    "  e<-SNPs[which(d$gene[i]==SNPs$gene),]\n",
    "  mylist[[length(mylist)+1]]<-e\n",
    "  cat(i,'\\n')\n",
    "}\n",
    "b1<-c()\n",
    "for (i in 1:length((mylist))){\n",
    "  Z<-merge(mylist[[i]][,c(4,5)],Y[[2]][,c(5,14)],by='SNP')\n",
    "  b1[i]<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "  b1<-as.data.frame(b1)\n",
    "}  \n",
    "\n",
    "#\"AT2G16365_AT3G47010.csv\"\n",
    "d<-a[which(\"AT2G16365_AT3G47010.csv\"==a$file_1),]\n",
    "d<-unique(d$gens_trans)\n",
    "d<-as.data.frame(d)\n",
    "colnames(d)<-\"gene\"\n",
    "mylist<-c()\n",
    "for (i in 1:nrow(d)){\n",
    "  e<-SNPs[which(d$gene[i]==SNPs$gene),]\n",
    "  mylist[[length(mylist)+1]]<-e\n",
    "  cat(i,'\\n')\n",
    "}\n",
    "b2<-c()\n",
    "for (i in 1:length((mylist))){\n",
    "  Z<-merge(mylist[[i]][,c(4,5)],Y[[3]][,c(5,14)],by='SNP')\n",
    "  b2[i]<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "  b2<-as.data.frame(b2)\n",
    "} \n",
    "#\"AT3G47010_AT2G16365.csv\" \n",
    "d<-a[which(\"AT3G47010_AT2G16365.csv\"==a$file_1),]\n",
    "d<-unique(d$gens_trans)\n",
    "d<-as.data.frame(d)\n",
    "colnames(d)<-\"gene\"\n",
    "mylist<-c()\n",
    "for (i in 1:nrow(d)){\n",
    "  e<-SNPs[which(d$gene[i]==SNPs$gene),]\n",
    "  mylist[[length(mylist)+1]]<-e\n",
    "  cat(i,'\\n')\n",
    "}\n",
    "b3<-c()\n",
    "for (i in 1:length((mylist))){\n",
    "  Z<-merge(mylist[[i]][,c(4,5)],Y[[4]][,c(5,14)],by='SNP')\n",
    "  b3[i]<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "  b3<-as.data.frame(b3)\n",
    "  \n",
    "} \n",
    "#\"AT3G47010_AT4G00970.csv\"\n",
    "d<-a[which(\"AT3G47010_AT4G00970.csv\"==a$file_1),]\n",
    "d<-unique(d$gens_trans)\n",
    "d<-as.data.frame(d)\n",
    "colnames(d)<-\"gene\"\n",
    "mylist<-c()\n",
    "for (i in 1:nrow(d)){\n",
    "  e<-SNPs[which(d$gene[i]==SNPs$gene),]\n",
    "  mylist[[length(mylist)+1]]<-e\n",
    "  cat(i,'\\n')\n",
    "}\n",
    "b4<-c()\n",
    "for (i in 1:length((mylist))){\n",
    "  Z<-merge(mylist[[i]][,c(4,5)],Y[[5]][,c(5,14)],by='SNP')\n",
    "  b4[i]<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "  b4<-as.data.frame(b4)\n",
    "  \n",
    "} \n",
    "#\"AT4G00970_AT3G47010.csv\" \n",
    "d<-a[which(\"AT4G00970_AT3G47010.csv\"==a$file_1),]\n",
    "d<-unique(d$gens_trans)\n",
    "d<-as.data.frame(d)\n",
    "colnames(d)<-\"gene\"\n",
    "mylist<-c()\n",
    "for (i in 1:nrow(d)){\n",
    "  e<-SNPs[which(d$gene[i]==SNPs$gene),]\n",
    "  mylist[[length(mylist)+1]]<-e\n",
    "  cat(i,'\\n')\n",
    "}\n",
    "b5<-c()\n",
    "for (i in 1:length((mylist))){\n",
    "  Z<-merge(mylist[[i]][,c(4,5)],Y[[6]][,c(5,14)],by='SNP')\n",
    "  b5[i]<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "  b5<-as.data.frame(b5)\n",
    "} \n",
    "\n",
    "#AT2G18735\n",
    "setwd(\"/home/s417377/trans_genes/AT2G18735\")\n",
    "Y <- list()\n",
    "csvfiles<-list.files(pattern = \"*.csv\")\n",
    "Y <- lapply(csvfiles, function(x) {\n",
    "  file <- read.csv(x)\n",
    "  return(file)\n",
    "})\n",
    "for (i in 1:length((csvfiles))){\n",
    "  Y[[i]]<-mutate(Y[[i]], SNP=paste(CHR, sep=\"- \", POS))\n",
    "}\n",
    "#\"AT2G29710_AT2G24600.csv\"\n",
    "d<-a[which(\"AT2G29710_AT2G24600.csv\"==a$file_1),]\n",
    "d<-unique(d$gens_trans)\n",
    "d<-as.data.frame(d)\n",
    "colnames(d)<-\"gene\"\n",
    "mylist<-c()\n",
    "for (i in 1:nrow(d)){\n",
    "  e<-SNPs[which(d$gene[i]==SNPs$gene),]\n",
    "  mylist[[length(mylist)+1]]<-e\n",
    "  cat(i,'\\n')\n",
    "}\n",
    "f<-c()\n",
    "for (i in 1:length((mylist))){\n",
    "  Z<-merge(mylist[[i]][,c(4,5)],Y[[1]][,c(5,14)],by='SNP')\n",
    "  f[i]<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "  f<-as.data.frame(f)\n",
    "} \n",
    "#\"AT2G24600_AT2G29710.csv\" \n",
    "d<-a[which(\"AT2G24600_AT2G29710.csv\"==a$file_1),]\n",
    "d<-unique(d$gens_trans)\n",
    "d<-as.data.frame(d)\n",
    "colnames(d)<-\"gene\"\n",
    "mylist<-c()\n",
    "for (i in 1:nrow(d)){\n",
    "  e<-SNPs[which(d$gene[i]==SNPs$gene),]\n",
    "  mylist[[length(mylist)+1]]<-e\n",
    "  cat(i,'\\n')\n",
    "}\n",
    "f1<-c()\n",
    "for (i in 1:length((mylist))){\n",
    "  Z<-merge(mylist[[i]][,c(4,5)],Y[[2]][,c(5,14)],by='SNP')\n",
    "  f1[i]<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "  f1<-as.data.frame(f1)\n",
    "} \n",
    "#\"AT2G24600_AT4G00970.csv\"\n",
    "d<-a[which(\"AT2G24600_AT4G00970.csv\"==a$file_1),]\n",
    "d<-unique(d$gens_trans)\n",
    "d<-as.data.frame(d)\n",
    "colnames(d)<-\"gene\"\n",
    "mylist<-c()\n",
    "for (i in 1:nrow(d)){\n",
    "  e<-SNPs[which(d$gene[i]==SNPs$gene),]\n",
    "  mylist[[length(mylist)+1]]<-e\n",
    "  cat(i,'\\n')\n",
    "}\n",
    "f2<-c()\n",
    "for (i in 1:length((mylist))){\n",
    "  Z<-merge(mylist[[i]][,c(4,5)],Y[[3]][,c(5,14)],by='SNP')\n",
    "  f2[i]<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "  f2<-as.data.frame(f2)\n",
    "} \n",
    "#\"AT3G12020_AT3G19040.csv\" \n",
    "d<-a[which(\"AT3G12020_AT3G19040.csv\"==a$file_1),]\n",
    "d<-unique(d$gens_trans)\n",
    "d<-as.data.frame(d)\n",
    "colnames(d)<-\"gene\"\n",
    "mylist<-c()\n",
    "for (i in 1:nrow(d)){\n",
    "  e<-SNPs[which(d$gene[i]==SNPs$gene),]\n",
    "  mylist[[length(mylist)+1]]<-e\n",
    "  cat(i,'\\n')\n",
    "}\n",
    "f3<-c()\n",
    "for (i in 1:length((mylist))){\n",
    "  Z<-merge(mylist[[i]][,c(4,5)],Y[[4]][,c(5,14)],by='SNP')\n",
    "  f3[i]<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "  f3<-as.data.frame(f3)\n",
    "} \n",
    "#\"AT3G19040_AT3G12020.csv\"\n",
    "d<-a[which(\"AT3G19040_AT3G12020.csv\"==a$file_1),]\n",
    "d<-unique(d$gens_trans)\n",
    "d<-as.data.frame(d)\n",
    "colnames(d)<-\"gene\"\n",
    "mylist<-c()\n",
    "for (i in 1:nrow(d)){\n",
    "  e<-SNPs[which(d$gene[i]==SNPs$gene),]\n",
    "  mylist[[length(mylist)+1]]<-e\n",
    "  cat(i,'\\n')\n",
    "}\n",
    "f4<-c()\n",
    "for (i in 1:length((mylist))){\n",
    "  Z<-merge(mylist[[i]][,c(4,5)],Y[[5]][,c(5,14)],by='SNP')\n",
    "  f4[i]<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "  f4<-as.data.frame(f4)\n",
    "} \n",
    "#\"AT4G00970_AT2G24600.csv\" \n",
    "d<-a[which(\"AT4G00970_AT2G24600.csv\"==a$file_1),]\n",
    "d<-unique(d$gens_trans)\n",
    "d<-as.data.frame(d)\n",
    "colnames(d)<-\"gene\"\n",
    "mylist<-c()\n",
    "for (i in 1:nrow(d)){\n",
    "  e<-SNPs[which(d$gene[i]==SNPs$gene),]\n",
    "  mylist[[length(mylist)+1]]<-e\n",
    "  cat(i,'\\n')\n",
    "}\n",
    "f5<-c()\n",
    "for (i in 1:length((mylist))){\n",
    "  Z<-merge(mylist[[i]][,c(4,5)],Y[[6]][,c(5,14)],by='SNP')\n",
    "  f5[i]<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "  f5<-as.data.frame(f5)\n",
    "} \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fb5c63a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import libraries\n",
    "from matplotlib_venn import venn2, venn2_circles, venn2_unweighted\n",
    "from matplotlib_venn import venn3, venn3_circles\n",
    "from matplotlib import pyplot as plt\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b10c9384",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-9-fea871b11ebf>:25: UserWarning: Legend does not support None instances.\n",
      "A proxy artist may be used instead.\n",
      "See: https://matplotlib.org/users/legend_guide.html#creating-artists-specifically-for-adding-to-the-legend-aka-proxy-artists\n",
      "  ax.legend(handles=h, labels=l, title=\"Legend\",loc=3)\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZQAAAFjCAYAAAAaUv0BAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAABLAUlEQVR4nO3deXxU1f3/8dedfSaTfSdhDyGEACKKglrGtWCVqtiKaG1dWhdqq7ZV+612de1PW4ul1m7WqtWqrYhLsagErVsV2cK+BchC9n2SzHZ+f0wSE2QJuTfMTPJ5Ph55kIHJmQ/bvHPuuedzNKUUQgghhF6mSBcghBBiaJBAEUIIYQgJFCGEEIaQQBFCCGEICRQhhBCGkEARQghhCAkUIYQQhpBAEUIIYQgJFCGEEIaQQBFCCGEICRQhhBCGkEARQghhCAkUIYQQhpBAEUIIYQgJFCGEEIaQQBFCCGEICRQhhBCGkEARQghhCAkUIYQQhpBAEUIIYQgJFCGEEIaQQBFCCGEICRQhhBCGkEARQghhCAkUIYQQhpBAEUIIYQgJFCGEEIaQQBFCCGEICRQhhBCGkEARQghhCAkUIYQQhrBEugAh9NKKsQMOwA7YgCDgB3xdP/qVh0DkKhRieNCUUpGuQQgAtGLMQBqQDqQAqV0/pgBJhEOjOzi6f7QDWj+GV3SFCxAAOoEmoB5o6Pqx/qDHjcpDyJDfnBDDgASKOO60YpKAscA4YCSQ2fWRSnRdhlWEw6Uc2Avs6/5ReWiNZGFCRCMJFDFotGJMQA7h4BjLZyGSFMGyjFLPZwHT/bFbLq2J4UwCRRhGK8YJFAHTgEJgNOE1jeHCB2wFNgElwFblwRfZkoQ4fiRQxIBpxdgIB8fUro88wBzRoqJLANhJOFxKgC3KgzeyJQkxeCRQRL9pxWhAPjCdcIAUAFYjxjaH8NsC+Bzhj4A9QMCkUJpCA5QWXs8A1XcFPqShdVgxt1uwdVqwdlixB03G1DQIQoQD5iPgA+Vhf4TrEcJQEijiiLpCpAA4A5hNeOH8mDj8eFO8tKV66UhtIxjficnpx2oPYLEGcVhCOEwGBRNACAIhE50BEz6/Cb/Pgr/VRqDBCfVxWOqdOBqdxAfMEb8cVwZ8QDhcdkS4FiF0k0ARn9MrRE7r+kg72teYQgQzWqnPaKUjxUsoqR1TfCdORwC3WeEY7JoHIqDh7bDS1mKno8lBsNqNuTyRhCYniREop5ZwuHwIlMjtyiIWSaAIoM/lrDM4WogoVHI7jblNtI5oQqW34YzzkaINkfWToEZHi52GKje+ikQsFQkktDiIP44lNAP/BVYoD3uO4+sKoYsEyjCnFRMHnAPMI3yL7+dYgvhGN1A7shFfZgv2hE6So3XWMVgCGm1NDpoOJODbmUZcZTxpaP3aUKnXDuANYLXy0HEcXk+IAZNAGaa0YsYDXwK+QHi3eR8pbTTk19A0pgF7YgcZQ2X2YZSgRnttHLW7U1E7U0lps+Me5JdsB1YDbygPOwf5tYQYEAmUYaTrNt8zgPMJX97qYQniG1tPzYRaAlnNJNtCJESkyBjVbqG+IoGWnWnY9iWTNsh3mu0iPGspVh7aB/F1hDgmEijDgFZMBuHZyLnw2VqALUBHYRXVBdVYZRZinBD4quKpKsnCsjuVDKUN2p9rO7ACWKY81A/SawjRbxIoQ5hWTDZwGeChKyxMIYLj6zgwpRKV1kamkbfris8LaHjLE6nZkI2zPJH0QVp3CQCrgH8qD+WDML4Q/SKBMgQdKkhymqieWkF7bhMZZoUzkvUNVz4TzaUpNGwYQXxtHCmD8BIKeB94TnkoHYTxhTgiCZQhpCtIFhIOEpPLR9uMMmryakmyB4dEQ8Yhw2uldkM23o3ZZA/CeosivKflWQkWcTxJoAwBBwdJRgu1p+6jPbuZEbIuEt0CGt5dqdT8bxSpg3CnWHew/E0uhYnjQQIlhmnFJANXAWeh0PJrqThpP+aETrIiXZs4NgqC1W4q/zcKa3kimQYPHwBeIzxjaTN4bCF6SKDEIK0YC3AhcLk1iHl6OZWTD5Aol7WGhjYrNety6NiUyYiQydAZZjPwDOEd+NLaRRhOAiXGaMVMB75lD5B22h4OjK8j26w+vzFRxD6/iZZ1I2hYm0OOwcGyF/iT8rDOwDGFkECJFVoxmcB1liAnnrqXyknVZA639ifD1SAGy8fAn2V9RRhFAiXKde1u/4o5xPyT9lMzpZJ0i8IV6brE8ecz0bJmJI0bsslRGiaDhg0ArwBPy+mSQi8JlCimFXOypvjmtAr8M8pItoaOa8dbEaU6LDR8MJq2bRnkGjhsBfCI8rDFwDHFMCOBEoW6zmb/5qQqJp26lzh7kORI1ySiT6uN6pX5aFXxpBs0pAKWE77NWGYr4phJoEQZrZjJqW1ce+523EkdZEe6HhHdFKh9Sex7awKZPotha2rlwG9ktiKOlQRKlNCKsdoCXD67lJMm1jBSA0ukaxKxI6Dh/d8oajdkM9KgfmEKWIasrYhjIIESBbRixhVU8fVZexkne0mEHi12qt7Ix1LrJtWgIcsJr61sNWg8MYRJoESQVowps4X5p+/h3PQ2RkW6HjE0KAjuSaFsVR7ZfjM2Y4bkGeB55UHeMMRhSaBEiO1NUr6wm2+Or2OmCUP+0wvRh99E69t5NO9JZYRBQ34MPCztW8ThSKBEwMy/MXlSNTe6/IyOdC1iaFOgSpPZ92Y+IwzqanwAuF952G3AWGKIkUA5zhY+wvzEdr5mQna5i+Onw0LD6wUEq+NJM2A4H/B75WGlAWOJIUQC5Tj5/rWtrtbxcTdrmnZ6pGsRw5OC4MZsyt4fzSiD7gT7D+Fg8RswlhgCJFCOgx99uW5iw0nJtyiLycidzUIMSJOdA69MJq7VbkjnhZ2EL4FVGzCWiHESKINoqVaiVVyYPb/h5ORFymKS/lsiagQ1Ov47ltotmYa0b2kB7lMeSgwYS8QwCZRBslQriSu/eMS1DScln4WmyamJIirtTKX0zQmMNuASmJ/wHWDvGVGXiE0SKINgqVYyct/Ckdc1T008MdK1CHE09U4qlhWRYkDrFgX8QXl41Yi6ROyRQDHYb80lJ5d+Y8yitjx3XqRrEaK/Os00vjwZVR9nSCPSfwJPyibI4UcCxSBLtRIt4DR/pfSaMZ6OHOfISNcjxLEKgW9VHjU70skxYLhVwBLlIWDAWCJGSKAYYKlW4vDHW769+4ZxBf5kW2ak6xFioBSozZnsfXesIesqawkv1ncYUZuIfhIoOi3VSpJ8SdY7dy0enx2Ms6REuh4hjFATR9myIjIN2F2/C/ip8tBoQFkiykmg6LBUK8n2JVrv3HXz+OSgyyKHYIkhpdnOgRenkuyzYNc5VDnwQ+WhwYi6RPSSQBmgpVrJOH+C5Y6d38lzSpiIocprpfaFqbjabejdR7WfcKg0GVGXiE4SKAOwVCuZ4o+33Lrzu3kWCRMx1HWaaXhhGhYDdtaXAv+nPLQYUJaIQhIox2ipVjLbH2+5ced385SEiRgufCZaXppCoMGl+7bincBd0gJ/aJJAOQZLtZLz/G7LN3bekheQMBHDTUDDu3wyXgM6Fm8D7lYe2o2oS0QPU6QLiBVLtZJzgw7TdTu/k+eTMBHDkUXh+vIm4nMbqdI51ETgp1qxHOEw1Eig9MNSreTMkFm7cee385qCbotRZ3ULEXPMCvv5W0gxIFQKgbu1YjmtdCiRQDmKpVrJGQq+s+dbY6v8KbasSNcjRKSZwDpvK0kZLdTqHGoqcKdWLO9DQ4X8RR7BUq1kNvC9/VeMLG8f6ZKzTIToYlbY52/ClezVvbfkZOA6I2oSkSeBchhLtZKTgR9Uzssqa56cKGe/C3EQi8J1yUYs8R006xzqQq2Y8w0pSkSUBMohLNVKTgB+WDcrpbLu9NRRka5HiGhlDRF/6QZCLp/u24C/pRUz3ZCiRMRIoBxkqVYyGvhh6/i4+soLskegaUacvS3EkGUPkvSV9XTYA7qaQJqBO7RipFN3DJNA6WWpVpIC/MSfYAntvWq0W05aFKJ/nAFSv7KeZksQn45h4oAfa8UkGFWXOL4kULos1UocwI+VmZRdN4z3KqspLtI1CRFL3D4yLiqhBqXrYK0s4C6tWHeXYxEBEijAUq3EBNwOjN/7tdHlgSRrRqRrEiIWpXnJ8exir85hJgE3G1GPOL4kUMK+BZxc7Unf15ofL4vwQuhQUMOYSVWU6RzmTK2YCwwpSBw3wz5QlmolXwa+1DbWVV19boYRR58KMeydsZtMAzY+XqMVM9aQgsRxMawDZalWUgRcE4gze0u/PsYli/BCGMME1gs2Y3f68OoYxgrcLj2/YsewDZSlWkky4XUT055rxjYom8kd6ZqEGEpsIeIvLqHFFCKoY5hc4HqjahKDa1gGSq9F+OTqM9P3dmY75FKXEIMgoZPMudt0r6ecoxXzBUMKEoNqWAYK8DWgqCPT3lB9dsaISBcjxFA2qpHRJ5SzT+cwi7VipDlrlBt2gbJUK5kJLFBmgqXXjAli0uR+dyEG2cx9ZCV5adQxhIvweorFoJLEIBhWgbJUK8kEbgO0sgW5+wPxVr0nzwkh+sEEtgs349eUrvWUCcBVRtUkjDdsAqXXuklc88T4A03TpIOwEMdTnJ90zy726xzmIq2YKYYUJAw3bAIF+AqQH7SZfGULc13S9FGI4y+/htE6T3vUgJvlpMfoNCwCZalWMhZYCFB+SU5FyG6W5nNCRIAG2nnbcNr0dSbOBhYZVZMwzpAPlKVaiQW4FbB4c521zVMSpLWKEBFkC5Ewbys1Ooe5WCtmvCEFCcMM+UAhPDMZqzRC+64cpdC04fB7FiKqZbcwsvCArvUUE/AdOY8+ugzpv4ylWskE4FKAqnMz9wcSrOkRLkkI0eW0UtKd+k56HAdcaFQ9Qr8hGyhLtRIb4UtdZl+StaX2jLTsSNckhPiMWeH44jbqdQ5zhVZMiiEFCd2GbKAQvqtrJMC+r41qxqzJXSFCRJmsVkaOraNCxxBO4JtG1SP0GZKBslQryQIuAWgujK/syHZKry4hotSZu4gzh/DrGOJ0rZjphhUkBmxIBgpwHWBToMovypGZiRBRzBYk8Qu7Kdc5zDVaMbK3LMKGXKAs1UpmAKcA1J2eWhZ0W1IjXJIQ4ijya8hNbNfV62sM4DGkGDFgQypQuvacfAsgZNX8VedkJke4JCFEP2hgOW+7rju+AK6U5pGRNaQCBbgYGAFwYG5WuRyaJUTsSPWSk1er69JXBnC+UfWIYzdkAmWpVpIGfBUgEGf21s9MkXNOhIgxZ+wmTmdH4q9qxTgNK0gckyETKITbWjsAyhbk1shtwkLEHnuQpGkVumYpiYSvVIgIGBKBslQrGU3Xglxniq2pdaI7N7IVCSEGakYZyaYQAR1DXKwVk2hYQaLfhkSgED7SVwOo+PKIJjTNHOF6hBADZA0RP6NM1yzFAVxmVD2i/2I+UJZqJRPpuk3Yl2RtacuLk02MQsS4aRWkWYK6NjvO04rJNKwg0S8xHyjAld2fVMwfUS+zEyFin0URd8o+XbMUC3CRQeWIforpQFmqlUwCTgDwx1vaWvPdMjsRYogorCJT50Fc52jFyNaB4yimA4Vep7YdmJdVi0mTTU1CDBFmhXN2KQd0DOEA5hpVjzi6mA2U3rOToMPU0TQlMSuyFQkhjJZfwwi7n3YdQ1wou+ePn5gNFHpdH60+O+MAZs0ewVqEEIPABLaTyqjSMUQKcIZR9Ygji8lAWaqVZAKzABSohpOS5YAdIYaogmpSde6ev8ioWsSRxWSgED72UwNomppYEbKbEyJcjxBikFhDxE+qolLHEOO0YqYaVpA4rJgLlKVaiQs4r/txzZnpMfd7EEIcmxPL0dtK6SIj6hBHFotvxucRPvaTzlRbU2eGXRbjhRji3D4yspuo1jHESVox0pJpkMVUoCzVSkyEL3cBUH1ORiOaJqe0CTEMnLKPTh1friG3EA+6mAoUwgvxGQAhi+ZvnpyQEeF6hBDHSWYrI+I6adUxxBytOObe82JKrP3h9hyeUz8z5YCymOTcAyGGCQ3MM/dTq2OIJOBEg8oRhxAzgbJUK0kHpnQ/rj81RTYrCTHMjK8lwxTSdQvxmYYVIz4nZgIFOIuuW4X98ZY2X6pNFuOFGGYsCte4Ol3tWE7VinEZVpDoI9YCBYC6Wam1shgvxPA05QAhHV9uA04zqhbRV0wESlffrp4z4htPSJK1EyGGqfRWsqxBfDqGOOvoTxEDEROBApzd/Ulnqq0pkGSVu7uEGKZMYC2o0nXZa7JWjLyHDIKoD5SlWokNOL37cd3s1IYIliOEiAKFVVh1fLkGeAwqRfQS9YECnArEdT9ompaYGMFahBBRIKmDTKePNh1DeIyqRXwmFgKlZwGtPdtRH3RZkiNZjBAi8jQwTTmga0/KSK0YuVPUYFEdKEu1Eiu9NiI1zEhujmA5QogoMrFa9+2/JxlSiOgR1YECTCV8jCcALQXxjiM8VwgxjMT5SU/ooEnHEBIoBov2QDml+5Ogw9ThT7amR7IYIUR0mVBDo44vn6IV626LL3qJ2kBZqpVowMzux01FibVomjmCJQkhoszYel13e9no1c5J6Be1gQLkAandD5qmJaoI1iKEiEIpXtJ09vaSy14GiuZA6bncpUB5R7nk3HghRB8msOU2UaNjiBmGFSOiOlB6Lne1j3TWK6sp7khPFkIMTxNq6NDx5dlaMTmGFTPMRWWgLNVK4oEx3Y8bpyW1RK4aIUQ0y2kiQecQctnLIFEZKMBkulrVA7SNi5M7MYQQh+QKkKLzJMdphhUzzEVroBT1fuBLs8nueCHEYeXVoafH30TDChnmojVQJnd/0pFhb5CjfoUQRzK2XteXJ0gbFmNEXaAs1UqcwLjuxy0T4/VMZYUQw0Cyl3idQ+QbUsgwF3WBAhTSq67WCW7ZfyKEOCJ7kCS7n3YdQ0igGCAaA2Vy7wcdIxx6v/MQQgwDI5plHSXSojpQAi5ze9BpTopgLUKIGJHbpOtY4HFaMRbDihmmoipQuvp39ayfeEe7GtE07QhfIoQQAGS26Gr0aKPX3jcxMFEVKEAWvdrVe0e59HzHIYQYRpI60Huaq6yj6BRtgTK294P2HKfMToQQ/WIJEefuRE9XDVlH0SnaAmVM7wedaXY5UEsI0W85TboO3Bp79KeII4naQFGgAvGWpMiVIoSINTlNBHR8ebZhhQxTURso/mRrC2ZNengJIfotuV3XgVsOrRhp86RD1Nwmt1QrccBn7Q+8I10toLuLqBBiGHF3ovcyeTbo2s/yOZqGCRgJjAcSATugAB9QD+wEKpQi5jdxR02gAKPp1WG4faRTz9RVCDEM2QO6W7CMADbrrUPTGAGcR3hf3VjCIXIk7ZrGLqAEeEMpavXWEAnRFCh9rl92ZDrkDi8hxDExgc3lw+u14RrgEANeR+maicwEvkS4Jf6xvIc5CXdZLwK+qml8DLwKrI+lmUs0BUpG7wf+RKuea6FCiGEquZ0WHYEyYiBfpGlMAL5L+EqLXibCR6CfAmzTNB5RijIDxh100bQon977QTDOLLcMCyGOWUobnTq+/JhmKJqGRdO4CngIY8LkYBOB32gal3TNgKJaNBXYN1DsZjlDXghxzFK9hHR8eb8DRdPIBB4BvsLgvpfagKuBBzVNdzeAQRVNgdJzySvoMHXILcNCiIFI6sCs48tdWvHR37Q1jZHALxmcWcnhFBAOlfSjPjNCoilQev6QfMk2byQLEULELnfnUe+oOpojvmFrGtnAfUCKztcZiBzgXk2Lzv0yUREoS7WSeHo1hfSl2joiWI4QIobZAwNekO/mPtwvaBrxwD1Aks7X0CMb+Lmm6equPCiiIlA46DuCzjS7P1KFCCFimzmke3PjkfayXM9Bd6RGyBjga5Eu4mDREih9pm/+RGvM3HcthIguJrCYQrp6eh0yUDSNU4E5OsY12pc1jUmRLqK3aAmUPlPUoMssmxqFEAPm9Ou6dfhzl7y6LnXdpGPMwaAB342mS1/REih9bhEOOiRQhBADZw/oOg74UDOUiyEqF8JzgLmRLqJbVAZKyGGKlrqEEDHI6UfPOmyfQNE0LIT7ckWreZEuoFu0vHH3ueQVspv03EcuhBjmXH5daygHX/I6HaJ6Q2GupjE10kVA9ARK3xmKTQJFCDFwTn2BcvAlr/P11HKcfCnSBUD0BErfGYrVJI0hhRAD5vTr6tDbEyhdi/FRdSfVYcyIhl5fES+gS58pprJo0dQFWQgRY3QGirPX5+P11nKc2Akf4hVRUfHGvY2Psp/lZ1/oftxwb61r6mnf3XbKF+/fE8m6ROR4W6ssbz63cFpT3a54DY0zLvrd+tLNyzPKdq7M0jSTsjtTOs9Z+Pd1ial5nQF/u7by2a9Ora/alKRpmjp17i83jStaUBfp34OIHNPBcdJQZeHnC6dRviseTYNbf7eeVc9nsXZVJhaLIn1kGz99fh3JmQH6fqOdp7+a+dPgnUxI7oQ9qz/7+bvGwJNjwazAUwV/3aLzhfKAvTrH0CUqAmUipzT+lNffAQgS4AeWM8+dcMLXDkS6LhE5xf+8pih3/Lk1869btSbgb9d8nc3mzNxTWuZc/Pg2gP/950djP1xxR/4Xr/jnxrWr7x8NcMUPdq9uadxre/XP554ypvCid02yFDdsaQcHyoPXFDHj3Bp+vWoNne0a3mYzbU0Wvv/HrVhtigevnsQf7pzAHU9sgT7NJQ2YoVyzH24phWtP+OznXkyFFVmwbTW4QrDXiL0k44G3DBhnwKLlkldPHWv5T3pcwojWlMzC9kgWJCKno63WUlO+JnW65859ABarU7ncmQFHXFrPQqvf12bWug7Ea6je6h4x7sxagPik0T6rPd5fvvPNpEjULqJDn41sTbUWtq9JZVH43xN2pyI5M8DZi2qw2sLRUzirgbqK7pYtvd8XB3TgVl8X1UPGQftiHhsDP9gZDhOA0Xr2zXQzoFZ9omKGQq+//094fcTYSRfFxOlkYnDUV5W47M7kzn//bf4JjbXbElIyChvPvPTJTXZnUrD4X98sKN28LNdqc/u//K13PgBIzZrSvHfrq1lFs75d0VizzdFYsy2puWGPHNA2jJlUr0zZU+IiPrmTH80/gf3bEhhT2MidT27CnRTsec5/nhzJGQsqur+811B6Oxcfxt44WJUCPy0Aewh+uQnOa9I5aMR3zEfLDEUD8NOh7eSTzIKTri2PdEEickIhv9ZUuz2xaNa3Sxd9b8c7Fqsr+OGKO/IAPJf8ces37qp5c0zhl8s/Lb53DMAJc+7YHxef3fGPXxee8e7LNxWlZk2pN5ks0g9uGOtzySvg1yjbnsjF3y7l6R3vYHcFefyOz9ZGHv1OHiaL4tJbut93ek9wBuk9MqhBoxU2/Rce3AxXnYSuc8GAKJggREugKICPeS0jgzFNCclj9fThETEuISWvwxGX3jEqf24jwPipCyvrqzb22Vg26eRvle/f/kY2gNlsU2df9symRd/f+c5F17/7sa+zxZqSWdQWgdJFlFC9IyEnr4PE9A5mhv89cebCSnZ3/Xt67v/lsuatTO5bvpbPGnT0fmfXs5/lCDI64CsHwm/BX2wM30ZQrneGEfEu7dESKCGANazImc65FShdt/yJGJeQMrbT6c5ory77OA5g/44VaYmpE1pqyj/t2QC7c8PfM+NTxrQC+DqazJ3tjWaAneufTdNMZpU56tTWyFQvokFI6/Uekj22k+SMdraG/z3x8Yo0cie0sPLpdJb9No9frviYuMRg7y/v9fkgfWNy/gFYmRb+fE0c+E2Qo3cdJeIHE0Z8itRFtdNqLmVj+jd4cMNeA+Z+IradfsGSkreev/LEUNBvcifmtp17+fPr33zu8mktDaVxaBpx8dlez6V/3QjQ2rTf9toT807VNJNyuNM7zrns2bWRrl9EVujg9rI3Lynh3itPJOg3kZ7bxk+eX8+3ZpxBwGfi1jNPBWD8tAZ+/s+N9H3/KSV89K4O550In6RCsw3Sz4HbtsHt++DiE2DsHLCE4HdrDfj+PuLbLDSlIj8ZWKqV/AI4ofvxpp8XepXFpPfUNSHEMLUtndJVeYwZ4JfXKw9fB9A0vgh827DCBtcvlOJ/kSwgWi559ZmqaQEV8WuBQojYFdD3ztZ73WSnvkqOq4jXGpWBYvKHBmkhTAgxHLRbdb239V5/2wu6zlY5XuqUoj7SRURLoPRZ+NL8MkMRQgyc14qeQ/p6AkUpAsC7+isadKsiXQBEaaCYfCFZlBdCDFi7FT19d1oOevyanlqOAwX8O9JFQPQESp9LXuaOYPBwTxRCiKNpt6LnCIw+gaIUO4Ad+ioaVJ8oRXWki4DoCZS+M5QOmaEIIQZOZ6Acag/Tch3jDbaoqS1aAuXgGUrk72UWQsSsDouuHlwHX/ICWA2s1zHmYFmtFOsiXUS3aAmUPt8RmL1yxUsIMTAKVKcFPc1BPxcoKty9YwkQTV3QG4DHI11Eb9ESKH0OQ7LV++QgCyHEgIQ0OtF03eV1qBkKXesUT+gY12hLlTp0rZESLYFS0/uBvaZTzpQXQgxIwESHziGO1AduBVCsc3wjvKwUH0W6iINFRaAsVkU+oOcsAFu9z3mEpwshxGG1W3UHymHPJem69PUIRPTNfCXw5wi+/mFFRaB06bntzdbocxMNTcaEEDGn2YGe4y8UHPkWXKUIAg8QmQ2PrwKPqijtyB5NgdJz2UsLYtYCKuKtmIUQsafBqevNtkF5jh5IXTvo/x/wDIN2ZkofncAflOLxaA0TiK5A6fNdgbk9GE13UwghYkS9S9celIqjPyVMKZRSPAfcBuzW8ZpHswm4WSleGcTXMEQ0BUqfhXlLa0BObRRCHLM6l65bhiuP9QuUYg/wPeApDnOH2ADVA38AfqjUsdcVCdFywBYcNEOxNfgCHSNkbV4IcWwancTr+PJ+z1B667oE9rym8RJwBnA+MHGANWwg3D/so671mpgRTYFS1vuBo7yD5smJh3uuEEJ8TsCEN2BGz+F8umYCSuEH3gbe1jTGAIVAXtfHKPhc08oA4ZMWdxE+z6REKcr11BBJ0RQoFYTPHbABuPZ79UxbhRDDULuFFohcoPSmFKWEjxAGQNOwAm7C73GK8Ptda9fsZkiImjWUxaooBOzrfuwsb5fpiRDimLTou2UYBnjJqz+Uwq8UDUpRpRTVStE4lMIEoihQuuzp/sTcEXJovtCRdqwKIUQf1XHo6VTeoDy6N0UOa9EWKKW9H9gafVHVp0YIEd3KE3Xd4RWzaxfRIqoDxV7VGQtnOQshooACVZlAio4hdhpWzDAVbYGyp/cDZ1l7tNUnhIhSHRYaAubwTT0DtM2wYoapqHrDXqyKWujVyt61T+70EkL0T72r78mvA7DdkEKGsagKlC49f6muMm8KIeWPZDFCiNhQkaBrQb5JeaLjXPZYFo2Bsqn7Ey2I2dror49kMUKI2FCeSJyOL5fLXQaIxkAp6f3Atc8rTSKFEEekIFDtJlnHEHK5ywDRGCi7gZ7W9fHbWvQssgkhhoE2Gw0h0+famhwLmaEYIOoCZbEqUsCW7sfuHa3JctiWEOJIqt3oOT9JATuMqmU4i7pA6dJz2cviDTrN7cHDHskphBA70nTdLlyuPLrvEBNEb6Bs6v3AUdkhO+aFEIcUAv++ZNJ0DLHVsGKGuWgNlB3wWZO3+K0tcslLCHFIjU5qgyZdpzSuMayYYS4qA2WxKgoAG7sfJ25sSpF1FCHEoexNRk+LphCw1qhahruoDJQuH3V/Ym0OuC2tAdmPIoT4nO3p6DnqYqusnxgnmgPlf4TvvgDAvaNV1lGEEH34TDQ3uEjSMcQnRtUiojhQFquienp1/0xa2ygHzAsh+jiQQKPOISRQDBS1gdKl57JX3O62dC0Qkl3zQogeO9N0bWasU56+Hc6FPjETKJrC5KjoqI1kMUKI6BEC/54U0nUMIXd3GSyqA2WxKioFqrofJ26Q/Y1CiLAaN1V+feefSKAYLKoDpUvPLCVpXWMaSgUjWYwQIjqUZKHp+PIAcruw4WIhUN7p/sTiDTrtVZ1VR3qyEGLoC2p07EolS8cQ65UHWZM1WNQHymJVtA0o636c8nG9zFCEGObKE6nS2V14lWHFiB5RHyhd3ur+JGlNY6ac4ijE8LYhGz3Hg7cDHxhVi/hMrATKKro2OZp9IZuzvF0uewkxTPlMtJQlkqFjiPeUR1e7FnEYMREoi1VRHbCu+3HqB3V6FuOEEDGsNIU6NF0L8nK5a5DERKB06bnslbixKYug6ohkMUKIyNgwggQdX15Dr8azwlixFCgf0HU0sBbEHLe7rTrC9QghjrM2KzW1caToGKJYeZDO5YMkZgJlsSry0esW4oziaj2LckKIGLQuB71XJt42pBBxSDETKF1e7f4kbo83w9Lil1YsQgwTARNtmzIZoWOIHcrz2RYEYbyYCpTFqmgvvRfn36vzRq4aIcTxtC2dGp17T2R2MshiKlC6LOv+JPWDumxZnBdi6FMQ+CRX163CbfS6sUcMjlgMlE/p2jlv8itrwpZm2ZMixBBXkUBFuw2XjiH+I61WBl/MBcpiVaSAl7sfZ7xZnSDnzQsxtH00SleYBIHlRtUiDi/mAqXL20ALgKO6M9lW65NZihBDVLOdA9XxpOkY4j3lQW7gOQ5iMlC6biH+d/fjjFXV0jBSiCFqTS4BnUO8ZEgh4qhiMlC6vALhfjyJ65pGmL2BhgjXI4QwWLuF+m3p5OgYokR52GlYQeKIYjZQFquiRuB1AA209FU1LZGtSAhhtA9H066zb9cyo2oRRxezgdLln0AnQOoHdTmmzmBzhOsRQhikzUrNtgxds5MK4H9G1SOOLqYDpc8sJYQ57d1auewlxBDx3lj0nnv0svTtOr5iOlC6vEj4wBzSVtfmaL5Qa4TrEULo1GKnaneqrjYrtcBKo+oR/RPzgbJYFTXTdZ3UFFSWtPdq6yJbkRBCr3fG6p5ZPKM8umc44hjFfKB0eQloBkhfVTNCZilCxK4GB5X7k8nSMcR+pG9XRAyJQFmsitqBfwCYAsqa8XZ1fYRLEkIM0OrxWHQO8aTyEDKkGHFMhkSgdHmN8HcmpL1TO9LcFpBQESLG1MRRdiCBdB1DbFUePjKsIHFMhkygLFZFQeD3EN6XMuLlCulCLEQMCYH/P/m6jvcFeNKQYsSADJlAAVisijYA/wVILGkeYavprIxwSUKIftqSSXmLQ1egfKI8lBhWkDhmQypQuvyZrs2OuS+UWaUTsRDRz2em6f0xujYxKmR2EnFDLlAWq6Ja4HkAV1l7mmuPV478FCLKvTOO1qAJq44hipWHUqPqEQMz5AKly0tAJUDui2VJhJTcjy5ElKqJo2xnmq7ZSTsyO4kKQzJQFqsiP/A4gK3RH5/6QV15hEsSQhxCUKNzxUSSdA7zjPIgG5qjwJAMFIDFqmgNXWdIZ/37wEhza0D+wQkRZdaNoKrNjlvHEHsIH2UhosCQDZQufwRqtRDmUc/uC8oCvRDRo9VG9ccjGaljCAX8TjYxRo8hHSiLVVEbsAQgbo83I35ry74IlySEAELge3USNp1nnbyhPGw1rCih25AOFIDFqmgtsAIg9/myLOnzJUTk/W8UBxpdutZO6oAnDCpHGGTIB0qXPwNV5s6QfcQrFY2RLkaI4awmjrJ1OYzSOcxjyoPXkIKEYYZFoCxWRR3AbwCVvKYx11HRLnd9CREBfhOtr04iVecw/5V+XdFpWAQKwGJVtJGuc1PGPFGarPlDbZGtSIjhRYF6awItnVacOoZpoqtnn4g+wyZQuvwN2GJpC7pyXyhrinQxQgwne1LYV5pCto4hFPBr5UH+70apYRUoi1VRAHgQaEosaR4Rv7lZ7voS4jhot1D35gRdu+EBlikPawwpSAyKYRUoAItVUR3wEKBGPrc/W85NEWJwhcD3SiGmkEnXwVk7CF9hEFFs2AUKwGJVtA74uymgrGOeKFWEVCDSNQkxFClQxXnU1MeRrGOYduD/KQ/y/zTKDctA6fIPYI2zoiM17b+1cteXEINgcyZ7t6frvtS1VHmQs41iwLANlMWqSAEPAzWZK6pG2Ss7JFSEMFB1HGXvjmW0zmHeUh5WG1KQGHTasba3WrNmTYbFYvkTUMQQCCQVwBJoCU/Hgy6LUkPg9zScKDTV7LW1rFw7cn1bh90X6XpEmNdK3d+nEx8wY9MxTDlwi/Igx3nHiGNeJLNYLH/KysqalJ6e3mAymYZEs0VfS7DZWxHIxqYFOxOsZtD09BcSx5FSIVqa69znsn/asg/yPo50PQICGt6XirDqDBMf8EsJk9gykO/Gi9LT05uHSpgA2OLNXke6pUb5lMXWEfSHb3cXsUDTTMQnpAYSXL74SNciQEHwjYk06zwbvnu/yW6j6hLHx0ACxTSUwqSbI8XcbE8yN9AatFmDoc5I1yP6T9NMaCiZVUaBj0dStj+ZLJ3D/F15+K8hBYnjStYLenFmmOuscaZWrSFgN6vQEY8NHjfKnXi86jqcvaW7TWfMLpTvzEVU2JRJ6ae5uhfhi5WH5wwpSBx3Eii9aJpG3AhLlcWhdZgb/WYTsj9FiP7YlcLed8cxRucwW+g6v0jEJgmUg2gmTcXlWCvMZs1vbfJrGqrfp8Ht3rXD9NUF58ad7ZnuvmDebPe2bZtM3T8/99yZ7nPPmuG+757/c/Se3Sx55H77uWfNcM85vSj+/nt/5IDwzOO0UybGf/fmq52nz5oUf+nFZ8e1t4c7da/99H9mzxlT4uedO9P95z89qmfRUwhDlCWyb+VE3TOTKuBe5eGIVwZEdJNAOQSTRQu5R1orTBC0tQRC0L97q7936zed9z3w2/a3ite2/uRnD7Xf8f2bnAA/+uF3nNd+8+bOlW+vac3KGtETUG+9+bpl9+4dpv+89XHrqnfWt2xY/6n5vf+uMgOUlu4yXXvdt33//WBLS0JCklq+7HkrwC3fucZ1z32/8f575f/koDARcVVuyl6dpOsYXwAv8DNp+hj79PTWGdJMFi3oHmktb9nvz7V5AwGfy2I50u3Era0tfLrmI8s3r/lKXPfP+Xzhtf21n35kfurvy/0Al371a757fn6nE6B41X8s765+y3rWF06wALR5W9m1a7s5N3d0IHfk6NDUaTOCAFOnnRjct6/U1NTUSHNzk3b6GWcFAb562VW+VW+/YR20PwQhjqDeScXLk8nWeYxvCHhQedhvVF0iciRQjsBkDYdK6z5/js0c9PvsZhuH+b8TCoVISEhUq97d0NLf8ZVSfPs7t3dcfe3iPhvy9pbuNtls9p7HZrOZjo52TSml6fu/K4Qxmu1U/XMq6SETZp1DPaY8fGpIUSLi5JLXUZitWsA90lpuag+ZbP7D306ckJBI7sjRoZf+9awVwhvuNm741ARwwokzgy93XbL654vP9Kx7nHnWFwPPPftXW2trOIMqyvdr1dUHDpsYSUnJKj4hQb3/3mozwIvPPy1rKOK4a7NS8+JUEoMm9M6O/6g8rDCkKBEVZIbSD2abFnDnWspb9/tzrAmaz2822drbvUwryunZvPXN67/b+djjf/fe/v0bnL/59X2OgN/P/Iu+6psy9cTOe+79TfviG690Pf7Yr+1nnzPPHx+foADOPuf8wPZtW3znf/HUeIA4V5xa+vtnvGbz4b/pe2TJX7y3fvdal9PhVHPOPE/uQhPHVZOdAy9OI8Wvbxc8wBPKw3JDihJR45h7ea1fv7502rRptYNUT1QL+kKW1rJATjDeovwmU7//Q3m9bTidTjTNxIsvPG1d9q/nbE8/+6ocQWygPbs22Z58a/Jbka5jKKt1Uf7SFDIMmJk8rTz8w5CiRFSRGcoxMNtMgfhR1rLW/f4RuC2+/obK2k8/Nv/ohze7lFIkJCSqR5Y84R3sWoUwUmU8+5dPZoTSdK+Z/EPCZOiSQDlGJosWdI+ylreV+7Nx9i9UTjvdEyx+d2O/F+uFiCZ7ktn7xkRG6bybC+BfysPThhQlopIsyg+AyayF3LnWCltn0GcNhaRluhiyNmdQ+kYBow0Ik+XKwxOGFCWilgTKAIV31FsO2APBdgkVMdQoUGty2PvOeN3tVABeVx7+aMA4IsrJJS8dNE0jLttaY6oJJBNUbr/ZbD/6VwkR3ULgWz2e6m0ZutupQLhz8LMGjCNigASKAZzplgZzc9DX1h5M9llMjsNtfhQi2vnMNC2fTLA2jlydQ4UInwX/HyPqErEh5i55HThwwFxQUFBYUFBQmJaWNi0jI2Nq9+OOjo5Beye/7bbbRvz4xz/O7P1zOTk5UyorKy0Ap8w5aWR8olZduXNH5/P/eHJQg/qB++5yPPqbB2Q2JAzV6KDymenYauNI0TlUJ3CPhMnwo/uNb9lZe/KMKKTbRW+P3XmkX8/Kygpu3bp1M4Tf5N1ud/DnP/95Vfev+/1+rNbj395q7dq1WwH2V21veun5pxMuu+wqn0KLucAWw9PuFPauzGek0nR/k9lMuNHjdiPqErFlSLzhLViwYMx1112Xe8opp+TfdNNNuatWrXJNnz69YNKkSYXTp08vWL9+vR1gyZIlqeedd974M844Y8Lo0aOLbrjhhlyAQCDAggULxkyYMGFyfn5+4c9+9rOMY63B5XJNB7jr7h+N+HjNh9Y5pxXFLX30l9bNmzeYzjv7JPeZZ0yNn3N6UfzOHds+92f+wvNPWb9w2uT4M2YXxv/07u85un9+3Ch34i9+dofDc8aU+LnnznRXVVX2mYHt3rXDdPacE9zdj3fu2GY62zPdjRD9pCDw/mj2/Wciow0IkyrgBxImw9eQWUPZtWuX47333ttusVior683/e9//9tqtVpZtmxZ/O233577xhtv7ALYvHmza/369ZudTmcoLy+v6Pvf/35VZWWltbKy0rpjx45NALW1tYfcvPX73/8+8/nnn0/tflxdXf25qdC9995b/vDDD2euWrVqZ0dLyH3Dd24afd03b+786sKv+32+ToLBYJ/nV1aUaffd83/O/7z1SUtycoq69OJz4l5d/qL1gvmX+r3eNk466dTA3T95sOMnd9/m+NtfH7f94I6f9vQTGzd+Qig+IUFtWL/GPHXajODfn/mz7auXXSV3nIl+8ZtofX0S7ZUJjDJguF3AT5WHRgPGEjFqSMxQAC655JIGiyWcj/X19ebzzz9//IQJEybffvvtI7dv397zXf/pp5/enJqaGnS5XCovL69j165d9oKCgs79+/fbv/71r4988cUXE5KTk4OHeo0bbrihauvWrZu7PzIyMo54GJAj3tR6ysnTq5c8cr/tkV/da9+3t9TkdLr6POfTNR+ZT511RiA9PVNZLFYuuXSR/4MP3jEDWG025p7/5QDA1GkzgmX7Sz/397Xoimt9f3/mL7ZgMMAry1+wXvrVK+WAInFUtS7K/z4dU2UC6QYM9xFwp4SJGDKB4na7ew6uuuOOO3LmzJnTsmPHjk2vvPLKTp/P1/P7tNlsPc3LzGaz8vv9Wnp6erCkpGTzmWee2fK73/0uY+HChWOMquvGm66vWfbSP7c6HRbv5V+dG7e6eGWfWaFS6rA3ElgtVrSuZRiz2UwgEPjcc+df9FV/8ao3LCtef9kyZcqJwdTU9GNrziaGlRD4PhrJvhenkdNuw3X0rzjacDxJ+KTFDgPKEzFuyARKb83Nzebc3FwfwOOPP552tOdXVlZagsEg3/jGNxrvueee8o0bNw74P1piYmKwtbW155LZ5s2bbZMKJ3X++O479pxzzjlNm0rW9QmFGSfPCnz0wX8ttbXVWjAYYNm/nrPOnj2n312EHQ4nc+acG/jhHd92LVz0DbncJQ6r1UbN89Pwrs015BJXE3C38vCi8iDfxAhgCK2h9HbHHXccuO6668YuWbIk64wzzmg+2vNLS0ut11577ZhQKKQB/PznPy8b6GvPnDmz3WKxqIkTJxYuWrSotqOjw/TCCy+kWiwWlZ6e7v/FvffuMVtIDQZwAmRn56g7f3RP+yVfPtOtlOKss+f6v3ThgmNqS7/gK1f6Vvz7ZevZ58yTdvbicxSEtqWzb/V4RhrQ3BFgK/CA8lBnwFhiCJH29REQVGhN7aT5O0hA6d8F+ehvHrA3NzdpP7r7/mF72UHa1x+az0zTiol0ViRyzHcuHsYrwF+UB/nmRXzOkJyhRDuzhkpxUdNkobOznSQVGPhhRVddMd+1d+9u8z+Xvd1qZI0itilQ+5PY9598sgNmEg0YsgN4VHl4x4CxxBAlgRJBiTaaOy20N3aQSidOQsd+OeJvzyyXs1VEH21Walbmw4EEQ3pxAewFHlQe9hs0nhiiJFAizG7Cn+HkQJ2F5EAnbs2HDWkGJgYgqNHxaS5Va3IMObsEIAj8E3hWLnGJ/pBAiQKaBmk2GtrNeOvMpFl9WAjqPmZVDBMKVEUC+9+cQFq7zbBZyT7g18rDEVshCdGbBEoUcZrpHOGkotpCatCHy+LDghqat3YLY3it1L01gUB5oiG3AkN4b8mLyKxEDIAESpQxaagsG7WtZpw1ZtIdATD5sRpxN5gYOgIa3vUjqPlkpCENHbvtJzwr2WHQeGKYicnvfvft22e54IILxo0cObJo/Pjxk+fMmZO3YcMGe2lpqXXu3Lnj+jtO7/bzAK+++mr8mWeemQfwzDPPJP7f//1fFsBTTz2VtGbNGsfhxjHCwbW4zbSPclBmsuNtcRBSVnxosoFsuAtqdG7KpPSvJ2P5eJQhDR3hs1nJdyVMhB66ZyhnrcPQ9vVvn3Dka7ahUIj58+fnLVq0qO7VV1/dDfD+++87KyoqrFOnTm1dsWLFbiPquOKKK5oI7wZm2bJlSYFAoGnGjBnHdZ+HWSOUaaM20UJTtY90vxlzXJCg5seKLNwPKyHw706l/N2xZHZaDTmWt9tm4I+yViKMEHMzlFdffTXeYrGo22+/vab752bPnt0+d+7c1m3bttkmTJgwGeCTTz5xTJkyZVJBQUFhfn5+4caNG4/pQKolS5akXnXVVaNWrlwZ9+abbybddddduQUFBYWbNm2y33PPPRnjx4+fnJ+fX3jBBRd8bkbk9Xq1Sy+9dEx+fn7hpEmTCl955ZX47jEP1T6/t+9+97sjfvGLX/RsQrv55ptzHrrvnuRRDirS7VS3WdHaHISwIG1WhoEQ+HelsvepGfjezGdMpzXcYcEA1YRvBb5DwkQYJebWUDZs2OCcNm3aUfdePProo+k33XRT1Y033ljf0dGhBQKHXl+cM2dOvskUzlWv12saP358n1nIueee23bOOec0XnDBBU1XX311A8CZZ56ZtXfv3o1Op1MdqtX9gw8+mAGwffv2zWvXrnWcf/75E3bt2lUCh26fn5eX19Mh+Kabbqq9+OKLx999993VwWCQZcuWJX/88cdbABIstLnNeOv8JDdoJFstBFwBQgRkxjLUhMC3J4XK/4419M4tCG9QfAFYpjzyTYkwVswFSn/NmjWr7aGHHsouKyuzLVy4sGHKlCmdh3re6tWrt2dnZwcgPPt5+OGHMw/1vN4mTpzYfvHFF4+dP39+4xVXXNF48K+///777ptvvrkaYPr06R0jRozwbdy40QGftc8H6G6f3ztQJk6c6EtKSgq89957zsrKSuvkyZO9WVlZPe30TRoq3UZ9ooXmaj9pjRpui4WgK0Sga/E+5mad4jOdZho3Z9K0Nocsn8XQIFHA28DflId6A8cVokfMvflMmTKlff369UftBnzDDTfUv/zyyzudTmdo3rx5+cuXL483qoZVq1btWLx4cc2aNWvipk2bVuj39z2C5Ej90Q7VPv/g51x99dW1f/rTn9KeeOKJtKuvvvqQDfhsJgK5dg7k2Cm3mPE3m7E3OdACdnyY5HbPWNPooPLNPCqeOJnEj0Yz2mfhmC7RHsVm4Dbl4REJEzGYYi5QLrzwwhafz6c9/PDDPW3pV69e7Xrttdf6HH27efNm26RJkzrvuuuu6vPOO69x3bp1A7727Ha7g83NzSaAYDDIrl27bBdeeGHL7373u7KWlhZzU1NTn8tep59+euvTTz+dArBhwwZ7ZWWlberUqf1e0P/a177WuGrVqsT169fHLViwoOmItZlpH+2gLMdOud1EZ6sJW6MdS6cDPxbksK0oFgL/viT2PT+Vhuemk70znREG7XDvtge4T9ZJxPESc5e8TCYTy5cv33XTTTeNfOSRR7LsdrvKzc3tfPTRR/v0GXrqqadSereNv//++ysG+ppXXHFF/Y033jjm97//feZzzz2365prrhnT0tJiVkpp119/fVVaWlqfEx5vv/326q997Wuj8/PzC81mM48//nip0+ns9y2/DodDzZ49uzkpKSnYfQrl0bjNtLvNlLUGcdb5SW0P4Wi3gtVC0BkiaApgHkivMGE8r4X67ek0r80lq9Ni2IbE3nYDzwIfyVkl4niS9vVRKBgMMnny5MIXXnhh1+HWfo6mO1g6QvTsn3Eo/PYgSgsMvR340d6+3memaU8KDRuySaiLI2WQXmYb8ILy8NEgjS/EEcXcDGWoW7NmjePLX/7yhHnz5h32RoL+OGjGktIRwtmhYe2wgGZBOUP4bEEgKLvwB0tAw1uWRO3GbBzl4fNIjGgjfyhrgH8qDxsHaXwh+kUCJcrMmDGjo6yszLA3hq5gKe8IYWvwk9gSJF6ByWvC5jWByUrIGcJvDYfLkJu5HG8BDW9VPLWbsrDuSSFDaYNySQvAD7wHvKQ8GLKZVwi9JFCGCYcJX7admgxFXVOA+MYAiX6FLQSmNhM2TIAV5VD4bSFCpiBmgvLv42gUBJsd1OxLomNHGu5qN6kMXohAuN/WG8DbykPLIL6OEMdM3jCGGbNGKMVKU4qVptYgzsYAiW1B4ghvjNQ6NKwdZsAMZkXIoQjI7KWvTjMNFQk070rDUppMesBM1iC/pI/wbGSF8rB5kF9LiAGTQBnGui6HtftDmBsDJLQESfCrz85hCWqY2rSe2QtWRcCmCFlCoIUwEcTMEN+hHwJfq536ajcdFQlY9ieR0OIgGUg+Di+/l/BsZJXyIEc8i6gngSKwmgim22hIh4b2ILbmIPFtQdy9wwXAr2Hxa/TsXtJA2RQBW4iQuStkVCh2QyYEfq+Nxpo4vBWJmMoTcNe7SEIb9BlIb7XAh8Bq5WHrcXxdIXSLyUDRNG3GddddV/XHP/6xDODHP/5xZmtrq/lXv/rVgPeadHO5XNO9Xu/a7sdLlixJ/eSTT+L+9re/7fvlL3+Z7nK5Qt/+9rfrlixZkjp//vzmMWPGDKnNg04zPqeZOqCuPYi9OYi7NYg7oD5/gqQCrVPD0tl1iQzCIWNRBCwKZVEokwJNoRHCRAgTEQ6bEPg6LbS02ehochBocKHVu7A0OHE1OolXGukRKKsc+AB4H9gpe0dErNLfvv4sg9vXv330Hb02m029/vrryZWVlQe6+3AdD707HD/99NNpJ5xwQvtQC5TenGY6nWY6M6HOG8TeEsR9qJlLbwq0npnMQbrCJmiBkKk7bAAtHDp03b6sobo+Dq/3r4WfbSLU5OCAz0yg00KoI/xBuxWt3Yqp0Ym9wYm73YYLSB3Yn4ihdhEOkA+Uh/1He7IQsSAmZyhms1ldddVVNffdd1/mo48+Wt7717Zv3277+te/Pqaurs6Smpoa+Nvf/lY6YcIE34IFC8bEx8cH169fH1dTU2P9xS9+UdbdPbi/brvtthFutzs4duxYX0lJieuqq64a53A4Qp988smW22+/fcQbb7yRZDablcfjaf7DH/5QZuzvOrJcZjpdZjqBOl8IS1sQlzeEsz2EM6j69++oK2zMfjAbOE/RANotmJ6dflwvTR2rFmATsIHwDvbqCNcjhOFiMlAAfvCDH1RPmTJl8k9/+tMDvX/+hhtuGLVo0aK6m2++ue6RRx5JvfHGG0e++eabuwCqqqqsn3zyydZ169Y5Lr744rxDBUpnZ6epoKCgsPtxU1OT+dxzz+3TT+vqq69ueOyxxzIeeuih/V/4whe8VVVV5tdffz159+7dJSaTiUO1tB9KbCYCNhPNydAM0BHC6u0OmCDOENLiBagnHCAlXT/uk0tZYqiL2UBJSUkJfeUrX6l74IEHMpxOZ6j759euXRv373//exfAjTfeWP+zn/2s5xCr+fPnN5rNZmbMmNFRV1d3yMs2drs9tHXr1p5bM7vXUI5SS9But4cWLlw4+ktf+lLTZZdddsSGjkONw4TfYaIpBZqUgo4Qdm8IR2cIuy+E3aewqRhdqO8nBRwg3NV3E1CiPFRGtiQhjr+YDRSAH/7wh1Unnnhi4cKFC/vVW8zhcPR8h3isPcyOxGq1sm7dui3Lly9PeO6555Ife+yxjA8//HC7YS8QQzTts7WX7p9TCjoVto4Q9s4Qts5wyNiDKuZmMgqoAfZ1fezt/lwOqxIixgMlMzMzeOGFFzb8/e9/T7v88svrAKZPn972pz/9KXnx4sX1jz/+eMpJJ500KPfvu93uYHfb+qamJlNra6vpsssua/J4PK35+flTBuM1Y5WmgUPD5zD1fdP1hzB3Kmz+EFa/wuJXWAPhD0tARez24zbCl6saun6sB8r4LDjaI1CTEDEhpgMF4Ec/+tGBJ598sudWz8cee2zf17/+9TG/+c1vsroX5Qfjda+66qram2++efQPfvCD0BtvvLHjggsuyOvs7NQA7rnnHrlrpx+sJoJWaMf8+TdppSCoMPsV5u6ACYFJgSmk0Pr8GO536SZ8yckKBAn3uvIBga7Pux/3/ryJz0KjAaiXmYYQAyft68WQsH79+rRp06aNiXQdQgxn0ptJCCGEISRQhBBCGEICRQghhCEkUIQQQhhCAkUIIYQhJFCEEEIYIiYDZd++fZYLLrhg3MiRI4vGjx8/ec6cOXkbNmywl5aWWufOnTvueNWRk5MzpbKyss9enqlTpxYUFBQUZmdnT0lOTp5WUFBQWFBQULht2zbbYNayZMmS1KuuuqrP0bMzZ86c+M4777gA5syZk1dbW2uura01P/DAA7pbtC9YsGDME088MeiHTN12220jfvzjH2cO9usIIfTTvbHx188Y277+1iuO3L4+FAoxf/78vEWLFtW9+uqruwHef/99Z0VFhXXq1KmtK1as2G1kPcdqw4YNW6HvOSq9f93v92O1Hrb7+6BZvXr1ToBt27bZ/vznP2fceeedNUf7msESCoVQSmE2x1rnFSHEkcTcDOXVV1+Nt1gsqvfZJLNnz26fO3du67Zt22wTJkyYDPDJJ584pkyZMqmgoKAwPz+/cOPGjfaDx7riiitGFRUVTcrLy5t86623juj++ZycnCm33nrriMLCwkn5+fmFa9eudQAcOHDAfNppp02YNGlS4aJFi0b3d1PobbfdNuLyyy8ffdppp0245JJLxm7bts02Y8aMiYWFhZMKCwsnrVy5Mq779zZz5syJc+fOHTd27NjJ8+fPHxsKhfte3nTTTTnjx4+fnJ+fX/itb30r94gveAjds6nvfe97ufv377cXFBQUXn/99bl79+61nnTSSRMLCgoKJ0yYMHnFihXug7/2cK+9evVq9/Tp0wtyc3OndM9WmpqaTLNmzcrv/rN7+umnkyAcZOPGjZt85ZVXjpo8eXLhrl27bHfffXdmUVHRpPz8/MLef/533HFH1pgxY4pmz56dv2PHjs/9vQkholPMtV7ZsGGDc9q0ad6jPe/RRx9Nv+mmm6puvPHG+o6ODi0Q+Pw5XL/61a/KMzMzg4FAgNmzZ0/86KOPnKeccko7QFpaWmDz5s1bHnjggfQHHngg8x//+MfeO++8c8SsWbNaH3roocrnnnsu8dlnn007hrpdH3300Va3261aWlpM77777naXy6U2btxov/zyy8eVlJRsAdiyZYtz3bp1u8eMGeOfMWNGwcqVK90nnHBCe3/a47/yyivJBQUFPYGwb9++z70ZP/zww2UXXHCBs7uj8k9+8pPMs88+u+nBBx88EAgEaGlp6fNNxpFa8x/qOACXyxV67bXXdqakpIQqKystp5xySsGiRYsaAUpLSx1//OMfS59++ul9//rXvxJ27tzp2LBhwxalFOecc07ev//9b7fb7Q699NJLKRs3btzs9/s54YQTCqdPn37Uv28hROTFXKD016xZs9oeeuih7LKyMtvChQsbpkyZ0nnwc5588smUv/71r2mBQECrqamxrl+/3tEdKIsWLWoAmDlzpnf58uXJAB9++GH8v/71r50ACxcubLr++uuD/a1n7ty5jW63WwH4fD7t2muvHb1582anyWRi7969PW/8U6ZMaRs/frwfYPLkyd5du3bZzjrrrNb+tMe/8MILG3pfYps5c+bEo9V16qmntl1//fVj/H6/6dJLL22YPXt2n75aR2rNf6jjAEKhkHbLLbfkfvjhh26TyUR1dbWtrKzMApCdne07++yz2wBWrFiR8M477yQUFhYWAni9XtPWrVsdLS0tpvPPP78xPj4+BHDeeec19vOPWAgRYTF3yWvKlCnt69evdx3teTfccEP9yy+/vNPpdIbmzZuXv3z58vjev75161bbb3/728zVq1dv3759++azzjqrqaOjo+fPo7vVvcViUYFAoKfrrck0sD+yuLi4njNb7r333syMjAz/li1bNnd9J94zqN1u77mOZjabCQQCWnd7/AULFjQuW7YsyePxTBhQEYcwb9681nfeeWdbTk6O7xvf+MbY3/72t32Oxz3Sax/qOIDHH388pa6uzrJx48YtW7du3Zyamupvb283AbhcrlDv599yyy2VW7du3bx169bN+/btK7n11ltrATRtKB+dIsTQFXOBcuGFF7b4fD7t4Ycf7rnctHr1atdrr73W59r/5s2bbZMmTeq86667qs8777zGdevWOXv/ekNDg9npdIZSUlKC+/fvtxQXFyce7bVPPfXUlr/85S+pAM8//3xCc3PzgFaVm5qazNnZ2X6z2czvfve71GDwyBOdpqYmU319vfmyyy5r+v3vf79/y5YtRw3Uw0lMTAy2tbX1/L1v377dlpOT4//e975Xe+WVV9Z++umnfcY+1tduamoyp6Wl+e12u3rllVfiKyoqDnl327x585qfeuqptKamJhPAnj17rOXl5Zazzjqr9bXXXktqbW3VGhoaTCtXrkwa6O9VCHF8xdwlL5PJxPLly3fddNNNIx955JEsu92ucnNzOx999NE+LeOfeuqplBdeeCHVYrGo9PR0//3331/R+9dnzZrVXlRU5J0wYcLkUaNGdc6YMeOo56Y88MADFQsWLBhXWFg4adasWa3Z2dkDanV+yy23VC9YsGD8smXLkk8//fSW3idOHkpjY6PZqPb4WVlZwRkzZrROmDBh8llnndVUVFTUvmTJkiyLxaJcLlfwmWee2aPnta+77rr6efPm5RUVFU2aPHmyd+zYsR2Het4ll1zSvGnTJsfJJ59cAOHZyzPPPLPn9NNP91588cX1RUVFk3Nycjpnzpw5KOfZCCGMJ+3rxZAg7euFiLyYu+QlhBAiOkmgCCGEMIQEihBCCEMMJFBCoVBI7usUUaPr3+MRb2wQQgy+gQRKSU1NTaKEiogGoVBIq6mpSQRKIl2LEMPdMd82HAgErjtw4MCfDhw4UIRcMhORFwJKAoHAdZEuRIjh7phvGxZCCCEORWYYQgghDCGBIoQQwhASKEIIIQwhgSKEEMIQEihCCCEMIYEihBDCEBIoQgghDCGBIoQQwhASKEIIIQwhgSKEEMIQ/x+/CDcvEHw4XAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 504x504 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from collections import Counter\n",
    "from matplotlib_venn import venn2, venn3\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "sets = Counter()                  \n",
    "sets['100'] = 7   \n",
    "sets['010'] = 62   \n",
    "sets['001'] = 16    \n",
    "sets['110'] = 639    \n",
    "sets['101'] = 0     \n",
    "sets['111'] = 0 \n",
    "labels = ('Cis', 'Trans', 'Non Hits')  \n",
    "plt.figure(figsize=(7,7))\n",
    "ax = plt.gca()\n",
    "colors = ['darkviolet','deepskyblue','blue']\n",
    "v = venn3(subsets=sets, set_labels=('','',''), ax=ax,set_colors=\n",
    "          colors,alpha=0.7)   \n",
    "\n",
    "\n",
    "h = [] \n",
    "for i in sets:\n",
    "    h.append(v.get_patch_by_id(i)) \n",
    "l = ['Trans Hits only','Cis Hits only','Non Hits','Cis and Trans Hits shared ','',\n",
    "     '','']\n",
    "ax.legend(handles=h, labels=l, title=\"Legend\",loc=3)  \n",
    "#plt.title('Number of SNPs Hits ')\n",
    "output_file = \"/home/s417377/lastversion_gwas_analysis/overlapping_result/Perm_GWAS_Hits.png\"\n",
    "plt.savefig(output_file, dpi=700, facecolor='w')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "b609a42d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-155-c91a4d062a27>:25: UserWarning: Legend does not support None instances.\n",
      "A proxy artist may be used instead.\n",
      "See: https://matplotlib.org/users/legend_guide.html#creating-artists-specifically-for-adding-to-the-legend-aka-proxy-artists\n",
      "  ax.legend(handles=h, labels=l, title=\"Legend\",loc=3)\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZQAAAFNCAYAAAAn9JlHAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAABIkElEQVR4nO3dd3zV9b3H8dfvd3ZyErIHCXuHBAQUBQenoF6xQou0VbF1VHtV6HC02lbtsNZqr7YWi6PaWqu2jqqIqFi1BKyzKAQie68MErLXWd/7xznBAAkJycn5nfF5Ph55AOckOZ+E5LzP7zs+X00phRBCCNFXutEFCCGEiA0SKEIIIUJCAkUIIURISKAIIYQICQkUIYQQISGBIoQQIiQkUIQQQoSEBIoQQoiQkEARQggREhIoQgghQkICRQghREhIoAghhAgJCRQhhBAhIYEihBAiJCRQhBBChIQEihBCiJCQQBFCCBESEihCCCFCQgJFCCFESEigCCGECAkJFCGEECEhgSKEECIkJFCEEEKEhASKEEKIkJBAEUIIERISKEIIIUJCAkUIIURISKAIIYQICQkUIYQQIWE2ugAhekorRgccgBXwAz7AG/y7R7lQBpYnRNzTlJLfQRF+WjEOIBvI6vCWAiQA9g5vjg5/t3bzab1Aa4e3FqAeqAaqgn8e+bty0RjKr0mIeCeBIvqNVswAYCSQxxeh0R4iTgNLa9fKFyFTAewCdgK7lItmIwsTIhpJoIiQ0IpJIhAeozr8mWFoUb2nCATMTgIhswvYqVwcMrQqISKcBIo4aVoxZmAsMIYvwiPb0KLCowHYAZQE37bLvI0QX5BAET2iFTMYmBR8G09gTiPeNfBFuKxVLioMrkcIQ0mgiE5pxViBU4DTgm/phhYUHcqBdcG3Epn0F/FGAkUcoRWTCJwJnAFMpPtVVaJrPgJXLv8BPpRwEfFAAiXOBfd2TAZmAVOREOkPXgJXLasJhEurseUI0T8kUOKUVsxQYCbgAlINLSa+tAIfAu8C62VSX8QSCZQ4EtwX4iIQJMONrUYQ2GD5LvCGcnHY6GKE6CsJlDigFTMKuBiYBpgMLkccz0tgOOxV5WKn0cUI0VsSKDFKK0YDpgDzgUKDyxE9twF4FfhEhsNEtJFAiTHBTYczgHnAEIPLEb1XBiwD3pFJfBEtJFBiRLDZ4gXAXKK35Yk4XhOwAnhFuagzuhghTkQCJcoF945cDHwZSDS4HNF/WoBXCASLXLGIiCSBEqWCQ1uzgcuAJIPLEeFTBzwPvKlceI0uRoiOJFCikFbMGcBVBNrCi/hUDjwDrJbJexEpJFCiSHD577eRVVviCzuBp5SLz4wuRAgJlCigFZMJXAmcA2gGlyMi03rgT8rFHqMLEfFLAiWCacXYgEuAryA9tkT3vMA/gReUC4/RxYj4I4ESobRiJgLfBXKMqsHsw53opiXRTZvTjcfZhjfBjTL7QVegK7Tgn0f+rinw6qg2M8ptQmszo7Va0NrM6K1mTG1mTI1WHE1WEtHkaquf7AMeUi42GV2IiC8SKBEmuAz4GuC8/n4skx9PehN1GU20pDfjS2vG5GzDZvNhM/tw6P14VaTA5zbR0GyhpcGOu86OqnFgrnFgrU4k2W2WA7z6SAFvEphfaTa6GBEfJFAiiFbM6cAi+qH7r9VL6+BaqvPq8GY0YUpuJdHqI0WLwDkZBcptoq7WQUN5Ev4DA7CVJ5HqNmMzurYoVAU8olx8YnQhIvZJoESA4FXJ/xLoAhwSiW00DqmhZnAt/uwGnHYvaZEYHj3VHjI1DhoPDMC3PYPkmgRpu38S/gM8plzUGl2IiF0SKAbTijkF+AF9bJdi9dI6qopDw6tRGU0k23ykhKK+SObWqS9PpmZHOqZdaWTKFUy3GoDFysVHRhciYpMEikGCO92vBubQyyuHBDdNYyupGlWFJaWFbC2OW9Mr8DXaqNqbQsumLJKqnKQbXVMEWw78RVaCiVCTQDGAVkwa8GNg3Ml+7IAWasdVUDuiGofTTVY0D2P1pxYz1dszaCgZSHqjTVrTdGIn8Fvl4oDRhYjYIYESZloxRcCt0PMhKbMPz/gKyovKsDvdZPZbcTFIgWqwUbEpC/fnOWTJ6rGjtBJYXrza6EJEbJBACSOtmIsJ7HjXe/L+mY1Un7qPpvw6skxKngj7SoG3KpHytXnoO9PIlX0wR7xGYAhMmk2KPpFACYPgWSU3AtO7e1+zD3dhOeWF5TjkaqT/tJmo/TyHunUDyZHJfAA2AfcpF9VGFyKilwRKP9OKGQTcTjedge0emqftoXJkFTlyNRI+fnDvTqPswyGkNthJNroeg9UBv5Yd9qK3JFD6kVbMmQSuTLoMiMQ2GqfvoXpYNbn9uTNdnJgCX4WTgx8MxVGZFNcnXrqBB5SLD4wuREQfCZR+ohUzF7iWLlZhJbXScNYuDg+qZaAOlvBWJ06kMpEDK0eSEMcbJxXwuHLxmtGFiOgigdIPtGKuAL7e2X1JrdTP2EFdXj0D43nfSKRT4D8wgP3FI0iN42XHrwBPygFeoqckUEJIK0Yn0CH4uMaOZh/us3dxcNQh8uSKJHr4wbMrnYOrh5HVZsFhdD0GWA38XlaAiZ6QQAkRrRgrgf0lpx91h0IVlbN/6l5SLX6chhQn+syn0bY5i7IPhpLn0+PuBcEGApP1TUYXIiKbBEoIBJs73sExR/Pm1lE5azs43WQZU5kINbeJupUjaN6VTq7RtYTZHuAXykWV0YWIyCWB0kfBNiq/BIa235bgpum8rRzOaSBfWqPEpvIk9r01mvQWKwlG1xJG1cBPlIsyowsRkUkCpQ+0YnKAu4Hs9tsKy9g/bQ/pJhWX4+1xxafR+t9BVK4byKA42nVfBdymXFQaXYiIPBIovaQVkwHcB4HhLIeb5gs3cziziXxjKxPhVm+jYsUYrIcT42aZcTmBKxUZ/hJHkUDpBa2YAQTCJA9gTCX7z95JmlnF1fCH6ECBrySXfR8NYUicXK0cIBAqNUYXIiKHBMpJCk7A3wMMt3lomb2Z6pxGuSoRAYcdHHytgJQ4mVvZRyBU6owuREQGCZSToBVjA34FjBtcQ/l5W3HKUmBxLI9O4zujaNiTFhcrwXYDP1UuGowuRBhPAqWHgics/gyYNH03u4vKGKz1sA29iD8K1NZM9hSPYJDSYr4jwjbgTtmnIiRQeiC4A/42s49T535OZZZMvIsearBSsWw8jjjoZLyZQKi0Gl2IMI4ESje0YjTgB6nNTP7K52D3xs1KHhEiPo2WFWOo25dKjtG19LOPgHuk91f8kiGb7l01ppIxX1tPgoSJ6A2TwnHhZjKLDrLX6Fr62RnAN40uQhhHrlBOQF/JLNcOvjr6EENkx7sIhS2Z7Fk5gsExvrT4fuVildFFiPCTQOlC4r8Ye/4WFuU0ftFSRYhQqHCyf9l4smO4yaSbwHLirUYXIsJLAqUTpzxLelE5P3e6GWZ0LSI2NVqpfKkIZwzvV6kBbpbd9PFF5lCO8cszy5LO2KPukDAR/cnpJuuytXgHtFBrdC39JBW4I7h3S8QJCZQOlmilGTVTU3+sadpIo2sRsc/qJ/lr6zGnNsds+5IRwE3BlZIiDkigBC3RSnPLZufc1JZln2B0LSJ+WPw4L96AJa0pZkPlTOBSo4sQ4SGBAizRSgc3jkj8UfVZ6eONrkXEn2Co2NKbOGx0Lf3kMq2YiUYXIfpf3AfKEq10qDfR9LO93xycg6bFeosMEaHMioR5G7BnNFJtdC39QANuDnbpFjEsrgNliVaaAfxi91VDPX6bKcnoekR8MysS5pWSkNUQkyuj0oAbZT4ltsVtoCzRSp3AXdVnpLW05jmkN5eICCaF4yufkxijcyqnAnONLkL0n7gMlCVaqQW4w5NsTi2/MCfd6HqE6MikcHy1FFOCOya7916pFTPE6CJE/4i7QFmilWrALcD43VcNrVNmXc5+FxEnuKS4xezDbXQtIWYhMJ9iNroQEXpxFyjAd4Azq85M39uWY88zuhghupLgIePiDVRpCp/RtYTYcGCB0UWI0IurQFmilV4MzHEPsDSWX5CTZXQ9QnQnrYWBX97EAaPr6AfztWLGGl2ECK24CZQlWukM4CqAPVcPqcek2Y2tSIieya9j8Nk72W10HSGmE9hFH6sNMuNSXATKEq10InAjoB2akbG3Lcs+0OCShDgp4ysYOr6cfUbXEWIDgXlGFyFCJ+YDZYlWOhD4KWB2p1gaKs7NjvVT80SMOnMX2TG4nPgbWjGZRhchQiOmA2WJVmoGfgiBFuF7rhzSgEmzGluVEL2jg3XuRoixlV824FqjixChEdOBQuA40lEAdYXJZW3ZMtQlopvdS+qXN1FhdB0hNl0r5hSjixB9F7OBEpw3uRhAafgPfmWgnMsgYkJuA4Mm72eP0XWE2PWyNyX6xWSgLNFKk4CbCZ4DX3VOxn5fojnN2KqECJ3T9pEXYz2/8oCvGl2E6JuYDBTgBwSa0eGz6u7KmVnSXkXEFA3MX96E2eql1ehaQugSrZgMo4sQvRdzgbJEK70QOL3932UX5R5UFj3RwJKE6Bc2HykXbKHS6DpCyA5cY3QRovdiKlCWaKWD6PAD6Uk2N9ZOTpH2KiJmDaxn8MhDMbWT/iytmHFGFyF6J2YCJdhB+FbgyLLg/V/Pr0HXZCeuiGkzdjLAFltDX9LnK0rFTKAAVwND2//RnOeobhqeKOeciJhn8eO8YDOHjK4jhE6RPl/RKSYCZYlWOga4qONt+7+R70bT5HQ4ERdyGxg09DBlRtcRQnKVEoViIlAItKQ/Eh71Y5PK3Jm2XAPrESLsZm4nwezDY3QdITJJK2aM0UWIkxP1gRLsInzUD17FBdkGVSOEcaw+Brh2xNQE/WVGFyBOTlQHyhKt1EawJX27loH26rYsu1ydiLg0oppBKc3UGl1HiEyRq5ToEtWBQqC1ylEbocovzInFc7iF6BENTLO2x9RZ9HKVEkWiNlCWaKUZwPyOt3mSzY1NwxJl34mIa5lN5OXVxkwDySlacaDBq4h8URsowJUEWl8fUfE/2VVomsmgeoSIGF/aQSytcLzE6AJEz0RloASXCc/oeJvPqrvrigbI4VlCAE43WeMq2G90HSEyVQ7hig5RFyhLtFKNY5YJA1TNyChTZl3OiRciaNoeknQ/PqPrCAENuMDoIkT3oi5QCFyZHLXyQ4GqnpaeYkw5QkQmq48Bp+2LmauU87RiZDg7wkVVoASP9L3i2NtrpqQe9NtNAwwoSYiINqGMrBg5MjgVOMPoIsSJRVWgAOfA8WOph2ZlSgNIITphUjgmHYiZliwy7BXhoiZQgnMnFx97e2u2rcaTYs0yoCQhokJRGamaiom5lIlaMbJpOYJFTaAApwJDjr2x6syMOgNqESJqWP0kjy+PiasUDZhtdBGia9EUKMddnQDUj09OCXMdQkSdyQdwGF1DiMzSipEh7ggVFYES3HdSeOztTUMTDvkdppTwVyREdEnwkD6smoNG1xECycCZRhchOhcVgQLM7ezGqjMzmsNdiBDR6oy9UfP73p1ZRhcgOhfxP2BLtNJUYPqxtysNf+NoZ0YnHyKE6MSAVnKyG2LiZMcJWjHJRhchjhfxgQL8D2A+9saGsUkVyqInGlCPEFFryn5i4apeB6YZXYQ4XkQHyhKt1EQXa8+rz0yPlZPphAibvDqyTf6YONXxLKMLEMeL6EAh8Cok/dgb/WbN0zQkUY5lFOIkmRT2sZUx0dq+SIa9Ik+kB8p5nd1YOzGlEpNm6+w+IcSJFZXFRE8sE3C60UWIo0VsoCzRSp3AhM7uqzk9VYW5HCFixoBWcpxtNBhdRwhMNboAcbSIDRQCPyzHTcYrUK05dlndJUQvaaCdcoDDRtcRAqdoxViNLkJ8IZID5bilwgAtgx1Vcu6JEH0zqopY6M5tByYaXYT4QkQGyhKt1A5M6uy+usIBsbDsUQhD2XykDKyj0ug6QkCGvSJIRAYKgUaQnV7KNo52Sh8fIUJgbCUtRtcQAp2+8BTGiNRA6XS4S5nwtWXY5GxpIUJgUC1JRtcQAtlaMWlGFyECIi5QlmilFuC0zu5rHO6sQtfkCkWIEHB4SUtqpd7oOkJgvNEFiICICxQCl7CdTrrXFybHwiW6EBFjzCFqjK4hBAqMLkAERGKgdDrcBdA40hkrZzoIERGGV8fEslsJlAhx3D4PIwV7d3W6asNn1d2eFIvsPxEihFJbyDT7cHtNUR0sw7RiEpTL2MaXmkYCkEJgQVH70Hxb8K1aKbwGlRY2ERUoBMZCO50obBztrELTBoa5HiFimgbm4Yc5sDWTPKNr6QMNGAd8GrYHDITHCGBkh7fcYC2d8Wgau4HtHd72xlrIRFqgjOvqjvqCZHc4CxEiXoyswrs1+tdOFtDPgaJp6MAZwJeBIroOj85YgFHBt3ZtmsZq4HWl2BGyQg0UaYEyuqs7WgfapRmkEP0gowmn0TWEQL/No2gaacD5BI7SOK77eR/YCDTAPU/T2Aq8AbynFFH74jnSAmVUV3e4U6yxsGZeiIjj8JAaA/Moo7VizMoVuiEkTSMJ+Dbgov+fK0cH367RNJ4HlilF1DXBjZhVXku00kwgtbP7fHa9VVn1WHgVJUTE0UAfWB/1zSKtwJBQfTJN43RgCXAu4X3hnQRcC9yraUTdnHHEBAonGO5qyXPEwuYrISLW4BrajK4hBPL7+gk0jSRN4xbgDrp4gRsmBcBDmsZXNO2k5moMFRWB0jw4oTWchQgRb3IbIm74uzf6FCiaxikErkpcoSgmBKx8cbUSyrmbfhMVgdKSL/sZhehPA1oMfTUeKr0OFE3jLODnGHtV0pUC4LeaRq7RhXQnIgJliVaqEVjH3am2LFs0TxYKEfHMioQBLdQaXUcf9SpQNI1zgVuJvEVKHWUB92kag40u5EQiIlCAwXTRvwvAM8CSHMZahIhLeXVRfyzwQK345OYbNI1zgO9zcntKjJIK/ErTyDG6kK5ESqB0OdzlcZqblVlPCGcxQsSjjCb8RtfQR1agx1s0NY0pwE1ER5i0SwPuDu6NiTgRHygt+bLCS4hwSG2JmOeDvhjUk3fSNFKBW4jsYa6uZAM3Gl1EZyLlB6jL9eMtgxyxsJxRiIiX1Nb1sHMU6ek8yiK66BsYJSZpGucbXcSxIiVQulwS15Zui7rdokJEI4cnqp9g23Xb5FLTcAGn938p/e4aTSOiOrAbHijBFV5djgd6k8ymMJYjRNwyKex2j7Et4EPghLvLg0Nd/xumWvpbAoEFBRHD8EABkjnBOKYvwSSBIkSYpDXTaHQNfTSgm/tvILqHuo41SdM4z+gi2kVCoJxwB6jPYZI9KEKESXoT0d6VosstBsE9HNPCWEu4fD1S2rNEfKD4bbq0rRciTNKbo37p8ImuPr4ctirCKxeYbHQREA2BYtGl74oQYZLgiYjnhL6waMXHr1bTNOzAlwyoJ1wuNLoAiIxA6XJC3mfX29C1aFwnLkRUsnkj4jmhrzob9poJxPKL09M0jSyji4iEH54ur1C8SeZoH88VIqrYvFG50e9YnU3Mzw57FeGlEThR0lARHSieZIsEihBhZPXFRKAcNY8SnIwfakwpYTXD6AIiIVC6HPLyDLCE7DhPIUT3LD5iYRHMsUNeXR4tHmOygscWGyYSAuVEVygSKEKEkckfE+1Xjg2ULo/GiEGGfq2RECiJXd3hSzRJ2xUhwkgHi+4n2l/IHfsqXQIlTCIhULrcCa/0iNirI0RccXiifnPjkWE7TUMHhhtYS7gZGiiGTcBpmvYX4KI0Btrv4q1VAKt5LvcdnhxzmDLnd3nsvSTTfEmUCOb3eXjuwYJzHIlZrfOuf/+T1/587uSGmt1OAI+70WKxOj0Lfrh9tc/bqr317PyJNZWbBii/Txs2ft7+My96cLvR9YvOmf34+vQJ9m6286sFk6ivtqFpcO6CPVx7zy6qyyzcPncK1QcdpA9s4Z7XPiUtx0PVAQs/mXMqezelMH3uPn7+fGkfv4SOL5QHETgnpR816nDadHDr4NPhgoPw6NYv7v/RcLi/APb9C/Ld/VtLnAYK8Nck0h4F3mm/YTAFDdfwwH//zi8mAChdkyGvCPbJ23cMT04d3uBxN1oA5lzzzmft9737wrcKrPZkD8DnHz860Od165f/aOcqd2ud6R+/G+Mad9q1B9KyC1uMql10TVP07ffObFHccP9GJs+so77axDUTz2H63EMse3QQRWdVsej321ly00ge/8lIbntyE/ZEP1f9Ygvb1yaxe2MoJpU7BkqPD9zqvQQ/fPAhpPqgVYNJZ8KblTC7FjbbYVUmZIXrZz0MX2/XDBvyUkqtTiG7ruNtQ5nQOITCpiM36JHRn0Ycr/bQFvv+7e9kjTvt2r3H3qeUn31bVwwcO+WagwCapimfp9nk87k1j7tR13Wz3+ZIj/Zx+pil0cdAGTiijckzA7/byek+coc1Ur7Lzmfv5jDvu/sAmPfdfXz6TuAoW2eKjzPnHsbq6NuV0Rc6DqOHoRegTiBMANp08GpfHAJ5w3j4v01hPBRS0zQs4XqwYxm65txBsrmJui7vlyuUyLX61evHn3HBbze5W2uP+xnavXFZmi0hrS0zb3ITQMHU68p2b1qe89Tdmef5fG2mya7bP09MzvWEv2rRE3pfr1A62lXqYN+WZKacV0tDjY2BIwIH5g0c0UZjbX892Xd8oRymZdAeYNw5cCARLt0NF9TCn7MhpxVmhPvUWWuwoLCLhEl5EWU2f/pklj0hw50/clanrwa2rXsmb9i4rxxo//e+bW+naLqurry94u1Lb97y7sZP/jSiumx9QvgqFicjZK+lG2pM/OziU/n2rz5nQEaMX5FagO2rYffbUJIC7yXB/aNg8RajKwsnQwOlhfoT/pBpfiVDXhGofPd/0g7uLM5+6p6cWate+c7kQwc+TX/9yQsmAfh8bm3/jndzx54aGO4C2Lr26bxBI8+vNJntKilliDtj4KTDZbtXd3duhTCILxSt0N2tGrf+z6mcNe8Ac64rByAptY2DOwJXDAd32HCm9NcEdceOyWE+QjzbC2dUw/M5UJYARTMgZxYcssPks2FXOK6Y+nviv0sGB0rjiVtl+0N46S1CxjX/z5uvuuPQO1f+tPzdGfMe/ywzb0r1l69esRZgx/oXMpJShjSmZI45svTUOSC/5eCu4gyl/LS11Jqqy0tS03MnRvtBTjFL9fUixe+Hn108kbyRjVx3384jt0+aWc4rfxwEwCt/HMTkWeV9epyudZyLCcOT6z4rVASHfht0+E8GTKmD2n9B+buBt8xW+Ow9GNbfAaeUMma4C4xdNvwPwKWjO2/j7HNncsUWJymeZSwubKHR+mduOd35zh+Gzj+1tMSoGsXJ277+ubxh4+cd6HjblJl37n7r2fmnPPt/w12gGFH49X25Q89uMKhE0Q3V1yuUD5en8fGb+eQOa+DykecAcMXPNvOde7bz07lT+HreIFJzWvjN8k+PfMzFObNobTLj8+p8NTOHe177iIIzevuio+ML1UO9/0J6arcNvj0JfMH1DBcehKsr+/9xO1Vl0OMCoCll7EXAEq30ta7uOzgnd/fhaelDw1iOEHHv6ck0NtlwGl1HH7yiXPwFjmxsfIGwTc4b7gOl+I1RDx4Jk/JdLhXU/DLiJUS4tViivp/XkWElpfADuwysJdx2GPngkRAoTV3dYWryyaS8EGHkB49fj/oW9scOp8ZTVwZDv9ZICJTqru6w1Hui/QdbiKji06O+jxfAsfs+4ilQthn54JEQKIe7usNSJ4EiRDh5TMYtOQ2heL1CqVTquK89rCIhUE50hRLtY7lCRBWPybglpyF01BWKUuwB9hhUSzitNrqAiA4Uc4NXAkWIMGo1R/1ZKECn/ZzeCHsV4aWAN40uIhICpcshL1Or34ZfxcIPuBBRoc3MiTcbR4fOhn1WArHc3XqNUhi19+WISAiULq9QAHSPP5Z/CISIKM2WqA8Uj3IdHxxK0UIgVGLV60YXANEQKG5/mHvxCBG/Did0fYJqlDjRpHSsDnuVA591+15hEPGBYmrxxcIkoRBRoSox6neUd9kqPjg5/1EYawmXF1Qojxzog0gIlHroeiLQ1OSTORQhwuRwQlS3XIHOJ+Q7egSIpcaka5XibaOLaGd4oCxShYoTTMybG7yhOsVNCHECPo3WVgvRfk5N2YnuVIrDwJ/CVEt/awYWG11ER4YHSlCXw17Ww25pvyJEGLRYjN0UFyL7u3sHpVgJfBKGWvrbX5QytrvwsSIlULrcdJSwrznax3SFiAoNtphou9JtoAQtIbqHvtYqxVtGF3GsSAmUrV3d4djfkhzOQoSIVzWOrjt/R5EeBUpw6OsBTjB/G8EqgQeNLqIzkRIoXTY0szR4EzSvvzmcxQgRj6oSo37JsBt6vrlPKdYAv4fIWCHVQzXAHcFAjDiREih7oOvLbXOdNxbGdoWIaAcGkGR0DX10ULlOLhyUYjXwENERKrXAnUqdeOGBkSIiUIIrvbrsCGqvbJXNjUL0I69Gc52DFKPr6KOezp8cJbjs9v+I7OGvSuC24F6aiBURgRLU5bCXY790XxGiP9U5qDW6hhDoVaAAKMV7wF0Qkd+HTQTC5KDRhXQnkgKly4n5hL3N0nVYiH5UlhQT56D0OlAAlGItsBBYFZpy+swN/JlAmETU8uCuRNIBVl2v9DogK72E6E97U6P+HHnoY6AABA+oul/T+A+wCAwbBtwE/EEpDhj0+L0SMVcoi1RhJV1cbppa/XbN7Y/mNeNCRCwF/oPJpBldRx+5CeEhWkrxEYGrlXchrMupG/jiqiSqwgQiKFCCupxHsda6ZaWXEP2gxUyt14TV6Dr6aKtyhXZSXSkalOJB4NvA3zlBi6gQ2EZgb8lVSrE0Upo9nqxIGvKCwLDXaZ3dYT/Y2taWFQtX5UJElionDRD1Vygb++sTB/d8/EPTeAE4A7gQKAL62hbKTeDY3jeU6vrFdDSJtEDp8ocieWO9te6UlDCWIkR82J4Rcc8DvdFvgdJOKXzA+8D7mkYiMAIY2eEth65DxgvsJrA9ov1tj1IRvVT5pEXaD9LnBPrrHNdC27m1MQOlfGhatO/mFSJiKPDuTCPT6Dr6SBGYxA7fAyqagPXBNwCCIZMCWAAb4CdwFeIGDsVaeHQmouZQFqlCH/BxZ/eZ3H6rpdYTFUvnhIgWNQ4OxcD8yW7lwvD2TErRpBQHlGK3UmxRim1KsUcpyuIhTCDCAiXow67ucO5ojIVuqEJEjF1pxMKJqJ8bXYAIiMRAWUsXfb2SN9RLK3shQmhzVtS3W4EwzJ+Inom4QFmkCt3Ams7uc+5szMSvYuEVlRCGazFzuMFOLGwalkCJEBEXKEHvd3aj5sNkrXbLPIoQIbAvhXqjawiBSuXq+sRXEV6RGihroPPeQklbG6TzsBAhsCULh9E1hMBnRhcgvhCRgbJIFbYSmEs5TnJp/XFLioUQJ6fNRO2BAWQbXUcIxMLZ8DEjIgMlqNPVXgl7mtM1r19WewnRB9szIrJN+8lqA0qMLkJ8IZID5WM6OfBGA81e3irzKEL0kgK1No90o+sIgXXKFRNt92NGxAbKIlXYCGzo7L7UT2rCXI0QsaPOTnmjLeqP+4UuNkEL40RsoAT9q7MbU9bVZuNTMjkvRC+U5oS1HXt/6bKrhjBOpAfKh3TSMlr3KkvinqYKA+oRIqr5NFo3ZcfEZPwG5YqJZc8xJaIDJdjb683O7kv/oNoS5nKEiHoHBlDh04mF351O96oJY0V0oAS9RSeT80mbGrI1j7/JgHqEiFqf5ZFgdA0h4OcEPf+EcSI+UBapwhrgg2Nv1xS6c1ujrPYSoofq7JSXJ0d9q3oIDHfVGV2EOF7EB0rQss5uzHi/KhZ2+goRFh8Pjs5jZTvxrtEFiM5FRaAsUoVb6KRFdeKu5iy9xVcb/oqEiC4tZqp3ppNrdB0h0AD8x+giROeiIlCCXursxuSN9XLpK0Q3Ps03/gCqEHlHuWLiDJeYFE2BsgbYe+yNGf+pioX220L0G7dO/ec5DDS6jhBQwAqjixBdi5pAWaQKFfDysbfbK9pSzbWeSgNKEiIqlOZSozRMRtcRAuuVi4NGFyG6FjWBErQKOG5lV9a/K6WfjxCd8Gm0fJYXE3Mn0MWeNBE5oipQFqlCL/C3Y29P/bRmoN7qk7kUIY6xIYdKrwmr0XWEQA3SaiXiRVWgBBUDWzreoCn09A+raw2pRogI5TZR98lg8o2uI0TeVq7jNziLyBJ1gRKcS3kcjl5Tn7GqKlfOSRHiCx8NpsGvx8TciSLQMUNEuKgLFDiyL2V1x9tMbr81ubS+3KCShIgojVYObcyJmauT/yoXsvAmCkRloAT9lcCJbUfkrCjPQKlYaM0tRJ8Uj8BvdA0h9LzRBYieidpAWaQKqzhmGbGl3utM3NV0wKCShIgIVQkc2J8SEy3qAT5VLrYaXYTomagNlKCXOGYZcc4b5YkG1SKE4RT43h0VEx2F2/3D6AJEz0V1oCxShW3AUx1vcxxsTbceaiszqCQhDLUjnX01CaQaXUeIfKZcR6/oFJEtqgMlaBXHLCPOWSFz8yL+uE3UFY8gz+g6QkiuTqJM1AdKZ8uIkzc15MpViog3K0fQ7DXFxGmMAGuVi81GFyFOTtQHChxZRry84235L+y3olSsnP8gxAmVJbFvV2y0p28nVydRKCYCJeivwO72fyQcaElP3NW0z7BqhAgTr07TijFkGF1HCK1TLjYZXYQ4eTETKItUoRv4P+BIo8i8Fw+k4VdydoKIaauGU9tmIZZOL/270QWI3omZQAFYpAr3An9u/7e1zuNMWVsr7a5FzCpLYt+2zJiaiP+PXJ1Er5gKFIBFqvANOnQlzX2tLFfz+JsMLEmIftFmovbNsWQaXUcItdLhBaGIPjEXKEF/AA5DoMdX5spD1QbXI0RIKfC+Pg6v24zd6FpC6HnlOv68IxE9YjJQFqnCBuB3BJcSZ646lG9q8h42tiohQue/gzhQmRRTE/EHgKVGFyH6JiYDBWCRKiwBXoHAeSkDXz0ore1FTChLYt9n+Qwxuo4Qe0zOO4l+MRsoQU8D2wAGlNYPtFW2ygS9iGqtZmpeHxczjR/bfaBcrDW6CNF3MR0owSOD7weaAYY8tceJT8n58yIq+cG9rABi5Ejfdm3AE0YXIUIjpgMFYJEqPAjcA3itNZ7krHcrpdGXiEofDKXicGLMNH5s94JyccjoIkRoxHygwJH5lD8AKqv40GDp8yWizaYsdpfmMsjoOkLsIMF5ThEb4iJQABapwmKCre6HPrnbiU/JJL2ICgeS2btqBEONriPEFPCgciGdLGJI3AQKwCJV+BLwmrXWk5TzVrlcZouIV+Pg4PKCmNoJ3+4l2REfe+IqUIIeBz7I+E/1IFt5qxwXLCJWs4Wql4rIUBomo2sJsV3As0YXIUJPO9kO759++mmW2Wx+AigkWgNJoXkbSFF+zex1mAA0o0sSvaPQVH2zteHttYNKmlptMbOCz61T//wp6E02nEbXEmIe4Gbl+qIzuIgd5pP+ALP5iZycnHGZmZk1uq5H7Xkjfq+qbtzryffZddw2cywtw4wrSvlpqK92nse+iUs/HPlfo+sJBZ9G69JCfE02ko2upR88JWESu3pzhVGYmZlZH81hAqCbNX9ivuWg3uzTzcofM69s442m6SQlp3uTE9xJRtcSCj6NllfH0xiDy4MB1gDLjC5C9J/eBIoe7WHSzmTVvIn5loOmBq+moXxG1yN6R9N0NFTUD1t6NZqXFtIcYz262tUQWNUVE88donPROQcSQma77nbmmstsLV4vyJHBwhhejealRbQecpJudC39QAG/Uy7qjC5E9K+4DxQIhkqmqdzm8/d4b8rwwc4B/VlTT+zZvVM/e3pBTAz1xDOvTtPLRbRVJZJmdC395HnlYp3RRYj+J4ESZLLpnqRUvcKqqxajaxHxw6PT+FIR7hidMwF4HznSN25IoHRgsmjeAcmUmU2qrTcfv3PHNv0b889LnOWa5Lxo9nTnli2f6+23X3DeVOd5M6c477n7p/aOVzeLH/yN7byZU5wzzipM+s2vb7dD4MrjzNPHJP3ge1c7zpo2Lulr82YltrQ0A7D2s09MrrOLkmafN9X55ycektVpUcytU//PCXhrEmI2THYCv5d5k/ghgXIM3aT5k5O1Mt3MSYfKLTd9x3HPvX9sebd4bePPf3l/y20/XOgAuP0n33dc853vtb39708bc3IG+tvf/9133jDv3LlN/9e7/21cubqkYX3JZ6b3/7PSBLB79w79mmu/6/7Ph5sakpNT1LKlL1gAbvz+txPuvucPzW++/UljqL5mEX6NVir/MQlTnYMUo2vpJzXAr5Tr5H+PRPQ66X0o8cCi4010cqihiUw82HryMY2NDXz26cfm73z764ntt7ndgd+ltZ99bHr678s8AF/7xrfcd9/1YwdA8cp/md9b9a5l5jmnmAGamhvZsWOrKT9/iDd/0BD/hIlTfAATJk727d27W6+rq6W+vk476+yZPoBvXHKFe+W/37KE9IsX/a7Cyf5l48n26cTq/50b+LUc5xt/JFC6kGCi1e2gtkUjRXd3Hyp+v5/k5AFq5XvrG3r6GEopvvv9W1uvvmbRUftg9uzeqVutXzykyWSitbVFU0ppmmzqj2pbMtm9cgRDiO3/yIeUiy1GFyHCT4a8TiDFQoNuo9ltw4124nHg5OQB5A8a4n/l5X9YILCDe8P6z3SAUyZP9b0aHLJ66Z/PHpn3+NLM//E+94+/WhsbAxl08MA+rbKyvMsnmpSUVJWUnKw+eH+VCeCfLzwjcyhRQoHvgyHsXTmSoTEeJv9ULoqNLkIYQ65QupFlobpMYWrS0BPdgD/wPWtpaWZiYd6R1hjfue4HbY889vfmW394veMPv7/H7vV4mPvVb7iLJkxuu/vXf2hZdMM3Ex575Pe2WefO9iQlJSuAWede6N26ZZP7wv85IwkgMSFRLXn02WaTqetegA8u/kvzTT+4JsFhd6gZXzpfzuCOAj6NlhVjqNuXymCja+lnHwN/M7oIYZyTbg5ZUlKye+LEiXE1NqoUHHCT0+IjMdmDB+/JHcHa3NyEw+FA03T++eIzlqUvP2d95h/Lm/qr3ni0a8fn1qfeHf+u0XUcq8FK5bLx2BvsMdmXq6MtwB3KhZwzFMfkCqUHNA0GWinf38bAWgsJSTpukxsLPexSvPaz/5pu/8n3EpRSJCcPUA8ufrK5n0sWBlOgtmWwZ+VIBsVg+/ljbQd+LmEiJFB6SNcg30bZvjbyGsDusOOxtWFCdT8PdeZZLl/xext6PFkvoptXp+mdUdTvTou5UxY7sxv4mXIhV9xCJuVPhq6h8m0ctGq0tWhYGuyAGelULI447ODgs5PQdqeRa3QtYbCPwDCXvFgSgATKSTNp+AfZOWjRcPtAr7VgDa4C83f/0SJWKfCV5LLnhYnktlhJMLqeMDhIIEyk4aM4Qoa8esGs4Rtk4+DeNvK9CnOzjrXNhj/Ji/tkJ+xF9GuwUfHmGKyHExlidC1hUgHcrlwcNroQEVkkUHrJouPNt3FgfxsDvQqLTwtcrSSacFvcmHsytyKim0+jdU0+lWvzGBTje0s6qiIQJnG10lP0TJ8DZenMXSNDUUi7r/572PYT3V9eXm5yuVxjAKqqqiy6rqu0tDQvwLp16zbZ7fZ+aUR38803D3Q6nb677rqrov224YPyxn70yZotelpu9lfPOT39hVUfN27es9O+9v33tCsuubKlv65W7r3nDntiolN97wc/lj5JBil3su+tMaS1WGN+b0lHhwmESUW37yniUtRdoeTk5Pg2b968ETp/kvd4PFgs4WuRZNbxD7RzYNl7H/tb/Tj2796lv/bP52xzv3mlz2bG4/Cg4Yu+77PonNtEXfEImnamM8joWsJsL4GlwXJlIroUE8My8+fPH3rttdfmn3766aMXLlyYv3LlyoRJkyaNHTduXMGkSZPGlpSU2AAWL16cfv755484++yzRw0ZMqTw+uuvzwfwer3Mnz9/6KhRo8aPHj264Je//GXWyTy+ScM/Li0hK0Gn6YGf/cT+2UcfmOdOnZD08IP/5/x463rrubOmJLrOLnLOOKswafu2Lcd9z1984WnLOWeOTzp7ekHSL+68xd5++/DBzgG/+uVtdtfZRUkXnDfVWVFRdtSwys4d2/RZM05xtv97+7Yt+izXJCci5HwabZ9ns/tvU0jYmc5Ao+sJs1LgVgkT0Z2YeeW8Y8cO+/vvv7/VbDZz+PBh/ZNPPtlssVhYunRp0q233pr/1ltv7QDYuHFjQklJyUaHw+EfOXJk4Q9/+MOKsrIyS1lZmWXbtm2fA1RVVXW6Ee3RRx/NfuGFF44c0VpZWXnUpVC+jbKf3nW3+ZE//D7t8aVvNgH84gc3OC5f9H33/MuvdFub29yqzWelw4bIsoP7tXvu/qnjX++uaUhNTVNfm3du4vJl/7RcNPdrnubmJk499QzvnT+/r/Xnd95s/9tfH7P+6LZfHBnmGj5ilD8pOVmtL/nUNGHiFN/fn/2z9RuXXCHLmENIgXdnGgfeG05mqyUu9pUc6z0CZ5p4jC5ERL6YCZSLL764xmwOfDmHDx82XXLJJcN2795t1zRNeTyeI0/gZ511Vn16eroPYOTIka07duywTZ48uWXfvn22K6+8ctCcOXPq5s2bV9/ZY1x//fUVHYfX8vLyijrer2mQpNy1ms975FjeU06f5v3T//3GXnFgv/4/X/2aafSoMR7lw6d5AjvtP/v0Y9MZ0872ZmZmK4CLv7bA8+GHq00Xzf2ax2K1csGFX/ECTJg4xbe6+O3j/r8WXH6N++/P/sU6vnBiy2vLXrSskHNSQkKB/8AA9hePILXRFjert471CvCkHJAleiomhrwAnE7nkX0gt912W96MGTMatm3b9vlrr7223e12H/k6rVbrkV8Ok8mkPB6PlpmZ6SstLd34pS99qeHhhx/OuvTSS4f2pRbd73VnWALDA19dcIXn4ReXNdnsDnXtVy5ILP7327Y6M9ZGO35lwa1O8LtqMVvQNL29Vrxe73ErieZ+9Rue4pVvmVe88aq5qGiyLz09U375++hQIvtfmEjd8gIGN9pI6v4jYo4CHlcu/iJhIk5GzARKR/X19ab8/Hw3wGOPPZbR3fuXlZWZfT4fV111Ve3dd999YMOGDb3emDZgwABfY2OjKd1CbY6V8r07tjFk+Aj/1d+/2e2a/WXvlg0lJgCvhqnOjHXsmdPcH334nqmquhKfz8vSl5+zTJ8+o8ddhO12BzNmnOf9yW3fTbh0wVUy3NVLCvzlTva9XEjVSxPIj+FjebvjBu5VLpYZXYiIPn0e8upuma8RbrvttvJrr7122OLFi3POPvvsToevOtq9e7flmmuuGer3+zWAu+66a39vH3vq1KktZrNZjRkzpmDBggVVza1tLf98+eVMs8WqpWdl+797+y+OaqCXkZen/+CXv2796le+5MTvZ+bMCzxfnjP/pNrSz//6N90r3nzVMuvc2dLO/iT5wb0nlbIPhpLaYI+7lVvHqgPuUS42Gl2IiE7Svj4MfAqtzE12k49uV2DpoBL8eMxe9J4uN37oD/fa6uvrtNvv/E3cdns92fb1bSZqN2ZTtzaPbLcZe/cfEfO2ELgykd9t0WsxMykfyUyBppLlVR5Sqj2kc4K2937QGnWsWMGm8Np9+DVv1zvvr7h8bsKePTtNLy39t0zGd0OBrzqB8nV5sD2dgWikGF1ThFgO/Fm5kCtc0ScSKGGUYaHWrtNW7ibHp7o/I6NNw9xmBs2MSvDjtnRy1fK3Z5fJ2SonoEA12KjYnIX78xyy2szkGV1TBGkF/qhcrDK6EBEbJFDCzGmiZYiNvQfc5Lb5ezbUokBrCl61WBQ+ux+vyYeOj/C1BIgyLWYOb8+gfn0u6Q12coyuJwLtAu5TLg4YXYiIHRIoBrDo+Abb2F/pIaPOe3LDLh4Nk8eECROYFH6Hwmv2oeHFTA9PkIxJGjRYqdibSsvmLJIOOUkH0owuK0K9TmCISzYripCSQDGIrkGOlSqniaYKN9ledfL/Fz4NvVHDig66Bb/Dj8fiB3yY8Mf8sbOg41MmfB4dWs1Ynp1CttElRbhGYLFy8aHRhYjYJIFiMKeJFoedvRVuMht8vd9E5we9SQ+EC2YwK3w2hc/sB92Hrvwx8H+t40PH59OhVUf3aJghEJwqftrH99b7wGPKRY3RhYjY1ecnmZnrCGn7+n+fQrf7Wvbu3WteuHDh4JKSkgSr1ary8/PbHnrooX3Jycn+66+/ftCKFSt29uSx8vLyitasWbMpNzfXC7B8+fKkBx54IHvlypXbn3322QGff/6545577il/+umnUwoKClqnTJnSL8tyTRr+04bnZRV//OkeLT0nvScT9t3xapi8Gqb2gNHBb/fjNSvQ/YA/wq9iOoSHR0dza5j8gfCI3JojUzXwiHLxsdGFiNgXda9a/X4/c+fOHblgwYLq5cuX7wT44IMPHAcPHrRMmDChsadh0p3LL7+8jsBGL5YuXZri9Xrr+itQ2jlNqjnTTmOlm4wGH8mh/Nx+0Jv14PkswadkDZRV4TMr/Kb2oFFoKPTgW3++6lfo+NHwKx3l11B+Dc2roXkkPEJBASuAvyoXshJQhEXUBcry5cuTzGazuvXWWw+13zZ9+vQWgC1btlgvuuiiUdu2bft8zZo19quvvnqYx+PR/H4/L7300o6ioqIeH0i1ePHi9DVr1iR+61vfqn7nnXdSPvroo6T77rsv96WXXtrxyiuvDHjyySczTSaTGj16dGt7sLVrbm7WrrjiiiHr169PMJlM/Pa3v903Z86chsWLF6cvX748paWlRd+7d69t9uzZtY8++uhRu/JvufEHORkZGd6bfnrngQo3WffefltSela2/+rv3xzytioKtDYNc1t7bBzz9K2DMin8JvDrCqUDeqD5mAYoLbAnttPQUYFblR80pQWe3ZSG5gfNB5pXOxIYEhqhtx94SHa8i3CLukBZv369Y+LEid2+4nrooYcyFy5cWHHDDTccbm1t1bzezvdszZgxY7SuB/YMNjc36yNGjDjqKuS8885rOvfcc2svuuiiuquvvroG4Etf+lLOnj17NjgcDtVZq/v77rsvC2Dr1q0b165da7/wwgtH7dixoxQ6b58/cuTII6ttFi5cWDVv3rwRd955Z6XN4tu74qXnJrz43hpDdsD7QfNrmDxgkhmKqOAFXgKelxVcwggx2RwSYNq0aU0PPPBA7u23356zbds2q9Pp7LTHzKpVq7Zu3rx54+bNmzc+/PDDe3ryuceMGdMyb968YQ8//HCaxWI57vN+8MEHziuuuKIaYNKkSa0DBw50b9iwwQ5ftM9PSEhQ7e3zj/nc7pSUFO/777/veHXpK0njx41rmJyfudtpouHkvwsijqwHblQunpEwEUaJukApKipqKSkp6bYb8PXXX3/41Vdf3e5wOPyzZ88evWzZspC1IV+5cuW2RYsWHfr0008TJ06cWODxHP37e6L+aJ21zz/2fa6++uqqJ554IuPJJ5/MuPrqq6utOt48GxWDbOyz67SE6usQMWEngaN5b1cuevSCSIj+EnWBMmfOnAa326098MADR9rSr1q1KuH1118/qvHixo0brePGjWu74447Ks8///zadevWOXr7mE6n01dfX68D+Hw+duzYYZ0zZ07Dww8/vL+hocFUV1d31LDXWWed1fjMM8+kAaxfv95WVlZmnTBhQo+Hrb71rW/Vrly5ckBJSUni/Pnz69pvTzDRNsTOgVwrZRZNXoXGuQrgfgJXJZ8ZXYwQEII5lJ4s8w0lXddZtmzZjoULFw568MEHc2w225Flwx3f7+mnn0578cUX081ms8rMzPT85je/Odjbx7z88ssP33DDDUMfffTR7Oeee27Ht7/97aENDQ0mpZR23XXXVWRkZPg6vv+tt95a+a1vfWvI6NGjC0wmE4899thuh8PR47bOdrtdTZ8+vT4lJcXXfgplR8lmmpwmmmq8DDjsIc0vE9vxpA54HnhTmjmKSCPt6yOQz+dj/PjxBS+++GK3K9N8Cr3KQ2q9lwH+KLziDJWDWz+3zqztefv6KNRK4EjeV5RLhj1FZIq6VV6x7tNPP7V/5StfGTV79uyanixzNmn4s61UZ1ioqfGSXOslxdeLNi4iYjUB/wJeUi7quntnIYwkTzwRZsqUKa379+/fcLIfZ9LwZ1ioTTNTW+8lqcZLqlsFNzKKaFQGvAa8rVzE7cFpIrpIoMQYXYMUCw0DzDQ0+kg47CW11U+vFySIsCsFlgKfKBcnNx4thMEkUGKUpkGSmeYkM83NPmyHvaQ2+UgknlvcRy4v8B7wqnKxw+hihOgtCZQ4kGCiLcFEudePqc5HUr2XJLfC1v1Hin5WDbwLvK5cHDa6GCH6SgIljph1fOk6tekWalt8WOt9JDf4SApFd2PRY63Ah8C/gRIZ1hKxpO/t62eGuH39v7vf16Jp2pRrr7224vHHH98P8LOf/Sy7sbHR9Lvf/a7Xe03aJSQkTGpubl7b/u/2JpF/+9vf9v72t7/NTEhI8H/3u9+tXrx4cfrcuXPrhw4dGpUbDB0m3A4TVVmKqkYfCXU+kpt9JCoZEusPXqAEWA18IJPsIlZF5RWK1WpVb7zxRmpZWVl5+1km4dCxw/EzzzyTccopp7REa6C06zjX4lPo9V6cjT4SW/wkSLj0iY9Af633gI+US3qxidgXlYFiMpnUFVdcceiee+7Jfuihhw50vG/r1q3WK6+8cmh1dbU5PT3d+7e//W33qFGj3PPnzx+alJTkKykpSTx06JDlV7/61f727sE9dfPNNw90Op2+YcOGuUtLSxOuuOKK4Xa73b9mzZpNt95668C33norxWQyKZfLVf+nP/1pf/efMbKYNPypFupTLdT7FVqjj4QmHwlNfhJlb0uPVABrCVyNlEiIiHgTtU8SP/rRjyqLiorG/+IXvyjvePv1118/eMGCBdXf+973qh988MH0G264YdA777yzA6CiosKyZs2azevWrbPPmzdvZGeB0tbWpo8dO7ag/d91dXWm884776gNZVdffXXNI488knX//ffvO+ecc5orKipMb7zxRurOnTtLdV2ns5b20UbXUMlmmpLNNAGHWv1Ym3wkNPtIaPFjV3G8K7+DBgJXIeuAdcpF+YnfXYjYFrWBkpaW5v/6179efe+992Y5HA5/++1r165NfPPNN3cA3HDDDYd/+ctf5rffN3fu3FqTycSUKVNaq6urLZ19XpvN5t+8efORg4na51C6qcVns9n8l1566ZAvf/nLdZdccknM7Wi267jtOu50C7V+hdbix97ix97qx9bmx+ZVdPr9jDGNwA4CVyDrgO0yqS7EF6I2UAB+8pOfVEyePLng0ksv7VFvMbvdfuSX/2R7mJ2IxWJh3bp1m5YtW5b83HPPpT7yyCNZH3300daQPUCE0TVUoomWRNMXPaW8Cr3FFwwYha3Nj90b3cNkFQRaw+9q/1O5qDS2JCEiWzT/wpOdne2bM2dOzd///veMyy67rP1Aq6YnnngiddGiRYcfe+yxtFNPPbWxPx7b6XT62tvW19XV6Y2Njfoll1xS53K5GkePHl3UH48Zycwa/iQzzUl8cX65V2Fq8WFzK6wehdmrMHv8WLwKS4Q0smwjsBekGignEB67CIRHk5GFCRGN+t6+vgfLfPvT7bffXv7UU09ltv/7kUce2XvllVcO/cMf/pDTPinfH497xRVXVH3ve98b8qMf/cj/1ltvbbvoootGtrUFTme/++6793X38fHArOFLMtNMh5Bp51PoHoXZ7ccSDBuLV2HyK3QFuh80pQJ/tt/Wg1VnXgIh0RL8s5VAu/dqoOqYP6tl0lyI0JL29SJqKAV+0P0KTQPQUBooDdiwviR94sSJQ42tUIj4FtVDXiK+aBqYwG+S3TFCRKRIGMcWQggRAyRQhBBChIQEihBCiJCQQBFCCBESEihCCCFCos+rvH7/bGjb1990eff7Wvbu3WteuHDh4JKSkgSr1ary8/PbHnrooX3Jycn+66+/ftCKFSt2hrKmruTl5RWtWbNmU8eOxxMmTBjrdrv1uro6U2trq56dne0BePXVV7ePGTPG3V+1dGyz337b1KlTx7T3G5sxY8bIl156aRfAE088kfbjH//4UNefrXvz588fetFFF9WdbIPNk9XekPOuu+6q6M/HEUL0XdQtG/b7/cydO3fkggULqpcvX74T4IMPPnAcPHjQMmHChMZwhUlX1q9fvxk6f4IH8Hg8WCzhb3u1atWq7QBbtmyx/vnPf87qa6D0hd/vRymFyRT1PTSFEB1E3ZDX8uXLk8xms+p4Nsn06dNbLrjggsYtW7ZYR40aNR5gzZo19qKionFjx44tGD16dMGGDRuOO/L28ssvH1xYWDhu5MiR42+66aaB7bfn5eUV3XTTTQMLCgrGjR49umDt2rV2gPLyctOZZ545aty4cQULFiwY0tNNoTfffPPAyy67bMiZZ5456uKLLx62ZcsW65QpU8YUFBSMKygoGPf2228ntn9tU6dOHXPBBRcMHzZs2Pi5c+cO8/sDfS8XLlyYN2LEiPGjR48u+N///d/8Ez5gJ/Ly8orKysrMt9xyS/6+fftsY8eOLbjuuuvy9+zZYzn11FPHjB07tmDUqFHjV6xY4Tz2Y7t67FWrVjknTZo0Nj8/v+jJJ59MhUAbmmnTpo1u/94988wzKRAIsuHDh4//5je/OXj8+PEFO3bssN55553ZhYWF40aPHl3Q8ft/22235QwdOrRw+vTpo7dt2yZHFQsRJaLuCmX9+vWOiRMnHtfK41gPPfRQ5sKFCytuuOGGw62trZrXe/w5XL/73e8OZGdn+7xeL9OnTx/z8ccfO04//fQWgIyMDO/GjRs33XvvvZn33ntv9vPPP7/nxz/+8cBp06Y13n///WXPPffcgH/84x8ZJ1F3wscff7zZ6XSqhoYG/b333tuakJCgNmzYYLvsssuGl5aWbgLYtGmTY926dTuHDh3qmTJlyti3337becopp7T0pD3+a6+9ljp27NgjgbB3797jnowfeOCB/RdddJGjvaPyz3/+8+xZs2bV3XfffeVer5eGhoajXmScqDV/Z8cBJCQk+F9//fXtaWlp/rKyMvPpp58+dsGCBbUAu3fvtj/++OO7n3nmmb0vv/xy8vbt2+3r16/fpJTi3HPPHfnmm286nU6n/5VXXknbsGHDRo/HwymnnFIwadKkbv+/hRDGi7pA6alp06Y13X///bn79++3XnrppTVFRUVtx77PU089lfbXv/41w+v1aocOHbKUlJTY2wNlwYIFNQBTp05tXrZsWSrARx99lPTyyy9vB7j00kvrrrvuOl9P67ngggtqnU6nAnC73do111wzZOPGjQ5d19mzZ8+RJ/6ioqKmESNGeADGjx/fvGPHDuvMmTMbe9Ief86cOTXHzqF0V9cZZ5zRdN111w31eDz61772tZrp06e3dLz/RK35OzsOwO/3azfeeGP+Rx995NR1ncrKSuv+/fvNALm5ue5Zs2Y1AaxYsSJ59erVyQUFBQUAzc3N+ubNm+0NDQ36hRdeWJuUlOQHOP/882t7+C0WQhgs6oa8ioqKWkpKShK6e7/rr7/+8Kuvvrrd4XD4Z8+ePXrZsmVJHe/fvHmz9Y9//GP2qlWrtm7dunXjzJkz61pbW498P9pb3ZvNZuX1eo80+9D13n3LEhMTj5zZ8utf/zo7KyvLs2nTpo3BV+JHPqnNZjsyjmYymfB6vVp7e/z58+fXLl26NMXlco3qVRGdmD17duPq1au35OXlua+66qphf/zjH9M73n+ix+7sOIDHHnssrbq62rxhw4ZNmzdv3pienu5paWnRARISEvwd3//GG28s27x588bNmzdv3Lt3b+lNN91UBaBp0ltFiGgUdYEyZ86cBrfbrT3wwANHhptWrVqV8Prrrx819r9x40bruHHj2u64447K888/v3bdunWOjvfX1NSYHA6HPy0tzbdv3z5zcXHxgO4e+4wzzmj4y1/+kg7wwgsvJNfX1/dqVrmurs6Um5vrMZlMPPzww+k+34kvdOrq6vTDhw+bLrnkkrpHH31036ZNm7oN1K4MGDDA19TUdOT/fevWrda8vDzPLbfcUvXNb36z6rPPPjvqc5/sY9fV1ZkyMjI8NptNvfbaa0kHDx60dvZ+s2fPrn/66acz6urqdIBdu3ZZDhw4YJ45c2bj66+/ntLY2KjV1NTob7/9dkpvv1YhRHj1ecirJ8t8Q0nXdZYtW7Zj4cKFgx588MEcm812ZNlwx/d7+umn01588cV0s9msMjMzPb/5zW8Odrx/2rRpLYWFhc2jRo0aP3jw4LYpU6Z0e27Kvffee3D+/PnDCwoKxk2bNq0xNze3V8uAb7zxxsr58+ePWLp0aepZZ53V0PHEyc7U1taaQtUePycnxzdlypTGUaNGjZ85c2ZdYWFhy+LFi3PMZrNKSEjwPfvss7v68tjXXnvt4dmzZ48sLCwcN378+OZhw4a1dvZ+F198cf3nn39uP+2008ZC4Orl2Wef3XXWWWc1z5s373BhYeH4vLy8tqlTp/bLeTZCiNCT9vUiJpSUlGRI+3ohjBV1Q15CCCEikwSKEEKIkOhNoPj9fr8swxERI/jzeMJ5KCFE/+tNoJQeOnRogISKiAR+v187dOjQAKDU6FqEiHcnvcrL6/VeW15e/kR5eXkhMmQmjOcHSr1e77VGFyJEvDvpVV5CCCFEZ+QKQwghREhIoAghhAgJCRQhhBAhIYEihBAiJCRQhBBChIQEihBCiJCQQBFCCBESEihCCCFCQgJFCCFESEigCCGECIn/B0BVW2H2B8NkAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 504x504 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from collections import Counter\n",
    "from matplotlib_venn import venn2, venn3\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "sets = Counter()                  \n",
    "sets['100'] = 11   \n",
    "sets['010'] = 201     \n",
    "sets['001'] = 34    \n",
    "sets['110'] = 478     \n",
    "sets['101'] = 0     \n",
    "sets['111'] = 0 \n",
    "labels = ('Cis', 'Trans', 'Non Hits')  \n",
    "plt.figure(figsize=(7,7))\n",
    "ax = plt.gca()\n",
    "colors = ['darkviolet','deepskyblue','blue']\n",
    "v = venn3(subsets=sets, set_labels=('','',''), ax=ax,set_colors=\n",
    "          colors,alpha=0.7)   \n",
    "\n",
    "\n",
    "h = [] \n",
    "for i in sets:\n",
    "    h.append(v.get_patch_by_id(i)) \n",
    "l = ['Trans Hits only','Cis Hits only','Non Hits','Cis and Trans Hits shared ','',\n",
    "     '','']\n",
    "ax.legend(handles=h, labels=l, title=\"Legend\",loc=3)  \n",
    "#plt.title('Number of SNPs Hits ')\n",
    "output_file = \"~/lastversion_gwas_analysis/overlapping_result/GWAS_Hits.png\"\n",
    "plt.savefig(output_file, dpi=700, facecolor='w')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35db6b90",
   "metadata": {},
   "source": [
    "## Find common Trans genes \n",
    "The phenotype files of the 3 most significant trans genes in the duplicated files are determined. SNPs that cause these trans genes are determined from the results of gwas resulting from this phenotype. As an example, SNPs that cause significant trans genes in common are identified in files AT2G16365_AT1G13430 and AT1G13430_AT2G16365. In this way, these common SNPs are determined according to the results of gwas causing AT2G18735, AT1G08840 and AT2G21830 trans genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea666a76",
   "metadata": {},
   "outputs": [],
   "source": [
    "%% R\n",
    "#AT1G08840\n",
    "#\"AT1G13430_AT2G16365.csv\"\n",
    "setwd(\"~/trans_genes/AT1G08840\")\n",
    "Y <- list()\n",
    "csvfiles<-list.files(pattern = \"*.csv\")\n",
    "Y <- lapply(csvfiles, function(x) {\n",
    "  file <- read.csv(x)\n",
    "  return(file)\n",
    "})\n",
    "for (i in 1:length((csvfiles))){\n",
    "  Y[[i]]<-mutate(Y[[i]], SNP=paste(CHR, sep=\"- \", POS))\n",
    "}\n",
    "d<-a[which(\"AT1G13430_AT2G16365.csv\"==a$file_1),]\n",
    "d<-unique(d$gens_trans)\n",
    "d<-as.data.frame(d)\n",
    "colnames(d)<-\"gene\"\n",
    "mylist<-c()\n",
    "for (i in 1:nrow(d)){\n",
    "  e<-SNPs[which(d$gene[i]==SNPs$gene),]\n",
    "  mylist[[length(mylist)+1]]<-e\n",
    "  cat(i,'\\n')\n",
    "}\n",
    "\n",
    "for (i in 1:length((mylist))){\n",
    "  Z<-merge(mylist[[i]][,c(4,5)],Y[[1]][,c(5,14)],by='SNP')\n",
    "  b[i]<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "  b<-as.data.frame(b)\n",
    "}  \n",
    "  Z<-merge(mylist[[1]][,c(4,5)],Y[[1]][,c(5,14)],by='SNP')\n",
    "  b<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "write.csv( b,\"~/lastversion_gwas_analysis/PermGWAS_Scripts/AC_1_0_files/SNPsAT1G08840_.csv\", row.names = TRUE)\n",
    "  \n",
    "#\"AT2G16365_AT1G13430.csv\" \n",
    "d<-a[which(\"AT2G16365_AT1G13430.csv\"==a$file_1),]\n",
    "d<-unique(d$gens_trans)\n",
    "d<-as.data.frame(d)\n",
    "colnames(d)<-\"gene\"\n",
    "mylist<-c()\n",
    "for (i in 1:nrow(d)){\n",
    "  e<-SNPs[which(d$gene[i]==SNPs$gene),]\n",
    "  mylist[[length(mylist)+1]]<-e\n",
    "  cat(i,'\\n')\n",
    "}\n",
    "b1<-c()\n",
    "for (i in 1:length((mylist))){\n",
    "  Z<-merge(mylist[[i]][,c(4,5)],Y[[2]][,c(5,14)],by='SNP')\n",
    "  b1[i]<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "  b1<-as.data.frame(b1)\n",
    "}  \n",
    "\n",
    "#\"AT2G16365_AT3G47010.csv\"\n",
    "d<-a[which(\"AT2G16365_AT3G47010.csv\"==a$file_1),]\n",
    "d<-unique(d$gens_trans)\n",
    "d<-as.data.frame(d)\n",
    "colnames(d)<-\"gene\"\n",
    "mylist<-c()\n",
    "for (i in 1:nrow(d)){\n",
    "  e<-SNPs[which(d$gene[i]==SNPs$gene),]\n",
    "  mylist[[length(mylist)+1]]<-e\n",
    "  cat(i,'\\n')\n",
    "}\n",
    "b2<-c()\n",
    "for (i in 1:length((mylist))){\n",
    "  Z<-merge(mylist[[i]][,c(4,5)],Y[[3]][,c(5,14)],by='SNP')\n",
    "  b2[i]<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "  b2<-as.data.frame(b2)\n",
    "} \n",
    "#\"AT3G47010_AT2G16365.csv\" \n",
    "d<-a[which(\"AT3G47010_AT2G16365.csv\"==a$file_1),]\n",
    "d<-unique(d$gens_trans)\n",
    "d<-as.data.frame(d)\n",
    "colnames(d)<-\"gene\"\n",
    "mylist<-c()\n",
    "for (i in 1:nrow(d)){\n",
    "  e<-SNPs[which(d$gene[i]==SNPs$gene),]\n",
    "  mylist[[length(mylist)+1]]<-e\n",
    "  cat(i,'\\n')\n",
    "}\n",
    "b3<-c()\n",
    "for (i in 1:length((mylist))){\n",
    "  Z<-merge(mylist[[i]][,c(4,5)],Y[[4]][,c(5,14)],by='SNP')\n",
    "  b3[i]<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "  b3<-as.data.frame(b3)\n",
    "  \n",
    "} \n",
    "#\"AT3G47010_AT4G00970.csv\"\n",
    "d<-a[which(\"AT3G47010_AT4G00970.csv\"==a$file_1),]\n",
    "d<-unique(d$gens_trans)\n",
    "d<-as.data.frame(d)\n",
    "colnames(d)<-\"gene\"\n",
    "mylist<-c()\n",
    "for (i in 1:nrow(d)){\n",
    "  e<-SNPs[which(d$gene[i]==SNPs$gene),]\n",
    "  mylist[[length(mylist)+1]]<-e\n",
    "  cat(i,'\\n')\n",
    "}\n",
    "b4<-c()\n",
    "for (i in 1:length((mylist))){\n",
    "  Z<-merge(mylist[[i]][,c(4,5)],Y[[5]][,c(5,14)],by='SNP')\n",
    "  b4[i]<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "  b4<-as.data.frame(b4)\n",
    "  \n",
    "} \n",
    "#\"AT4G00970_AT3G47010.csv\" \n",
    "d<-a[which(\"AT4G00970_AT3G47010.csv\"==a$file_1),]\n",
    "d<-unique(d$gens_trans)\n",
    "d<-as.data.frame(d)\n",
    "colnames(d)<-\"gene\"\n",
    "mylist<-c()\n",
    "for (i in 1:nrow(d)){\n",
    "  e<-SNPs[which(d$gene[i]==SNPs$gene),]\n",
    "  mylist[[length(mylist)+1]]<-e\n",
    "  cat(i,'\\n')\n",
    "}\n",
    "b5<-c()\n",
    "for (i in 1:length((mylist))){\n",
    "  Z<-merge(mylist[[i]][,c(4,5)],Y[[6]][,c(5,14)],by='SNP')\n",
    "  b5[i]<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "  b5<-as.data.frame(b5)\n",
    "} \n",
    "\n",
    "#AT2G18735\n",
    "setwd(\"/home/s417377/trans_genes/AT2G18735\")\n",
    "Y <- list()\n",
    "csvfiles<-list.files(pattern = \"*.csv\")\n",
    "Y <- lapply(csvfiles, function(x) {\n",
    "  file <- read.csv(x)\n",
    "  return(file)\n",
    "})\n",
    "for (i in 1:length((csvfiles))){\n",
    "  Y[[i]]<-mutate(Y[[i]], SNP=paste(CHR, sep=\"- \", POS))\n",
    "}\n",
    "#\"AT2G29710_AT2G24600.csv\"\n",
    "d<-a[which(\"AT2G29710_AT2G24600.csv\"==a$file_1),]\n",
    "d<-unique(d$gens_trans)\n",
    "d<-as.data.frame(d)\n",
    "colnames(d)<-\"gene\"\n",
    "mylist<-c()\n",
    "for (i in 1:nrow(d)){\n",
    "  e<-SNPs[which(d$gene[i]==SNPs$gene),]\n",
    "  mylist[[length(mylist)+1]]<-e\n",
    "  cat(i,'\\n')\n",
    "}\n",
    "f<-c()\n",
    "for (i in 1:length((mylist))){\n",
    "  Z<-merge(mylist[[i]][,c(4,5)],Y[[1]][,c(5,14)],by='SNP')\n",
    "  f[i]<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "  f<-as.data.frame(f)\n",
    "} \n",
    "#\"AT2G24600_AT2G29710.csv\" \n",
    "d<-a[which(\"AT2G24600_AT2G29710.csv\"==a$file_1),]\n",
    "d<-unique(d$gens_trans)\n",
    "d<-as.data.frame(d)\n",
    "colnames(d)<-\"gene\"\n",
    "mylist<-c()\n",
    "for (i in 1:nrow(d)){\n",
    "  e<-SNPs[which(d$gene[i]==SNPs$gene),]\n",
    "  mylist[[length(mylist)+1]]<-e\n",
    "  cat(i,'\\n')\n",
    "}\n",
    "f1<-c()\n",
    "for (i in 1:length((mylist))){\n",
    "  Z<-merge(mylist[[i]][,c(4,5)],Y[[2]][,c(5,14)],by='SNP')\n",
    "  f1[i]<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "  f1<-as.data.frame(f1)\n",
    "} \n",
    "#\"AT2G24600_AT4G00970.csv\"\n",
    "d<-a[which(\"AT2G24600_AT4G00970.csv\"==a$file_1),]\n",
    "d<-unique(d$gens_trans)\n",
    "d<-as.data.frame(d)\n",
    "colnames(d)<-\"gene\"\n",
    "mylist<-c()\n",
    "for (i in 1:nrow(d)){\n",
    "  e<-SNPs[which(d$gene[i]==SNPs$gene),]\n",
    "  mylist[[length(mylist)+1]]<-e\n",
    "  cat(i,'\\n')\n",
    "}\n",
    "f2<-c()\n",
    "for (i in 1:length((mylist))){\n",
    "  Z<-merge(mylist[[i]][,c(4,5)],Y[[3]][,c(5,14)],by='SNP')\n",
    "  f2[i]<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "  f2<-as.data.frame(f2)\n",
    "} \n",
    "#\"AT3G12020_AT3G19040.csv\" \n",
    "d<-a[which(\"AT3G12020_AT3G19040.csv\"==a$file_1),]\n",
    "d<-unique(d$gens_trans)\n",
    "d<-as.data.frame(d)\n",
    "colnames(d)<-\"gene\"\n",
    "mylist<-c()\n",
    "for (i in 1:nrow(d)){\n",
    "  e<-SNPs[which(d$gene[i]==SNPs$gene),]\n",
    "  mylist[[length(mylist)+1]]<-e\n",
    "  cat(i,'\\n')\n",
    "}\n",
    "f3<-c()\n",
    "for (i in 1:length((mylist))){\n",
    "  Z<-merge(mylist[[i]][,c(4,5)],Y[[4]][,c(5,14)],by='SNP')\n",
    "  f3[i]<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "  f3<-as.data.frame(f3)\n",
    "} \n",
    "#\"AT3G19040_AT3G12020.csv\"\n",
    "d<-a[which(\"AT3G19040_AT3G12020.csv\"==a$file_1),]\n",
    "d<-unique(d$gens_trans)\n",
    "d<-as.data.frame(d)\n",
    "colnames(d)<-\"gene\"\n",
    "mylist<-c()\n",
    "for (i in 1:nrow(d)){\n",
    "  e<-SNPs[which(d$gene[i]==SNPs$gene),]\n",
    "  mylist[[length(mylist)+1]]<-e\n",
    "  cat(i,'\\n')\n",
    "}\n",
    "f4<-c()\n",
    "for (i in 1:length((mylist))){\n",
    "  Z<-merge(mylist[[i]][,c(4,5)],Y[[5]][,c(5,14)],by='SNP')\n",
    "  f4[i]<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "  f4<-as.data.frame(f4)\n",
    "} \n",
    "#\"AT4G00970_AT2G24600.csv\" \n",
    "d<-a[which(\"AT4G00970_AT2G24600.csv\"==a$file_1),]\n",
    "d<-unique(d$gens_trans)\n",
    "d<-as.data.frame(d)\n",
    "colnames(d)<-\"gene\"\n",
    "mylist<-c()\n",
    "for (i in 1:nrow(d)){\n",
    "  e<-SNPs[which(d$gene[i]==SNPs$gene),]\n",
    "  mylist[[length(mylist)+1]]<-e\n",
    "  cat(i,'\\n')\n",
    "}\n",
    "f5<-c()\n",
    "for (i in 1:length((mylist))){\n",
    "  Z<-merge(mylist[[i]][,c(4,5)],Y[[6]][,c(5,14)],by='SNP')\n",
    "  f5[i]<-as.data.frame(paste(as.character(na.omit((Z$SNP))),collapse=','))\n",
    "  f5<-as.data.frame(f5)\n",
    "} "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f55f9b2e",
   "metadata": {},
   "source": [
    "## Manhattan Plot\n",
    "The code used to visualize the GWAS results was written by Matthew Flickinger and modified according to the permgwas results \"https://genome.sph.umich.edu/wiki/Code_Sample:_Generating_Manhattan_Plots_in_R\" (manhattanplot2.r)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d94e8b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "%% R\n",
    "library(lattice)\n",
    "manhattan.plot<-function(chr, pos, pvalue, \n",
    "                         sig.level=NA, annotate=NULL, ann.default=list(),\n",
    "                         should.thin=T, thin.pos.places=2, thin.logp.places=2, \n",
    "                         xlab=\"Chromosome\", ylab=expression(-log[10](p-value)),\n",
    "                         col=c(\"black\",\"gray\"), panel.extra=NULL, pch=20, cex=0.8,...) {\n",
    "  \n",
    "  if (length(chr)==0) stop(\"chromosome vector is empty\")\n",
    "  if (length(pos)==0) stop(\"position vector is empty\")\n",
    "  if (length(pvalue)==0) stop(\"pvalue vector is empty\")\n",
    "  \n",
    "  #make sure we have an ordered factor\n",
    "  if(!is.ordered(chr)) {\n",
    "    chr <- ordered(chr)\n",
    "  } else {\n",
    "    chr <- chr[,drop=T]\n",
    "  }\n",
    "  \n",
    "  #make sure positions are in kbp\n",
    "  if (any(pos>1e6)) pos<-pos/1e6;\n",
    "  \n",
    "  #calculate absolute genomic position\n",
    "  #from relative chromosomal positions\n",
    "  posmin <- tapply(pos,chr, min);\n",
    "  posmax <- tapply(pos,chr, max);\n",
    "  posshift <- head(c(0,cumsum(posmax)),-1);\n",
    "  names(posshift) <- levels(chr)\n",
    "  genpos <- pos + posshift[chr];\n",
    "  getGenPos<-function(cchr, cpos) {\n",
    "    p<-posshift[as.character(cchr)]+cpos\n",
    "    return(p)\n",
    "  }\n",
    "  \n",
    "  #parse annotations\n",
    "  grp <- NULL\n",
    "  ann.settings <- list()\n",
    "  label.default<-list(x=\"peak\",y=\"peak\",adj=NULL, pos=3, offset=0.5, \n",
    "                      col=NULL, fontface=NULL, fontsize=NULL, show=F)\n",
    "  parse.label<-function(rawval, groupname) {\n",
    "    r<-list(text=groupname)\n",
    "    if(is.logical(rawval)) {\n",
    "      if(!rawval) {r$show <- F}\n",
    "    } else if (is.character(rawval) || is.expression(rawval)) {\n",
    "      if(nchar(rawval)>=1) {\n",
    "        r$text <- rawval\n",
    "      }\n",
    "    } else if (is.list(rawval)) {\n",
    "      r <- modifyList(r, rawval)\n",
    "    }\n",
    "    return(r)\n",
    "  }\n",
    "  \n",
    "  if(!is.null(annotate)) {\n",
    "    if (is.list(annotate)) {\n",
    "      grp <- annotate[[1]]\n",
    "    } else {\n",
    "      grp <- annotate\n",
    "    } \n",
    "    if (!is.factor(grp)) {\n",
    "      grp <- factor(grp)\n",
    "    }\n",
    "  } else {\n",
    "    grp <- factor(rep(1, times=length(pvalue)))\n",
    "  }\n",
    "  \n",
    "  ann.settings<-vector(\"list\", length(levels(grp)))\n",
    "  ann.settings[[1]]<-list(pch=pch, col=col, cex=cex, fill=col, label=label.default)\n",
    "  \n",
    "  if (length(ann.settings)>1) { \n",
    "    lcols<-trellis.par.get(\"superpose.symbol\")$col \n",
    "    lfills<-trellis.par.get(\"superpose.symbol\")$fill\n",
    "    for(i in 2:length(levels(grp))) {\n",
    "      ann.settings[[i]]<-list(pch=pch, \n",
    "                              col=lcols[(i-2) %% length(lcols) +1 ], \n",
    "                              fill=lfills[(i-2) %% length(lfills) +1 ], \n",
    "                              cex=cex, label=label.default);\n",
    "      ann.settings[[i]]$label$show <- T\n",
    "    }\n",
    "    names(ann.settings)<-levels(grp)\n",
    "  }\n",
    "  for(i in 1:length(ann.settings)) {\n",
    "    if (i>1) {ann.settings[[i]] <- modifyList(ann.settings[[i]], ann.default)}\n",
    "    ann.settings[[i]]$label <- modifyList(ann.settings[[i]]$label, \n",
    "                                          parse.label(ann.settings[[i]]$label, levels(grp)[i]))\n",
    "  }\n",
    "  if(is.list(annotate) && length(annotate)>1) {\n",
    "    user.cols <- 2:length(annotate)\n",
    "    ann.cols <- c()\n",
    "    if(!is.null(names(annotate[-1])) && all(names(annotate[-1])!=\"\")) {\n",
    "      ann.cols<-match(names(annotate)[-1], names(ann.settings))\n",
    "    } else {\n",
    "      ann.cols<-user.cols-1\n",
    "    }\n",
    "    for(i in seq_along(user.cols)) {\n",
    "      if(!is.null(annotate[[user.cols[i]]]$label)) {\n",
    "        annotate[[user.cols[i]]]$label<-parse.label(annotate[[user.cols[i]]]$label, \n",
    "                                                    levels(grp)[ann.cols[i]])\n",
    "      }\n",
    "      ann.settings[[ann.cols[i]]]<-modifyList(ann.settings[[ann.cols[i]]], \n",
    "                                              annotate[[user.cols[i]]])\n",
    "    }\n",
    "  }\n",
    "  rm(annotate)\n",
    "  \n",
    "  #reduce number of points plotted\n",
    "  if(should.thin) {\n",
    "    thinned <- unique(data.frame(\n",
    "      logp=round(-log10(pvalue),thin.logp.places), \n",
    "      pos=round(genpos,thin.pos.places), \n",
    "      chr=chr,\n",
    "      grp=grp)\n",
    "    )\n",
    "    logp <- thinned$logp\n",
    "    genpos <- thinned$pos\n",
    "    chr <- thinned$chr\n",
    "    grp <- thinned$grp\n",
    "    rm(thinned)\n",
    "  } else {\n",
    "    logp <- -log10(pvalue)\n",
    "  }\n",
    "  rm(pos, pvalue)\n",
    "  gc()\n",
    "  \n",
    "  #custom axis to print chromosome names\n",
    "  axis.chr <- function(side,...) {\n",
    "    if(side==\"bottom\") {\n",
    "      panel.axis(side=side, outside=T,\n",
    "                 at=((posmax+posmin)/2+posshift),\n",
    "                 labels=levels(chr), \n",
    "                 ticks=F, rot=0,\n",
    "                 check.overlap=F\n",
    "      )\n",
    "    } else if (side==\"top\" || side==\"right\") {\n",
    "      panel.axis(side=side, draw.labels=F, ticks=F);\n",
    "    }\n",
    "    else {\n",
    "      axis.default(side=side,...);\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  #make sure the y-lim covers the range (plus a bit more to look nice)\n",
    "  prepanel.chr<-function(x,y,...) { \n",
    "    A<-list();\n",
    "    maxy<-ceiling(max(y, ifelse(!is.na(sig.level), -log10(sig.level), 0)))+.5;\n",
    "    A$ylim=c(0,maxy);\n",
    "    A;\n",
    "  }\n",
    "  \n",
    "  xyplot(logp~genpos, chr=chr, groups=grp,\n",
    "         axis=axis.chr, ann.settings=ann.settings, \n",
    "         prepanel=prepanel.chr, scales=list(axs=\"i\"),\n",
    "         panel=function(x, y, ..., getgenpos) {\n",
    "           if(!is.na(sig.level)) {\n",
    "             #add significance line (if requested)\n",
    "             panel.abline(h=-log10(sig.level), lty=2);\n",
    "           }\n",
    "           panel.superpose(x, y, ..., getgenpos=getgenpos);\n",
    "           if(!is.null(panel.extra)) {\n",
    "             panel.extra(x,y, getgenpos, ...)\n",
    "           }\n",
    "         },\n",
    "         panel.groups = function(x,y,..., subscripts, group.number) {\n",
    "           A<-list(...)\n",
    "           #allow for different annotation settings\n",
    "           gs <- ann.settings[[group.number]]\n",
    "           A$col.symbol <- gs$col[(as.numeric(chr[subscripts])-1) %% length(gs$col) + 1]    \n",
    "           A$cex <- gs$cex[(as.numeric(chr[subscripts])-1) %% length(gs$cex) + 1]\n",
    "           A$pch <- gs$pch[(as.numeric(chr[subscripts])-1) %% length(gs$pch) + 1]\n",
    "           A$fill <- gs$fill[(as.numeric(chr[subscripts])-1) %% length(gs$fill) + 1]\n",
    "           A$x <- x\n",
    "           A$y <- y\n",
    "           do.call(\"panel.xyplot\", A)\n",
    "           #draw labels (if requested)\n",
    "           if(gs$label$show) {\n",
    "             gt<-gs$label\n",
    "             names(gt)[which(names(gt)==\"text\")]<-\"labels\"\n",
    "             gt$show<-NULL\n",
    "             if(is.character(gt$x) | is.character(gt$y)) {\n",
    "               peak = which.max(y)\n",
    "               center = mean(range(x))\n",
    "               if (is.character(gt$x)) {\n",
    "                 if(gt$x==\"peak\") {gt$x<-x[peak]}\n",
    "                 if(gt$x==\"center\") {gt$x<-center}\n",
    "               }\n",
    "               if (is.character(gt$y)) {\n",
    "                 if(gt$y==\"peak\") {gt$y<-y[peak]}\n",
    "               }\n",
    "             }\n",
    "             if(is.list(gt$x)) {\n",
    "               gt$x<-A$getgenpos(gt$x[[1]],gt$x[[2]])\n",
    "             }\n",
    "             do.call(\"panel.text\", gt)\n",
    "           }\n",
    "         },\n",
    "         xlab=xlab, ylab=ylab, \n",
    "         panel.extra=panel.extra, getgenpos=getGenPos, ...\n",
    "  );\n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98d3df8e",
   "metadata": {},
   "source": [
    "## Visualization\n",
    "Manhattan plots  are drawn according to the  common SNPs are found of gwas causing AT2G18735, AT1G08840 and AT2G21830 trans genes (manhattanplot.r)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e0d2f75",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "library(tidyverse)\n",
    "library(dplyr)\n",
    "vignette('qqman')\n",
    "ara11<-read.csv(\"~/lastversion_gwas_analysis/ara11_50.csv\")\n",
    "setwd(\"/home/s417377/trans_genes/AT2G18735/maf_filtered/\")\n",
    "Y <- list()\n",
    "csvfiles<-list.files(pattern = \"*.csv\")\n",
    "Y <- lapply(csvfiles, function(x) {\n",
    "  file <- read.csv(x)\n",
    "  return(file)\n",
    "})\n",
    "for (i in 1:length((csvfiles))){\n",
    "  Y[[i]]<-mutate(Y[[i]], SNP=paste(CHR, sep=\"- \", POS))\n",
    "}\n",
    "for (i in 1:length((csvfiles))){\n",
    "  colnames(Y[[i]])[4]<-\"chr\"\n",
    "  colnames(Y[[i]])[5]<-\"pos\"\n",
    "  colnames(Y[[i]])[6]<-\"pvalue\"\n",
    "}\n",
    "#\"AT2G29710_AT2G24600\"\n",
    "#make annotation factor\n",
    "ann<-rep(1, length(Y[[1]]$pvalue))\n",
    "ann[with(Y[[1]], chr==2  &  pos>=12648573 &  pos<= 12750343)]<-2\n",
    "ann[with(Y[[1]], chr==2  &  pos==9137974)]<-3\n",
    "ann[with(Y[[1]], chr==2  &  pos==9303977)]<-4\n",
    "ann[with(Y[[1]], chr==2  &  pos==17833766)]<-5\n",
    "ann[with(Y[[1]], chr==4  &  pos==840051)]<-6\n",
    "ann<-factor(ann, levels=1:6, labels=c(\"\",\"-\", \"9137974\",\"9303977\",\"17833766\",\"840051\"))\n",
    "#draw plot with annotation\n",
    "manhattan.plot(Y[[1]]$chr, Y[[1]]$pos, Y[[1]]$pvalue, annotate=ann, ylim=c(0,50))\n",
    "#\"AT2G24600_AT2G29710\"\n",
    "#make annotation factor\n",
    "ann<-rep(1, length(Y[[3]]$pvalue))\n",
    "ann[with(Y[[3]], chr==2  &  pos>=10402034 &  pos<= 10504593)]<-2\n",
    "ann[with(Y[[3]], chr==2  &  pos==9137974)]<-3\n",
    "ann[with(Y[[3]], chr==2  &  pos==9303977)]<-4\n",
    "ann[with(Y[[3]], chr==2  &  pos==17833769)]<-5\n",
    "ann[with(Y[[3]], chr==4  &  pos==840051)]<-6\n",
    "ann<-factor(ann, levels=1:6, labels=c(\"\",\"-\", \"9137974\",\"9303977\",\"17833769\",\"840051\"))\n",
    "#draw plot with annotation\n",
    "manhattan.plot(Y[[3]]$chr, Y[[3]]$pos, Y[[3]]$pvalue, annotate=ann, ylim=c(0,50))\n",
    "#\"AT2G24600_AT4G00970.csv\"\n",
    "#make annotation factor\n",
    "ann<-rep(1, length(Y[[2]]$pvalue))\n",
    "ann[with(Y[[2]], chr==4  &  pos>=368327 &  pos<= 471885)]<-2\n",
    "ann[with(Y[[2]], chr==2  &  pos==8122148)]<-3\n",
    "ann[with(Y[[2]], chr==2  &  pos==9345189)]<-4\n",
    "ann[with(Y[[2]], chr==2  &  pos==17833784)]<-5\n",
    "ann[with(Y[[2]], chr==4  &  pos==839000)]<-6\n",
    "ann<-factor(ann, levels=1:6, labels=c(\"\",\"-\", \"8122148\",\"9345189\",\"17833784\",\"839000\"))\n",
    "#draw plot with annotation\n",
    "manhattan.plot(Y[[2]]$chr, Y[[2]]$pos, Y[[2]]$pvalue, annotate=ann, ylim=c(0,50))\n",
    "#\"AT4G00970_AT2G24600.csv\"\n",
    "#make annotation factor\n",
    "ann<-rep(1, length(Y[[6]]$pvalue))\n",
    "ann[with(Y[[6]], chr==2  &  pos>=10402034 &  pos<= 10504593)]<-2\n",
    "ann[with(Y[[6]], chr==2  &  pos==8122148)]<-3\n",
    "ann[with(Y[[6]], chr==2  &  pos==9345189)]<-4\n",
    "ann[with(Y[[6]], chr==2  &  pos==17833784)]<-5\n",
    "ann[with(Y[[6]], chr==4  &  pos==839000)]<-6\n",
    "ann<-factor(ann, levels=1:6, labels=c(\"\",\"-\", \"8122148\",\"9345189\",\"17833784\",\"839000\"))\n",
    "#draw plot with annotation\n",
    "manhattan.plot(Y[[6]]$chr, Y[[6]]$pos, Y[[6]]$pvalue, annotate=ann, ylim=c(0,50))\n",
    "#\"AT3G12020_AT3G19040.csv\" \n",
    "ann<-rep(1, length(Y[[4]]$pvalue))\n",
    "ann[with(Y[[4]], chr==3  &  pos>=6517021 &  pos<= 6517021)]<-2\n",
    "ann[with(Y[[4]], chr==2  &  pos==8122723)]<-3\n",
    "ann<-factor(ann, levels=1:3, labels=c(\"\",\"-\", \"8122723\"))\n",
    "#draw plot with annotation\n",
    "manhattan.plot(Y[[4]]$chr, Y[[4]]$pos, Y[[4]]$pvalue, annotate=ann, ylim=c(0,50))\n",
    "#\"AT3G19040_AT3G12020.csv\" \n",
    "#make annotation factor\n",
    "ann<-rep(1, length(Y[[5]]$pvalue))\n",
    "ann[with(Y[[5]], chr==3  &  pos>=3775980 &  pos<= 3885146)]<-2\n",
    "ann[with(Y[[5]], chr==2  &  pos==8122723)]<-3\n",
    "ann<-factor(ann, levels=1:3, labels=c(\"\",\"-\", \"8122723\"))\n",
    "#draw plot with annotation\n",
    "manhattan.plot(Y[[5]]$chr, Y[[5]]$pos, Y[[5]]$pvalue, annotate=ann, ylim=c(0,50))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d417787f",
   "metadata": {},
   "source": [
    "## Save Jupyter Notebook session"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "af50af1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import dataframe_image as dfi\n",
    "dfi.convert('/home/s417377/PermGwas.ipynb',\n",
    "                to='pdf',\n",
    "                use='latex',\n",
    "                center_df=True,\n",
    "                max_rows=30,\n",
    "                max_cols=10,\n",
    "                execute=False,\n",
    "                save_notebook=False,\n",
    "                limit=None,\n",
    "                document_name=None,\n",
    "                table_conversion='chrome'\n",
    "                #chrome_path=None,\n",
    "                #latex_command=None,\n",
    "                #output_dir=None,\n",
    "                )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8a10b68",
   "metadata": {},
   "source": [
    "## Add accesion ids\n",
    "(accesion.r)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7387e5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "setwd(\"~/lastversion_gwas_analysis/PermGWAS_OverCo/\")\n",
    "setwd(\"~/lastversion_gwas_analysis/PermGWAS_Scripts/AC_1_0_files/\")\n",
    "Y <- list()\n",
    "csvfiles<-list.files(pattern = \"*.csv\")\n",
    "Y <- lapply(csvfiles, function(x) {\n",
    "  file <- read.csv(x)\n",
    "  return(file)\n",
    "})\n",
    "\n",
    "d<-read.csv(\"~/lastversion_gwas_analysis/overlapping_result/perm_overlapping_genes.csv\")\n",
    "d['accesion']<-NA\n",
    "for(i in 1:nrow(d)){\n",
    "  a<-Y[[i]][which(Y[[i]]$Phenotype==1),]\n",
    "  d[i,15]<-paste(as.character(na.omit(a$X)),collapse=',')\n",
    "}\n",
    "write.csv(d, \"~/lastversion_gwas_analysis/overlapping_result/accesions_overlapping_genes.csv\", row.names = FALSE)\n",
    "\n",
    "\n",
    "\n",
    "  \n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4c8050d",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(tidyverse)\n",
    "library(fs)\n",
    "library(readxl)\n",
    "library(dplyr)\n",
    "load(\"/storage/full-share/evolgen/snps_2029.rda\")\n",
    "small_gene<-read_excel(\"~/lastversion_gwas_analysis/Over_Occ/stringent_hits.xlsx\")\n",
    "#setwd(\"/home/s417377/lastversion_gwas_analysis/PermGWAS_Scripts/permgwas_premature/\")\n",
    "#b<-d[which(d$SNP==SNPs$SNP),]\n",
    "#d[which(\"AT2G27340\"==SNPs$SNP)&SNPs$prematSTOP==1,]\n",
    "#cc<-SNPs[which(\"AT2G27340\"==SNPs$SNP)&SNPs$prematSTOP==1,]\n",
    "setwd(\"/home/s417377/PermGWAS/\")\n",
    "list_of_files <- list.files(path = \"/home/s417377/PermGWAS\",\n",
    "                            recursive = TRUE,\n",
    "                            pattern = \"*gwasp_\",\n",
    "                            full.names = TRUE)\n",
    "Y <- lapply(list_of_files, function(x) {\n",
    "  file <- read.csv(x)\n",
    "  return(file)\n",
    "})\n",
    "for (i in 1:length((list_of_files))){\n",
    "  Y[[i]]<-mutate(Y[[i]], SNP=paste(CHR, sep=\"- \", POS))\n",
    "}\n",
    "setwd(\"/home/s417377/lastversion_gwas_analysis/PermGWAS_Scripts/permgwas_premature/\")\n",
    "for (i in 1:length((list_of_files))){\n",
    "  a<-SNPs[which(small_gene$...3[[i]]==SNPs$gene),]\n",
    "  a<-a[2:16]\n",
    "  b<-a[which(a$prematSTOP==1),]\n",
    "  for(x in 1:length((list_of_files))){\n",
    "    c<-b[which(b$SNP[x]==Y[[i]][\"SNP\"]),]\n",
    "    write.csv(c, file=paste( \"premature_\",small_gene$gen_name[[i]],\".csv\", sep=\"\"))\n",
    "  }\n",
    "}\n",
    "\n",
    "#a<-SNPs[which(small_gene$...3[[1]]==SNPs$gene),]\n",
    "#a<-a[2:16]\n",
    "#b<-a[which(a$prematSTOP==1),]\n",
    "#c<-b[which(b$SNP[4]==Y[[1]][\"SNP\"]),]\n",
    "#for(x in 1:length((list_of_files))){\n",
    " # c<-b[which(b$SNP[x]==Y[[1]][\"SNP\"]),]\n",
    "  #write.csv(c, file=paste( \"premature_\",small_gene$gen_name[[i]],\".csv\", sep=\"\"))\n",
    "#}\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
